<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.54">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>3&nbsp; Reading and Writing Files – Five days of R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./day4-manipulating-your-data.html" rel="next">
<link href="./day2-complex-structures.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://hypothes.is/embed.js"></script>
<script>
  window.document.addEventListener("DOMContentLoaded", function (_event) {
    document.body.classList.add('hypothesis-enabled');
  });
</script>


<link rel="stylesheet" href="custom.css">
</head>

<body class="nav-sidebar floating nav-fixed slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Five days of R</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="./index.html" aria-current="page"> 
<span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day1-intro-to-r.html"> 
<span class="menu-text">One</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day2-complex-structures.html"> 
<span class="menu-text">Two</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day3-reading-your-data.html"> 
<span class="menu-text">Three</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day4-manipulating-your-data.html"> 
<span class="menu-text">Four</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day5-visualization-and-statistics.html"> 
<span class="menu-text">Five</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./solutions.html"> 
<span class="menu-text">Solutions</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/bihealth/RCrashcourse-book/" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./day3-reading-your-data.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Reading and Writing Files</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day1-intro-to-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day2-complex-structures.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Complex data structures</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day3-reading-your-data.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Reading and Writing Files</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day4-manipulating-your-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Manipulating data frames</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day5-visualization-and-statistics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">R markdown, basic statistics and visualizations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix-wide-long.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendix: Wide vs long data</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix-more-stats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendix: More statistics and visualizations</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Solutions to Exercises</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#aims-for-today" id="toc-aims-for-today" class="nav-link active" data-scroll-target="#aims-for-today"><span class="header-section-number">3.1</span> Aims for today</a></li>
  <li><a href="#reading-data" id="toc-reading-data" class="nav-link" data-scroll-target="#reading-data"><span class="header-section-number">3.2</span> Reading data</a>
  <ul class="collapse">
  <li><a href="#data-types" id="toc-data-types" class="nav-link" data-scroll-target="#data-types"><span class="header-section-number">3.2.1</span> Data types</a></li>
  <li><a href="#where-is-my-file-relative-and-absolute-paths" id="toc-where-is-my-file-relative-and-absolute-paths" class="nav-link" data-scroll-target="#where-is-my-file-relative-and-absolute-paths"><span class="header-section-number">3.2.2</span> Where is my file? Relative and absolute paths</a></li>
  <li><a href="#reading-with-relative-paths" id="toc-reading-with-relative-paths" class="nav-link" data-scroll-target="#reading-with-relative-paths"><span class="header-section-number">3.2.3</span> Reading with relative paths</a></li>
  <li><a href="#more-on-relative-paths" id="toc-more-on-relative-paths" class="nav-link" data-scroll-target="#more-on-relative-paths"><span class="header-section-number">3.2.4</span> More on relative paths</a></li>
  <li><a href="#using-rstudio-to-import-files" id="toc-using-rstudio-to-import-files" class="nav-link" data-scroll-target="#using-rstudio-to-import-files"><span class="header-section-number">3.2.5</span> Using RStudio to import files</a></li>
  <li><a href="#reading-excel-files" id="toc-reading-excel-files" class="nav-link" data-scroll-target="#reading-excel-files"><span class="header-section-number">3.2.6</span> Reading Excel files</a></li>
  </ul></li>
  <li><a href="#diagnosing-and-cleaning-data" id="toc-diagnosing-and-cleaning-data" class="nav-link" data-scroll-target="#diagnosing-and-cleaning-data"><span class="header-section-number">3.3</span> Diagnosing and cleaning data</a>
  <ul class="collapse">
  <li><a href="#sec-diagnosing" id="toc-sec-diagnosing" class="nav-link" data-scroll-target="#sec-diagnosing"><span class="header-section-number">3.3.1</span> Diagnosing datasets</a></li>
  <li><a href="#sec-skim" id="toc-sec-skim" class="nav-link" data-scroll-target="#sec-skim"><span class="header-section-number">3.3.2</span> Using <code>skim()</code> to diagnose datasets</a></li>
  <li><a href="#checking-individual-values" id="toc-checking-individual-values" class="nav-link" data-scroll-target="#checking-individual-values"><span class="header-section-number">3.3.3</span> Checking individual values</a></li>
  <li><a href="#diagnosing-datasets-a-checklist" id="toc-diagnosing-datasets-a-checklist" class="nav-link" data-scroll-target="#diagnosing-datasets-a-checklist"><span class="header-section-number">3.3.4</span> Diagnosing datasets: a checklist</a></li>
  </ul></li>
  <li><a href="#mending-the-data" id="toc-mending-the-data" class="nav-link" data-scroll-target="#mending-the-data"><span class="header-section-number">3.4</span> Mending the data</a>
  <ul class="collapse">
  <li><a href="#correcting-column-names" id="toc-correcting-column-names" class="nav-link" data-scroll-target="#correcting-column-names"><span class="header-section-number">3.4.1</span> Correcting column names</a></li>
  <li><a href="#correcting-outliers" id="toc-correcting-outliers" class="nav-link" data-scroll-target="#correcting-outliers"><span class="header-section-number">3.4.2</span> Correcting outliers</a></li>
  <li><a href="#correcting-categorical-data" id="toc-correcting-categorical-data" class="nav-link" data-scroll-target="#correcting-categorical-data"><span class="header-section-number">3.4.3</span> Correcting categorical data</a></li>
  <li><a href="#incorrectly-imported-numeric-data" id="toc-incorrectly-imported-numeric-data" class="nav-link" data-scroll-target="#incorrectly-imported-numeric-data"><span class="header-section-number">3.4.4</span> Incorrectly imported numeric data</a></li>
  </ul></li>
  <li><a href="#regular-expressions" id="toc-regular-expressions" class="nav-link" data-scroll-target="#regular-expressions"><span class="header-section-number">3.5</span> Regular expressions</a>
  <ul class="collapse">
  <li><a href="#introduction-to-regular-expressions" id="toc-introduction-to-regular-expressions" class="nav-link" data-scroll-target="#introduction-to-regular-expressions"><span class="header-section-number">3.5.1</span> Introduction to regular expressions</a></li>
  <li><a href="#how-regexp-works" id="toc-how-regexp-works" class="nav-link" data-scroll-target="#how-regexp-works"><span class="header-section-number">3.5.2</span> How regexp works</a></li>
  <li><a href="#usage-of-regular-expressions-in-r" id="toc-usage-of-regular-expressions-in-r" class="nav-link" data-scroll-target="#usage-of-regular-expressions-in-r"><span class="header-section-number">3.5.3</span> Usage of regular expressions in R</a></li>
  <li><a href="#correcting-columns-in-iris_data-with-regular-expressions" id="toc-correcting-columns-in-iris_data-with-regular-expressions" class="nav-link" data-scroll-target="#correcting-columns-in-iris_data-with-regular-expressions"><span class="header-section-number">3.5.4</span> Correcting columns in <code>iris_data</code> with regular expressions</a></li>
  <li><a href="#sec-quick-look" id="toc-sec-quick-look" class="nav-link" data-scroll-target="#sec-quick-look"><span class="header-section-number">3.5.5</span> Just a quick look</a></li>
  </ul></li>
  <li><a href="#writing-data" id="toc-writing-data" class="nav-link" data-scroll-target="#writing-data"><span class="header-section-number">3.6</span> Writing data</a>
  <ul class="collapse">
  <li><a href="#keep-your-data-organized" id="toc-keep-your-data-organized" class="nav-link" data-scroll-target="#keep-your-data-organized"><span class="header-section-number">3.6.1</span> Keep your data organized</a></li>
  <li><a href="#functions-to-use-for-writing-data" id="toc-functions-to-use-for-writing-data" class="nav-link" data-scroll-target="#functions-to-use-for-writing-data"><span class="header-section-number">3.6.2</span> Functions to use for writing data</a></li>
  </ul></li>
  <li><a href="#principles-of-data-management" id="toc-principles-of-data-management" class="nav-link" data-scroll-target="#principles-of-data-management"><span class="header-section-number">3.7</span> Principles of data management</a>
  <ul class="collapse">
  <li><a href="#keeping-originals" id="toc-keeping-originals" class="nav-link" data-scroll-target="#keeping-originals"><span class="header-section-number">3.7.1</span> Keeping originals</a></li>
  <li><a href="#use-r-for-data-editing" id="toc-use-r-for-data-editing" class="nav-link" data-scroll-target="#use-r-for-data-editing"><span class="header-section-number">3.7.2</span> Use R for data editing</a></li>
  <li><a href="#working-with-excel-and-other-spreadsheets" id="toc-working-with-excel-and-other-spreadsheets" class="nav-link" data-scroll-target="#working-with-excel-and-other-spreadsheets"><span class="header-section-number">3.7.3</span> Working with Excel and other spreadsheets</a></li>
  </ul></li>
  <li><a href="#review" id="toc-review" class="nav-link" data-scroll-target="#review"><span class="header-section-number">3.8</span> Review</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Reading and Writing Files</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="aims-for-today" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="aims-for-today"><span class="header-section-number">3.1</span> Aims for today</h2>
<ul>
<li>Reading data</li>
<li>Cleaning data</li>
<li>Regular expressions</li>
<li>Data management</li>
</ul>
<p>Today is a special day. If there is one thing that I would like you to learn from this course, it is how to read and write the data.</p>
</section>
<section id="reading-data" class="level2 page-columns page-full" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="reading-data"><span class="header-section-number">3.2</span> Reading data</h2>
<section id="data-types" class="level3 page-columns page-full" data-number="3.2.1">
<h3 data-number="3.2.1" class="anchored" data-anchor-id="data-types"><span class="header-section-number">3.2.1</span> Data types</h3>
<p>In your work you will encounter many different types of data. Most frequently, you will work with tabular data, either as Excel files or as comma or tab separated values (CSV and TSV files, respectively). I am sure you have worked with such files before.</p>
<div class="page-columns page-full"><p> To read these files, we will use two packages: <code>readr</code> and <code>readxl</code>. The former, <code>readr</code>, is part of the <code>tidyverse</code> package, so when you load the tidyverse using <code>library(tidyverse)</code>, <code>readr</code> is loaded as well. The latter, <code>readxl</code>, is a separate package that you need to install and load separately.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>readr</code> and <code>readxl</code> packages</span></div></div>
<div class="page-columns page-full"><p> There are also “base R” functions <code>read.table</code>, <code>read.csv</code>, <code>read.tsv</code> (there is no function for reading XLS[X] files in base R). These are always available when you start R, but don’t use them. The tidyverse functions are not only faster, but also much better behaving and, which is most important, they are safer – it is less likely to mess up your data with them.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>read.table()</code>, <code>read.csv()</code>, <code>read.tsv()</code></span></div></div>
<div class="page-columns page-full"><p> tidyverse functions return <em>tibbles</em>, which, as you remember from yesterday, are a special flavor of data frames. Just to refresh your memory, here are key facts about tibbles:</p><div class="no-row-height column-margin column-container"><span class="margin-aside">tibbles</span></div></div>
<ul>
<li>in most of the cases, they behave exactly like data frames</li>
<li>when you print them, they are nicer</li>
<li>tibbles have no row names</li>
<li>when you select columns using <code>[ , sel ]</code>, you <em>always</em> get a tibble, even if you selected only one column</li>
</ul>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>read_table()</code>, <code>read_tsv()</code>, <code>read_csv()</code>, <code>read_delim()</code>, <code>read_xls()</code>, <code>read_xlsx()</code>, <code>read_excel()</code></span></div></div>
<table class="caption-top table">
<colgroup>
<col style="width: 24%">
<col style="width: 26%">
<col style="width: 26%">
<col style="width: 22%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Data type</th>
<th style="text-align: left;">Function</th>
<th style="text-align: left;">Package</th>
<th style="text-align: left;">Notes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Columns separated by spaces</td>
<td style="text-align: left;"><code>read_table()</code></td>
<td style="text-align: left;"><code>readr</code>/<code>tidyverse</code></td>
<td style="text-align: left;">one or more spaces separate each column</td>
</tr>
<tr class="even">
<td style="text-align: left;">TSV / TAB separated values</td>
<td style="text-align: left;"><code>read_tsv()</code></td>
<td style="text-align: left;"><code>readr</code>/<code>tidyverse</code></td>
<td style="text-align: left;">Delimiter is tab (<code>\t</code>).</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CSV / comma separated</td>
<td style="text-align: left;"><code>read_csv()</code></td>
<td style="text-align: left;"><code>readr</code>/<code>tidyverse</code></td>
<td style="text-align: left;">Comma separated values</td>
</tr>
<tr class="even">
<td style="text-align: left;">Any delimiter</td>
<td style="text-align: left;"><code>read_delim()</code></td>
<td style="text-align: left;"><code>readr</code>/<code>tidyverse</code></td>
<td style="text-align: left;">Customizable</td>
</tr>
<tr class="odd">
<td style="text-align: left;">XLS (old Excel)</td>
<td style="text-align: left;"><code>read_xls()</code> <code>read_excel()</code></td>
<td style="text-align: left;"><code>readxl</code></td>
<td style="text-align: left;">Avoid using XLS files. From the <code>readxl</code> package.</td>
</tr>
<tr class="even">
<td style="text-align: left;">XLSX (new Excel)</td>
<td style="text-align: left;"><code>read_xlsx()</code> <code>read_excel()</code></td>
<td style="text-align: left;"><code>readxl</code></td>
<td style="text-align: left;">From the <code>readxl</code> package. You need to provide the sheet number you wish to read. Note: returns a tibble, not a data frame!</td>
</tr>
</tbody>
</table>
<p>As you can see, the functions we recommend to use can be used by loading the packages <code>tidyverse</code> and <code>readxl</code>. If you haven’t done that yet, please install these packages now:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"tidyverse"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"readxl"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>However, before we start using these functions, we need to dive into a very important problem: where are your files?</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Remember!
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>For reading text files (csv, tsv etc.), use the <code>readr</code> package. This package is loaded automatically when you load the <code>tidyverse</code> package: <code>library(tidyverse)</code>. Then, use the functions <code>read_csv</code>, <code>read_tsv</code> etc.</li>
<li>For reading Excel files, use the <code>readxl</code> package: <code>library(readxl)</code>. Then, use the function <code>read_excel</code>.</li>
</ul>
</div>
</div>
</section>
<section id="where-is-my-file-relative-and-absolute-paths" class="level3 page-columns page-full" data-number="3.2.2">
<h3 data-number="3.2.2" class="anchored" data-anchor-id="where-is-my-file-relative-and-absolute-paths"><span class="header-section-number">3.2.2</span> Where is my file? Relative and absolute paths</h3>
<p>We are coming now to a crucial problem and a source of endless frustration for beginning R programmers: how to tell R where your file is. Fortunately, there are many ways to deal with that, both with R and RStudio. Still, this is a key problem and we would like you to spend some time on the following chapter.</p>
<div class="page-columns page-full"><p> In order to access a file, a program (any program) needs what is known a <em>path</em>. Path is a character string that tells the program how to get to the file and looks, for example, like this on a Windows computer: <code>C:/Users/johndoe/Documents/RProjects/MyFirstRProject/data.csv</code>, and like this on a Mac: <code>/Users/johndoe/Documents/RProjects/MyFirstRProject/data.csv</code>.</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Path</span></div></div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Path separators on different systems
</div>
</div>
<div class="callout-body-container callout-body">
<p>Most computer systems separate the directories and files in a path using a slash (<code>/</code>). However, Windows uses a backslash (<code>\</code>). This is quite annoying for R users, because in R character vectors, the slash has a special meaning. To use backslashes, you need to “escape” them by putting another backslash in front of each backslash. So instead of <code>C:\Users\johndoe\Documents</code>, you need to write <code>C:\\Users\\johndoe\\Documents</code>. Alternatively, you can use the forward slash even on a Windows system, so type <code>C:/Users/johndoe/Documents</code>. We recommend the latter approach.</p>
<p>This is also why simply copying a path from the Windows Explorer to your R script will not work in R – because the copied text contains single backslashes.</p>
</div>
</div>
<p>If R cannot find your file, it will return an error message. At first, you will be seeing this error <em>a lot</em>:</p>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside">Error: file not found</span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: 'data.csv' not found in the current working directory
('C:/Users/johndoe/Documents/RProjects/MyFirstRProject/').</code></pre>
</div>
</div>
<div class="page-columns page-full"><p> Before we proceed, you need to understand one important thing. When you start your RStudio and R session, the R session runs in a specific directory. This is called the <em>working directory</em>. You can check what is using the <code>getwd()</code> function<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>getwd()</code></span><div id="fn1"><p><sup>1</sup>&nbsp;You can change it using the <code>setwd()</code> function, but avoid doing that. This path leads to madness, trust me on that.</p></div></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "C:/Users/johndoe/Documents/RProjects/MyFirstRProject"</code></pre>
</div>
</div>
<p>Of course, the result will be different on your computer. By the way, the above is called an <em>absolute</em> path. That means that it works no matter where you are in your system, because a program can always find the file or directory using this path.</p>
<p>The easiest way to read the data is this: copy your data files to the directory returned by <code>getwd()</code>.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-simple-reading" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.1 (Reading your first file)</strong></span> &nbsp;</p>
<ul>
<li>Check your working directory using <code>getwd()</code></li>
<li>Load the tidyverse package using <code>library(tidyverse)</code></li>
<li>Go to the URL <a href="https://github.com/bihealth/RCrashcourse-book/Datasets">https://github.com/bihealth/RCrashcourse-book/Datasets</a></li>
<li>Click on “iris.csv”</li>
<li>Click on the “Download raw file” button on the right side of the screen</li>
<li>Save the file in the directory returned by <code>getwd()</code></li>
<li>Read the file using <code>read_csv("iris.csv")</code></li>
</ul>
</div>
</div>
</div>
</div>
<p>The following code should now work without an error:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>iris_data <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"iris.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now the contents of the file are stored in the <code>iris_data</code> object. There are many ways to have a look at it:</p>
<ul>
<li>type <code>iris_data</code> or <code>print(iris_data)</code> in the console</li>
<li>type <code>View(iris_data)</code> in the console</li>
<li>click on the little spreadsheet icon next to the <code>iris_data</code> object in the Environment tab in RStudio (upper right panel)</li>
</ul>
<p>Please make sure that the above works for you. If it does not, read the instructions again. In RStudio, there is a “Files” tab in the lower right panel. You should see your working directory as well as the “iris.csv” file there.</p>
</section>
<section id="reading-with-relative-paths" class="level3 page-columns page-full" data-number="3.2.3">
<h3 data-number="3.2.3" class="anchored" data-anchor-id="reading-with-relative-paths"><span class="header-section-number">3.2.3</span> Reading with relative paths</h3>
<p>OK, so far, so good. That was easy. Now comes a slightly harder part.</p>
<p>Saving your data files in the working directory works well if you have one or two. However, the more files you read and write, the more cluttered your project directory becomes. You will soon find yourself in a situation where you have no idea which file is which.</p>
<p>It is generally a good idea to keep your data files in a separate directory, or even multiple directories.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-simple-reading-directory" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.2 (Reading your first file from a data directory)</strong></span> &nbsp;</p>
<ul>
<li>In the working directory, create a new directory called “Datasets”</li>
<li>Move the “iris.csv” file to the “Datasets” directory</li>
<li>Read the file using <code>read_csv("Datasets/iris.csv")</code></li>
</ul>
</div>
</div>
</div>
</div>
<div class="page-columns page-full"><p> The path “Datasets/iris.csv” is called a <em>relative</em> path, because it is relative to the working directory and will not work from another location. So why should we use relative paths? Wouldn’t it be easier to use absolute paths all the time, for example <code>read_csv("C:/Users/johndoe/Documents/RProjects/MyFirstRProject/Datasets/iris.csv")</code>?</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Relative path</span></div></div>
<p>Actually, no. The problem is that if you move your R project to another location, of if you share it with someone else, the absolute path will no longer work. In other words, the absolute path is not portable.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Remember!
</div>
</div>
<div class="callout-body-container callout-body">
<p>Do not use absolute paths in your code. Always use relative paths.</p>
</div>
</div>
</section>
<section id="more-on-relative-paths" class="level3 page-columns page-full" data-number="3.2.4">
<h3 data-number="3.2.4" class="anchored" data-anchor-id="more-on-relative-paths"><span class="header-section-number">3.2.4</span> More on relative paths</h3>
<p>Some times the data files are not in your R project directory. For example, you are writing your PhD thesis. You have created a directory called “PhD”, which contains directories “manuscript”, “data”, “images”, “R_project” and so on. You use R to do a part of the calculations, but you want to keep the data files in the “data” directory. How to read them?</p>
<p>When you type <code>getwd()</code>, you will get the path to the “R_project” directory, something like <code>C:/Users/johndoe/Documents/PhD/R_project</code>. The date files are in the directory <code>C:/Users/johndoe/Documents/PhD/data</code>. To get the relative path from the R project directory to the data directory, think about how you would navigate from one directory to another in the Windows Explorer or Finder. You need to go up one level, to get into “PhD”, and then down again to “data”.</p>
<div class="page-columns page-full"><p> Getting “up one level” in a path is done by using <code>..</code>. So the relative path to the file “iris.csv” in your “data” directory is <code>../data/iris.csv</code>.</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Up one level with <code>..</code></span></div></div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-reading-directory-relative" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.3 (Reading a file from a data directory using relative paths)</strong></span> &nbsp;</p>
<ul>
<li>In the directory that contains the working directory create a directory called “Data”. That is, if your working directory is <code>C:/Users/johndoe/Documents/PhD/R_project</code>, create the directory <code>C:/Users/johndoe/Documents/PhD/Data</code></li>
<li>Move the “iris.csv” file to the new “Data” directory</li>
<li>Read the file using <code>read_csv("../Data/iris.csv")</code></li>
</ul>
</div>
</div>
</div>
</div>
<p>But what about the case when your data directory is at a completely different location? For example, on a different drive, or maybe on your desktop?</p>
<p>First, I don’t recommend keeping your data files separately from the R project directory. In general, try to put everything in one place, as part of one structure. This structure can be complex, but it should be coherent. If necessary, copy the data files into your R project directory.</p>
<p>However, sometimes it simply isn’t possible. Maybe the files are huge and you need to read them from a special drive. In this case, there are three options.</p>
<p><strong>Using absolute paths.</strong> Yes, I told you not to use absolute paths, but sometimes you have no choice.</p>
<p><strong>Create shortcuts.</strong> In all systems it is possible to create shortcuts to your data directories (on Unix-like systems like MacOS they are called “symbolic links”). You can put these shortcuts in your R project directory – R will treat them as normal directories.</p>
<p><strong>Create a complex relative path.</strong> Depending on how far “away” your data directory is, you can create a complex relative path. For example, if your R project directory is <code>C:/Users/johndoe/Documents/PhD/R_project</code> and your data directory is <code>D:/Data</code>, you can use the path <code>../../../../Data/iris.csv</code>. Unfortunately, despite being a relative path, this is not very portable (and it is easy to lose count on the <code>..</code>’s).</p>
</section>
<section id="using-rstudio-to-import-files" class="level3" data-number="3.2.5">
<h3 data-number="3.2.5" class="anchored" data-anchor-id="using-rstudio-to-import-files"><span class="header-section-number">3.2.5</span> Using RStudio to import files</h3>
<p>RStudio has a very nice feature that allows you to import files using a point-and-click interface. When you click on a data file in the “Files” tab (lower right panel), you will see two options: “View File” and “Import Dataset”. Choosing the latter opens a dialog window which shows a preview of the file and allows you to construct your <code>read_csv</code>, <code>read_tsv</code> or another command using a point-and-click interface. You also see a preview of the expression that will be executed.</p>
<p>This feature is particularly useful when you are not sure about the format of the data to import, e.g.&nbsp;what type of delimiter is used, how many lines should be skipped etc.</p>
<p>Then you can click on “Import” to actually run the command, however I would recommend another approach. Clicking on “Import” runs the command directly in the console, bypassing your script – and you should always enter the code in your script before executing it.</p>
<p>Rather than clicking on “Import”, click on the little Copy icon next to the “Code preview” field, and then cancel the dialog. Paste the copied code into your script, and then run it.</p>
<p>There is one more thing to modify. The code generated by RStudio often uses absolute paths. Try to modify it to use relative paths to ensure portability of your script.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-opening-data" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.4 (Reading data)</strong></span> &nbsp;</p>
<ul>
<li>Go to the URL <a href="https://github.com/bihealth/RCrashcourse-book/Datasets">https://github.com/bihealth/RCrashcourse-book/Datasets</a></li>
<li>Download the following files:
<ul>
<li><code>TB_ORD_Gambia_Sutherland_biochemicals.csv</code></li>
<li><code>iris.tsv</code></li>
<li><code>meta_data_botched.xlsx</code></li>
</ul></li>
<li>Alternatively, you can download the whole repository as a ZIP file and unpack it.</li>
<li>Save the files in the “Data” directory in your working directory – or another location of your choice.</li>
<li>Read the files using the appriopriate functions. Consult the table above for the correct function names, or use the RStudio data import feature. Make sure that you are using relative paths.</li>
</ul>
</div>
</div>
</div>
</div>
</section>
<section id="reading-excel-files" class="level3 page-columns page-full" data-number="3.2.6">
<h3 data-number="3.2.6" class="anchored" data-anchor-id="reading-excel-files"><span class="header-section-number">3.2.6</span> Reading Excel files</h3>
<p>Reading files sometimes requires diligence. This is especially true for Excel files – they can contain multiple sheets, tables often do not start on the first row etc.</p>
<div class="page-columns page-full"><p> The package <code>readxl</code> (which you hopefully successfully used to read the XLSX file in the previous exercise) contains several example files. They have been installed on your system when you installed the package. Manually finding these example files is annoying, and that is why the <code>readxl</code> package provides a convenience function, <code>readxl_example()</code>, that returns the absolute path to the file (yes, I know what I said about absolute paths; this is an exception).</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Reading excel files with <code>readxl</code> package</span></div></div>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>readxl_example()</code>,<code>read_excel()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>fn <span class="ot">&lt;-</span> <span class="fu">readxl_example</span>(<span class="st">"deaths.xls"</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(fn)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/home/january/R/x86_64-pc-linux-gnu-library/4.4/readxl/extdata/deaths.xls"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>deaths <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(fn)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>New names:
• `` -&gt; `...2`
• `` -&gt; `...3`
• `` -&gt; `...4`
• `` -&gt; `...5`
• `` -&gt; `...6`</code></pre>
</div>
</div>
<p>If you view this file (or if you use the RStudio data import feature), you will notice that the actual data in this file starts on line 5; the first 4 lines contain the text “Lots of people simply cannot resist writing some notes at the top of their spreadsheets or merging cells”. To skip these lines, we need to use an argument to the <code>read_excel</code> function. If you look up the help file for the function (e.g.&nbsp;using <code>?read_excel</code>), you will find the following line:</p>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside">skipping lines with <code>read_excel(skip=...)</code></span></div></div>
<pre><code>skip    Minimum number of rows to skip before reading anything,
        be it column names or data. Leading empty rows are 
        automatically skipped, so this is a lower bound. Ignored
        if range is given.</code></pre>
<p>Therefore, we can modify the code above to skip the first four lines:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>deaths <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(fn, <span class="at">skip=</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-day3-opening-data-with-options" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.5 (Reading data with options)</strong></span> If you take a closer look at the file <code>deaths.xls</code>, you will notice that there is some extra text at the bottom of the file as well. How can you omit that part when reading? <em>Hint: there are two ways to do that. If in doubt, look up the “examples” section of the <code>readxl</code> helpfile.</em></p>
<p>(<a href="solutions.html#exr-day3-opening-data-with-options">Solution</a>)</p>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="diagnosing-and-cleaning-data" class="level2 page-columns page-full" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="diagnosing-and-cleaning-data"><span class="header-section-number">3.3</span> Diagnosing and cleaning data</h2>
<section id="sec-diagnosing" class="level3 page-columns page-full" data-number="3.3.1">
<h3 data-number="3.3.1" class="anchored" data-anchor-id="sec-diagnosing"><span class="header-section-number">3.3.1</span> Diagnosing datasets</h3>
<p>Congratulations! You are now proficient in reading data into R. Believe me or not, this alone is an important step forward – many who try their hand at R get stuck at this point.</p>
<p>However, reading data is only the first step. In most of the cases, the data requires some treatment: cleaning, transformation, filtering etc. In many projects that I have worked on as a bioinformatician, data import, diagnosis and cleaning took up the majority of time spent on the project. Unfortunately, this is necessary before any real fun with the data can start.</p>
<div class="page-columns page-full"><p> Let us examine the file <code>iris.csv</code> that you have just read. The dataset comes from a famous paper by Ronald Fisher, who used it to demonstrate his newly developed method called linear discriminant analysis – an early machine learning algorithm, if you will. The dataset contains measurements (in cm) of 150 flowers of three species of irises: <em>Iris setosa</em>, <em>Iris versicolor</em> and <em>Iris virginica</em>. The measurements are the sepal and petal length and width, four measurements in total. Each row consists of these four measurements, plus a column which contains the species name.</p><div class="no-row-height column-margin column-container"><span class="margin-aside">The <code>iris</code> dataset</span></div></div>
<p>I have doctored the file to mimick typical problems with imported data, especially in clinical trial settings. Humans who enter data make errors, this is normal and expected, all of us do. Before we analyse them, we need to correct them. Before we correct them, we need to <em>find</em> them.</p>
<p>First, note that when you read the data using <code>read_csv</code>, the function conveniently shows what types of data were assigned to each column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>iris_data <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"Datasets/iris.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 150 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (3): Sepal Length, Petal?Length, Species
dbl (2): Sepal Width, Petal.Width

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p>This should already ring some alarm bells. First, it looks like the columns are not consistently named: we have <code>Sepal Length</code> and <code>Petal?Length</code>, and also <code>Petal.Width</code>.</p>
<div class="page-columns page-full"><p> Secondly, we would expect that the measurements are imported as numbers. However, both sepal and petal lengths are imported as characters. We can confirm this by using the <code>class</code> function:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>class()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(iris_data[[<span class="st">"Sepal Length"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "character"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(iris_data[[<span class="st">"Sepal Width"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numeric"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(iris_data[[<span class="st">"Petal?Length"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "character"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(iris_data[[<span class="st">"Petal.Width"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numeric"</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<code>class</code> and <code>typeof</code>
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <code>class</code> function returns the class of an object, which is a higher-level classification of the object. An object can have multiple classes. The <code>typeof</code> function returns the internal storage type of the object, which is a lower level classification. For example, both tibbles and data frames have the type <code>list</code>, but their classes are different. Another example:i if <code>mtx</code> is a matrix of numbers, <code>typeof(mtx)</code> is <code>double</code>, and <code>class(mtx)</code> is <code>matrix</code>.</p>
</div>
</div>
<p>Note that instead of using <code>iris_data$Sepal Length</code> (which will not work, because there is a space in the column name), we use the double bracket notation. An alternative would be to use quotes: <code>iris_data$'Sepal Length'</code>. This a reason why want to avoid spaces and special characters in column names (in a moment we will show you how to standardize column names). If you use tab-completion with RStudio, the quotes will be inserted automatically.</p>
<div class="page-columns page-full"><p> We can also use the <code>summary</code> function on the whole dataset:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>summary()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(iris_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Sepal Length        Sepal Width     Petal?Length        Petal.Width   
 Length:150         Min.   : 2.000   Length:150         Min.   :0.100  
 Class :character   1st Qu.: 2.800   Class :character   1st Qu.:0.300  
 Mode  :character   Median : 3.000   Mode  :character   Median :1.300  
                    Mean   : 3.411                      Mean   :1.199  
                    3rd Qu.: 3.375                      3rd Qu.:1.800  
                    Max.   :36.000                      Max.   :2.500  
   Species         
 Length:150        
 Class :character  
 Mode  :character  
                   
                   
                   </code></pre>
</div>
</div>
<p>Ops, another problem. Did you notice the maximum value for sepal widths? It is 36 cm. This cannot be, way too large for the flowers. Also, both the mean and median are around 3 cm, so this must be either an outlier or a clerical error.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
The <code>summary()</code> functions
</div>
</div>
<div class="callout-body-container callout-body">
<p>Under the hood, there is no single <code>summary()</code> function. Instead, different classes can have different types of summaries. Whenever you produce a result, always try the summary function with it.</p>
</div>
</div>
<div class="page-columns page-full"><p> Alternatively, we can use the tidyverse <code>glimpse</code> function. Rather then providing a summary, this function uses a terse format to show the data type for each column and first few values:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>glimpse()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(iris_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 150
Columns: 5
$ `Sepal Length` &lt;chr&gt; "5.1", "4.9", "4.7", "4.6", "5", "5.4", "4.6", "5", "4.…
$ `Sepal Width`  &lt;dbl&gt; 3.5, 3.0, 3.2, 3.1, 3.6, 3.9, 3.4, 3.4, 2.9, 3.1, 3.7, …
$ `Petal?Length` &lt;chr&gt; "1.4", "1.4", "1.3", "1.5", "1.4", "1.7", "1.4", "1.5",…
$ Petal.Width    &lt;dbl&gt; 0.2, 0.2, 0.2, 0.2, 0.2, 0.4, 0.3, 0.2, 0.2, 0.1, 0.2, …
$ Species        &lt;chr&gt; "setosa", "setosa", "setosa", "setosa", "setosa", "seto…</code></pre>
</div>
</div>
<div class="page-columns page-full"><p> Another way how we can diagnose the dataset is to use the <code>str</code> function. This provides a more detailed summary for each column:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>str()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(iris_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>spc_tbl_ [150 × 5] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ Sepal Length: chr [1:150] "5.1" "4.9" "4.7" "4.6" ...
 $ Sepal Width : num [1:150] 3.5 3 3.2 3.1 3.6 3.9 3.4 3.4 2.9 3.1 ...
 $ Petal?Length: chr [1:150] "1.4" "1.4" "1.3" "1.5" ...
 $ Petal.Width : num [1:150] 0.2 0.2 0.2 0.2 0.2 0.4 0.3 0.2 0.2 0.1 ...
 $ Species     : chr [1:150] "setosa" "setosa" "setosa" "setosa" ...
 - attr(*, "spec")=
  .. cols(
  ..   `Sepal Length` = col_character(),
  ..   `Sepal Width` = col_double(),
  ..   `Petal?Length` = col_character(),
  ..   Petal.Width = col_double(),
  ..   Species = col_character()
  .. )
 - attr(*, "problems")=&lt;externalptr&gt; </code></pre>
</div>
</div>
</section>
<section id="sec-skim" class="level3 page-columns page-full" data-number="3.3.2">
<h3 data-number="3.3.2" class="anchored" data-anchor-id="sec-skim"><span class="header-section-number">3.3.2</span> Using <code>skim()</code> to diagnose datasets</h3>
<div class="page-columns page-full"><p> Another option to diagnose the dataset is to use the <code>skim</code> function from the <code>skimr</code> package, which you installed (hopefully) yesterday. The <code>skim</code> package provides a more detailed summary of the dataset, including the number of missing values, the number of unique values and summary statistics. To use it, you need first to load the package:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>skim()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(skimr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This <code>skim()</code> function shows more than just the <code>str</code> function. In addition to column types, for character and factor columns it shows the unique values, ordered by their frequency. For numerical values, it shows their non-parametric summary statistics (range, median, quartiles):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">skim</span>(iris_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td style="text-align: left;">Name</td>
<td style="text-align: left;">iris_data</td>
</tr>
<tr class="even">
<td style="text-align: left;">Number of rows</td>
<td style="text-align: left;">150</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Number of columns</td>
<td style="text-align: left;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">_______________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Column type frequency:</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">character</td>
<td style="text-align: left;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">________________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Group variables</td>
<td style="text-align: left;">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: character</strong></p>
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 19%">
<col style="width: 13%">
<col style="width: 19%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 8%">
<col style="width: 12%">
<col style="width: 15%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">min</th>
<th style="text-align: right;">max</th>
<th style="text-align: right;">empty</th>
<th style="text-align: right;">n_unique</th>
<th style="text-align: right;">whitespace</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Sepal Length</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">37</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">Petal?Length</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">47</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Species</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: numeric</strong></p>
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 18%">
<col style="width: 13%">
<col style="width: 18%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 7%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">mean</th>
<th style="text-align: right;">sd</th>
<th style="text-align: right;">p0</th>
<th style="text-align: right;">p25</th>
<th style="text-align: right;">p50</th>
<th style="text-align: right;">p75</th>
<th style="text-align: right;">p100</th>
<th style="text-align: left;">hist</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Sepal Width</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3.41</td>
<td style="text-align: right;">3.16</td>
<td style="text-align: right;">2.0</td>
<td style="text-align: right;">2.8</td>
<td style="text-align: right;">3.0</td>
<td style="text-align: right;">3.38</td>
<td style="text-align: right;">36.0</td>
<td style="text-align: left;">▇▁▁▁▁</td>
</tr>
<tr class="even">
<td style="text-align: left;">Petal.Width</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1.20</td>
<td style="text-align: right;">0.76</td>
<td style="text-align: right;">0.1</td>
<td style="text-align: right;">0.3</td>
<td style="text-align: right;">1.3</td>
<td style="text-align: right;">1.80</td>
<td style="text-align: right;">2.5</td>
<td style="text-align: left;">▇▁▇▅▃</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="page-columns page-full"><p> Hm, when you look at the output you might notice one more thing: for the last column, “Species”, <code>skim()</code> shows that there are six unique values (take a look at the <code>n_unique</code> column). However, how can that be? We know that there are only three species in this dataset. We can check this using the <code>unique</code> function:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>unique()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(iris_data[[<span class="st">"Species"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "setosa"     "Setosa"     "versicolor" "Versicolor" "virginica" 
[6] "Virginica" </code></pre>
</div>
</div>
<p>There it is. Looks like whoever typed the data<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>, sometimes used a lower-case species designation, and sometimes upper-case. However, for R (and for computers in general), “versicolor” and “Versicolor” are two different things.</p>
<div class="no-row-height column-margin column-container"><div id="fn2"><p><sup>2</sup>&nbsp;In this case, it was probably me.</p></div></div><div class="page-columns page-full"><p> More information can be gained using the <code>table()</code> function:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>table()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(iris_data[[<span class="st">"Species"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    setosa     Setosa versicolor Versicolor  virginica  Virginica 
        45          5         42          8         46          4 </code></pre>
</div>
</div>
<p>Here we not only see the unique values, but also their frequency. The <code>table()</code> function is very useful for any categorical data, as you will see later.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Lower and upper case
</div>
</div>
<div class="callout-body-container callout-body">
<p>For computers in general and R in particular, “lowercase” and “Uppercase” are two different things. Variables <code>a</code> and <code>A</code> are different, as are <code>versicolor</code> and <code>Versicolor</code>. This is called <em>case sensitivity</em>.</p>
</div>
</div>
</section>
<section id="checking-individual-values" class="level3 page-columns page-full" data-number="3.3.3">
<h3 data-number="3.3.3" class="anchored" data-anchor-id="checking-individual-values"><span class="header-section-number">3.3.3</span> Checking individual values</h3>
<p>We have seen before that some measurement columns were imported as character vectors, while others were correctly imported as numbers. To understand why, we need to have a closer look at the individual columns.</p>
<p>One of the columns that has been imported as a character vector is the “Sepal Length”. We will extract the column as a vector and then try to manually convert it to a number:</p>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>as.numeric()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>sepal_length <span class="ot">&lt;-</span> iris_data[[<span class="st">"Sepal Length"</span>]]</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as.numeric</span>(sepal_length)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: NAs introduced by coercion</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] 5.1 4.9 4.7 4.6 5.0 5.4 4.6 5.0 4.4 4.9 5.4 4.8 4.8 4.3 5.8 5.7 5.4 5.1
 [19] 5.7 5.1 5.4 5.1 4.6 5.1 4.8 5.0 5.0 5.2 5.2 4.7 4.8 5.4 5.2 5.5 4.9 5.0
 [37] 5.5 4.9 4.4 5.1 5.0 4.5 4.4 5.0 5.1 4.8 5.1 4.6 5.3 5.0 7.0 6.4 6.9 5.5
 [55] 6.5 5.7 6.3 4.9 6.6 5.2 5.0 5.9 6.0 6.1 5.6 6.7 5.6 5.8 6.2  NA 5.9 6.1
 [73] 6.3 6.1 6.4 6.6 6.8 6.7 6.0 5.7 5.5 5.5 5.8 6.0 5.4 6.0 6.7 6.3 5.6 5.5
 [91] 5.5 6.1 5.8 5.0 5.6 5.7 5.7 6.2 5.1 5.7 6.3 5.8 7.1 6.3 6.5 7.6 4.9 7.3
[109] 6.7 7.2 6.5 6.4 6.8 5.7 5.8 6.4 6.5 7.7  NA 6.0 6.9 5.6 7.7 6.3 6.7 7.2
[127] 6.2 6.1 6.4 7.2 7.4 7.9 6.4 6.3 6.1 7.7 6.3 6.4 6.0 6.9 6.7 6.9 5.8 6.8
[145] 6.7 6.7 6.3 6.5 6.2 5.9</code></pre>
</div>
</div>
<p>Two things to note: first, R issued a warning, “NAs introduced by coercion”. That means that some values could not be converted to numbers, and were replaced by the special value <code>NA</code><a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>. Second, <em>there was no error</em>. R happily allowed you to create a vector of numbers from a vector even though there were some problems.</p>
<div class="no-row-height column-margin column-container"><div id="fn3"><p><sup>3</sup>&nbsp;See Day 1 for more information on <code>NA</code> values.</p></div></div><div class="page-columns page-full"><p>When you look at the output, you will easily spot the <code>NA</code> values. You could even figure out at which positions in the vector they occur. However, this is not the way to go – what if you have not 150 values, but 150,000?  Instead, use the <code>is.na</code> function:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>is.na()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to numbers</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>sepal_length_n <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(sepal_length)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: NAs introduced by coercion</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># how many are NA?</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(sepal_length_n))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># which are NA?</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(<span class="fu">is.na</span>(sepal_length_n))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  70 119</code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show the values</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>sepal_length[<span class="fu">is.na</span>(sepal_length_n)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "5,6"   "&gt; 7.7"</code></pre>
</div>
</div>
<p>One more problem remains: the value “36” in the sepal width column. We can check which values are greater than 10 with a similar approach:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>sepal_width <span class="ot">&lt;-</span> iris_data[[<span class="st">"Sepal Width"</span>]]</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="co"># how many are greater than 10?</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(sepal_width <span class="sc">&gt;</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># which are greater than 10?</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(sepal_width <span class="sc">&gt;</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  42 110</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show the values</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>sepal_width[sepal_width <span class="sc">&gt;</span> <span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 23 36</code></pre>
</div>
</div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-petal_length" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.6 (Petal lengths)</strong></span> Repeat the steps above for the <code>Petal length</code> column.</p>
</div>
</div>
</div>
</div>
</section>
<section id="diagnosing-datasets-a-checklist" class="level3" data-number="3.3.4">
<h3 data-number="3.3.4" class="anchored" data-anchor-id="diagnosing-datasets-a-checklist"><span class="header-section-number">3.3.4</span> Diagnosing datasets: a checklist</h3>
<p>Whenever you read a data set, you should check the following things:</p>
<p><strong>Column names.</strong> Are they consistent? Are they easy to type (no spaces, no special characters)? Are they in the correct order? Are they what you expect them to be? Do you understand what each column <em>is</em>?</p>
<p><strong>Data types.</strong> Are the columns of the correct type (e.g. numerical columns imported as numbers)?</p>
<p><strong>Categorical variables.</strong> Do you see what you expect? Is the number of categories correct? Do you understand what each category <em>is</em>?</p>
<p><strong>Numerical variables.</strong> Do the summary statistics make sense? Are there any outliers? Missing values? Were the variables imported incorrectly as characters? Why?</p>
<p><strong>Missing values.</strong> Are there any missing values? Why? Are they important? How to deal with them?</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Checklist for importing data
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Column names</li>
<li>Data types</li>
<li>Categorical variables</li>
<li>Numerical variables</li>
<li>Missing values</li>
</ul>
</div>
</div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-day3-diagnosing" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.7 (Botched metadata)</strong></span> Load the file <code>meta_data_botched.xlsx</code> using the <code>readxl</code> package. Diagnose the problems.</p>
<p>(<a href="solutions.html#sec-day3-diagnosing">Solution</a>)</p>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="mending-the-data" class="level2 page-columns page-full" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="mending-the-data"><span class="header-section-number">3.4</span> Mending the data</h2>
<section id="correcting-column-names" class="level3 page-columns page-full" data-number="3.4.1">
<h3 data-number="3.4.1" class="anchored" data-anchor-id="correcting-column-names"><span class="header-section-number">3.4.1</span> Correcting column names</h3>
<p>Column names should contain only letters, numbers and underscores. While all other characters are also allowed (so a column name can be <code>豚肉 "temp."? (°C) 😀</code>), it is <em>way</em> easier to work with columns with simple, standardized names.</p>
<p>You already know one way of dealing with this – simply assign new names to the columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(iris_data) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"sepal_length"</span>, <span class="st">"sepal_width"</span>, <span class="st">"petal_length"</span>,</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"petal_width"</span>, <span class="st">"species"</span>)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(iris_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "sepal_length" "sepal_width"  "petal_length" "petal_width"  "species"     </code></pre>
</div>
</div>
<div class="page-columns page-full"><p> However, there is a quicker way. The <code>janitor</code> package contains a function called <code>clean_names</code> that does exactly that, automatically and for all  columns. Install it with <code>install.packages("janitor")</code> if you haven’t installed it yet.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>janitor</code> package</span><span class="margin-aside"><code>clean_names()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(janitor)</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>iris_data <span class="ot">&lt;-</span> <span class="fu">clean_names</span>(iris_data)</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(iris_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "sepal_length" "sepal_width"  "petal_length" "petal_width"  "species"     </code></pre>
</div>
</div>
<p>My advice is to use <code>clean_names</code> on every dataset you import.</p>
</section>
<section id="correcting-outliers" class="level3 page-columns page-full" data-number="3.4.2">
<h3 data-number="3.4.2" class="anchored" data-anchor-id="correcting-outliers"><span class="header-section-number">3.4.2</span> Correcting outliers</h3>
<p>In the column with sepal widths, which now has the name <code>sepal_width</code>, we found two values that probably lack the decimal point: 23 and 36. They are at positions 42 and 110, respectively.</p>
<p>It is tempting to simply substitute the data in the original data frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># works, but don't do it</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>sepal_width[<span class="dv">42</span>] <span class="ot">&lt;-</span> <span class="fl">2.3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Unfortunately, this is not a good idea, because this solution is a bit like manual editing in a spreadsheet. There is a better way. First, <strong>generalize the issue</strong>. What seems to be the problem? The values that are larger than 10 are probably missing a decimal point. The simplest solution is to divide them by 10. Using logical vectors, we can do that for <em>all</em> the values that are larger than 10.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># numbers that are missing a decimal point</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>too_large <span class="ot">&lt;-</span> iris_data<span class="sc">$</span>sepal_width <span class="sc">&gt;</span> <span class="dv">10</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>sepal_width[too_large] <span class="ot">&lt;-</span> iris_data<span class="sc">$</span>sepal_width[too_large] <span class="sc">/</span> <span class="dv">10</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This a bit hard to read, so we make it more explicit:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># numbers that are missing a decimal point</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>sepal_width <span class="ot">&lt;-</span> iris_data<span class="sc">$</span>sepal_width</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>too_large <span class="ot">&lt;-</span> sepal_width <span class="sc">&gt;</span> <span class="dv">10</span></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>sepal_width[too_large] <span class="ot">&lt;-</span> sepal_width[too_large] <span class="sc">/</span> <span class="dv">10</span></span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>sepal_width[too_large]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>numeric(0)</code></pre>
</div>
</div>
<p>Looks good. Finally, we assign the corrected values back to the data frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>sepal_width <span class="ot">&lt;-</span> sepal_width</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="page-columns page-full"><p>As the last step, we need to check whether our approach worked for all  values in the column. For this, we will use the <code>any()</code> function, which returns <code>TRUE</code> if any of the values in a logical vector is <code>TRUE</code>:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>any()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(iris_data<span class="sc">$</span>sepal_width <span class="sc">&gt;</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
</div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-potential-problem" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.8</strong></span> Can you spot what the potential problem with this approach is? Hint: what would happen if the original value was 1.10? Can you think of a better approach? Is it actually possible to make sure that our guess is correct?</p>
</div>
</div>
</div>
</div>
</section>
<section id="correcting-categorical-data" class="level3 page-columns page-full" data-number="3.4.3">
<h3 data-number="3.4.3" class="anchored" data-anchor-id="correcting-categorical-data"><span class="header-section-number">3.4.3</span> Correcting categorical data</h3>
<p>In the column “species”, we found that there are two different values for the same species. There are many ways we can handle it. Obviously, we could manually assign the correct labels, but that would be a lot of work.</p>
<p>A better way would be to somehow automatically convert all the values. In our case, the problem is quite simple: it is sufficient to put all the values in uniform lower case, so that <code>Versicolor</code> becomes <code>versicolor</code>.</p>
<div class="page-columns page-full"><p> We can do that in one line of code thanks to the <code>tolower</code> function:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>tolower()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>species <span class="ot">&lt;-</span> <span class="fu">tolower</span>(iris_data<span class="sc">$</span>species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="page-columns page-full"><p> There is also a corresponding function, <code>toupper</code>, that converts all the letters to upper case.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>toupper()</code></span></div></div>
<p>Remember to check that your changes were successful:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(iris_data<span class="sc">$</span>species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    setosa versicolor  virginica 
        50         50         50 </code></pre>
</div>
</div>
<p>This was a simple case, but what if we had more complex typos? For example, “i.versicolor”, “vers.”, “V” etc. In such cases, you would need to use a more sophisticated approach. This approach involves <em>regular expressions</em>, and will also help us to correct the sepal and petal lengths.</p>
</section>
<section id="incorrectly-imported-numeric-data" class="level3 page-columns page-full" data-number="3.4.4">
<h3 data-number="3.4.4" class="anchored" data-anchor-id="incorrectly-imported-numeric-data"><span class="header-section-number">3.4.4</span> Incorrectly imported numeric data</h3>
<p>We have two columns which were imported as characters, but should be numbers. Their brand-new names are <code>sepal_length</code> and <code>petal_length</code>. All in all, there are 2 problems in the sepal length column and 5 in the petal length column, as shown in the table below:</p>
<div class="cell">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 20%">
<col style="width: 15%">
<col style="width: 15%">
<col style="width: 15%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Column</th>
<th style="text-align: center;">Position</th>
<th style="text-align: center;">Original</th>
<th style="text-align: center;">Correct</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">sepal_length</td>
<td style="text-align: center;">70</td>
<td style="text-align: center;"><code>5,6</code></td>
<td style="text-align: center;">5.6</td>
</tr>
<tr class="even">
<td style="text-align: center;">sepal_length</td>
<td style="text-align: center;">119</td>
<td style="text-align: center;"><code>&gt; 7.7</code></td>
<td style="text-align: center;">7.7</td>
</tr>
<tr class="odd">
<td style="text-align: center;">petal_length</td>
<td style="text-align: center;">12</td>
<td style="text-align: center;"><code>1.6!</code></td>
<td style="text-align: center;">1.6</td>
</tr>
<tr class="even">
<td style="text-align: center;">petal_length</td>
<td style="text-align: center;">13</td>
<td style="text-align: center;"><code>1 .4</code></td>
<td style="text-align: center;">1.4</td>
</tr>
<tr class="odd">
<td style="text-align: center;">petal_length</td>
<td style="text-align: center;">14</td>
<td style="text-align: center;"><code>1.1?</code></td>
<td style="text-align: center;">1.1</td>
</tr>
<tr class="even">
<td style="text-align: center;">petal_length</td>
<td style="text-align: center;">115</td>
<td style="text-align: center;"><code>5k.1</code></td>
<td style="text-align: center;">5.1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">petal_length</td>
<td style="text-align: center;">149</td>
<td style="text-align: center;"><code>5. 4</code></td>
<td style="text-align: center;">5.4</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The last column in the table above is our target – this is how we would like to correct the data. How can we do that?</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Correcting values
</div>
</div>
<div class="callout-body-container callout-body">
<p>Of course, without additional information we can only <em>guess</em> that <code>5,6</code> should be <code>5.6</code>. Maybe there were two measurements, 5 and 6? Maybe <code>5. 4</code> should be <code>5.04</code>, and not <code>5.4</code>? In a real-world scenario, you would need to consult the author(s) of the data, or check your lab book.</p>
</div>
</div>
<p>It is tempting to simply substitute the data in the original data frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># works, but don't do it</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>sepal_length[<span class="dv">70</span>] <span class="ot">&lt;-</span> <span class="fl">5.6</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As before, that is not a good idea. If you receive another version of the file, with added lines before the 70th, the position will change and you will have to correct the version manually again. Also, if there are new problems, like yet another typo in the data, you will have to spot it and manually add a line to correct it. Again, we need to generalize the issue.</p>
<p>What is the problem on line 70? A comma instead of a decimal dot – maybe someone who typed it was used to entering numbers in a German version of Excel, where a comma is used instead of a dot. So why not replace <em>all</em> the commas by a dot? This should do the trick not only for line 70, but also for possible future problems.</p>
<div class="page-columns page-full"><p> We will use for that the function <code>str_replace_all</code> from tidyverse<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>. Before we do that, however, we will make a copy of the column. That will allow us to easily see the changes will make, check that everything is OK and only then replace the column in the data frame. Later on, you will learn more efficient ways of doing this.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>str_replace_all()</code></span><div id="fn4"><p><sup>4</sup>&nbsp;In base R, you can use the <code>gsub</code> function. However, it has a slightly different syntax, which makes it less convenient when you start using pipes tomorrow.</p></div></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make a copy</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>sepal_length <span class="ot">&lt;-</span> iris_data<span class="sc">$</span>sepal_length</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a><span class="co"># record the problematic places</span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>problems <span class="ot">&lt;-</span> <span class="fu">is.na</span>(<span class="fu">as.numeric</span>(sepal_length))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: NAs introduced by coercion</code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>sepal_length[problems]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "5,6"   "&gt; 7.7"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># replace the comma </span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>sepal_length <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(sepal_length, <span class="st">","</span>, <span class="st">"."</span>)</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a><span class="co"># what was the result?</span></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>sepal_length[problems]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "5.6"   "&gt; 7.7"</code></pre>
</div>
</div>
<p>It worked! The <code>str_replace_all</code> is a general search-and-replace function. You can always use it with character vectors to replace stuff.</p>
<p>However, we still have the problem with <code>&gt; 7.7</code>. You will find such values quite often in clinical settings, along with the counterpart <code>&lt; 7.7</code>. These may indicate that the measurement was out of range of an instrument. However, we cannot usually treat them as missing values, because they contain some information. Depending on whether or not we want to keep this information, we could either replace them by NAs, or decide to keep the value as is (in this case, change <code>&gt; 7.7</code> to <code>7.7</code>).</p>
<p>In the latter case, we can use the <code>str_replace_all</code> function again:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>sepal_length <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(sepal_length, <span class="st">"&gt; "</span>, <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In any case, the last point that remains is to convert the vector to numbers and assign it to the column holding sepal lengths. We do it in one go and also check if everything is OK:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co"># finalize</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>sepal_length <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(sepal_length)</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether the column is numeric</span></span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a><span class="fu">is.numeric</span>(iris_data<span class="sc">$</span>sepal_length)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether our problems are gone</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>sepal_length[problems]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 5.6 7.7</code></pre>
</div>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check whether there are any NA's</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">is.na</span>(iris_data<span class="sc">$</span>sepal_length))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
</div>
<div class="page-columns page-full"><p> The new function, <code>is.numeric()</code>, checks whether the column <code>sepal_length</code> is indeed numeric. Finally, we make sure that no NA’s were produced in the conversion.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>as.numeric()</code></span></div></div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-str_replace" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.9</strong></span> Find the problems in the following vector and correct them:</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">" 5"</span>, <span class="st">"5,6"</span>, <span class="st">"5.7"</span>, <span class="st">"5. 4"</span>, <span class="st">"&gt; 5.0"</span>, <span class="st">"6.O"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="regular-expressions" class="level2 page-columns page-full" data-number="3.5">
<h2 data-number="3.5" class="anchored" data-anchor-id="regular-expressions"><span class="header-section-number">3.5</span> Regular expressions</h2>
<section id="introduction-to-regular-expressions" class="level3 page-columns page-full" data-number="3.5.1">
<h3 data-number="3.5.1" class="anchored" data-anchor-id="introduction-to-regular-expressions"><span class="header-section-number">3.5.1</span> Introduction to regular expressions</h3>
<div class="page-columns page-full"><p> Regular expressions, regexps for short, are for finding certain patterns in the data and handling them automatically. They are quite old (going back at least to 1960’s) and have remained virtually the same for many decades. You will find them in almost all programming languages, and while details may differ, the principles are usually the same. The regexps in R are very similar to regular expressions in other programming languages such as Python. As you may guess, they have a different syntax then R – they are something else altogether.</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Regular expressions</span></div></div>
<p>We will not go in depth in regular expressions in R here. Here, just a short primer.</p>
<div class="page-columns page-full"><p> Before we start, let us introduce a new function called <code>str_detect()</code><a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>. The function is available as soon as you load the <code>tidyverse</code> package. This function is used to search for a pattern in a character vector. For every element that matches the pattern, it will return a <code>TRUE</code>, otherwise a <code>FALSE</code>. Say, we have a vector (or data frame column) containing sample names, and we want to find all the controls.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>str_detect()</code></span><div id="fn5"><p><sup>5</sup>&nbsp;In base R, there is the <code>grep()</code> function. It is very similar, but has a different syntax. We will stick to <code>str_detect()</code> for now.</p></div></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ko_1_ctrl"</span>, <span class="st">"ko_2_ctrl"</span>, <span class="st">"ko_1_treat"</span>, <span class="st">"ko_2_treat"</span>,</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>            <span class="st">"wt_1_ctrl"</span>, <span class="st">"wt_2_ctrl"</span>, <span class="st">"wt_1_treat"</span>, <span class="st">"wt_2_treat"</span>)</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="fu">str_detect</span>(samples, <span class="st">"ctrl"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  TRUE  TRUE FALSE FALSE  TRUE  TRUE FALSE FALSE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>samples[<span class="fu">str_detect</span>(samples, <span class="st">"ctrl"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ko_1_ctrl" "ko_2_ctrl" "wt_1_ctrl" "wt_2_ctrl"</code></pre>
</div>
</div>
<p>In case that someone did not pay attention to lower or upper case, we can tell str_detect to ignore the case:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ko_1_ctrl"</span>, <span class="st">"ko_2_CTRL"</span>, <span class="st">"ko_1_treat"</span>, <span class="st">"ko_2_treat"</span>,</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>            <span class="st">"wt_1_CTRL"</span>, <span class="st">"wt_2_ctrl"</span>, <span class="st">"wt_1_treat"</span>, <span class="st">"wt_2_treat"</span>)</span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="co"># this does not work</span></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a><span class="fu">str_detect</span>(samples, <span class="st">"ctrl"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  TRUE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># but this does</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str_detect</span>(samples, <span class="fu">regex</span>(<span class="st">"ctrl"</span>, <span class="at">ignore_case=</span><span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  TRUE  TRUE FALSE FALSE  TRUE  TRUE FALSE FALSE</code></pre>
</div>
</div>
<p>Here we are looking for a literal string “ctrl”. However, we can use regular expressions to search for more complex patterns. What if the sample names are more hap-hazard, like <code>ko_1_control</code>, <code>ko_2_kontrol</code>, <code>ko_1_ctrl</code>?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ko_1_control"</span>, <span class="st">"ko_2_ctrl"</span>, <span class="st">"ko_1_trt"</span>, <span class="st">"ko_2_treatment"</span>,</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>            <span class="st">"wt_1_Control"</span>, <span class="st">"wt_2_kontrol"</span>, <span class="st">"wt_1_Trtmt"</span>, <span class="st">"wt_2_treated"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>OK, so we got <code>control</code>, <code>ctrl</code>, <code>Control</code> and <code>kontrol</code>. A pattern emerges:</p>
<ul>
<li>first letter is either <code>k</code> or <code>c</code> or <code>C</code></li>
<li>then we might have an <code>on</code> (or not)</li>
<li>next always comes <code>tr</code></li>
<li>then we might have an <code>o</code></li>
<li>then we always have an <code>l</code></li>
<li>then the word ends.</li>
</ul>
<p>All this “mights” and “ors” and “eithers” and “always” can be encoded in a regular expression:</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Pattern</th>
<th>Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>[kcC]</code></td>
<td>must have one of the letters <code>k</code>, <code>c</code> or <code>C</code></td>
</tr>
<tr class="even">
<td><code>o?</code></td>
<td>zero or one <code>o</code> (i.e., “there might be an o”)</td>
</tr>
<tr class="odd">
<td><code>n?</code></td>
<td>zero or one <code>n</code> (i.e., “there might be an n”)</td>
</tr>
<tr class="even">
<td><code>tr</code></td>
<td>must have literally the string <code>tr</code></td>
</tr>
<tr class="odd">
<td><code>o?</code></td>
<td>zero or one <code>o</code> (i.e., “there might be an o”)</td>
</tr>
<tr class="even">
<td><code>l</code></td>
<td>must have literally the string <code>l</code></td>
</tr>
<tr class="odd">
<td><code>$</code></td>
<td>here must be the end of the string</td>
</tr>
</tbody>
</table>
<p>Taken together, this gives us the regular expression <code>[kcC]o?n?tro?l$</code>. This looks weird, but has the magic power of matching all the control-like strings in our vector. Let us try it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str_detect</span>(samples, <span class="st">"[kcC]o?n?tro?l$"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  TRUE  TRUE FALSE FALSE  TRUE  TRUE FALSE FALSE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>samples[<span class="fu">str_detect</span>(samples, <span class="st">"[kcC]o?n?tro?l$"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ko_1_control" "ko_2_ctrl"    "wt_1_Control" "wt_2_kontrol"</code></pre>
</div>
</div>
<p>This works, but… what does that all mean?</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-regex1" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.10 (Quick question)</strong></span> Look at the table above. Can you think of other patterns that would match this regular expression? Would <code>kotrl</code> work? What about <code>cotro</code>?</p>
</div>
</div>
</div>
</div>
</section>
<section id="how-regexp-works" class="level3 page-columns page-full" data-number="3.5.2">
<h3 data-number="3.5.2" class="anchored" data-anchor-id="how-regexp-works"><span class="header-section-number">3.5.2</span> How regexp works</h3>
<div class="page-columns page-full"><p> A regular expression is a sequence of characters – a string. However, some of these strings have a special meaning. Here is a short table, we will discuss these in more detail in a moment:</p><div class="no-row-height column-margin column-container"><span class="margin-aside">regexp rules</span></div></div>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Character</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>a</code></td>
<td>a literal character <code>a</code></td>
</tr>
<tr class="even">
<td><code>.</code></td>
<td>any character</td>
</tr>
<tr class="odd">
<td><code>[abc]</code></td>
<td>one of the characters <code>a</code>, <code>b</code> or <code>c</code></td>
</tr>
<tr class="even">
<td><code>[^abc]</code></td>
<td>any character <em>except</em> <code>a</code>, <code>b</code> or <code>c</code></td>
</tr>
<tr class="odd">
<td><code>[a-z]</code></td>
<td>any character from <code>a</code> to <code>z</code></td>
</tr>
<tr class="even">
<td><code>[0-9]</code></td>
<td>any digit</td>
</tr>
</tbody>
</table>
<p>These are the so-called “atoms”. They stand for something that we want to match. In the simplest case (like the <code>a</code> above), they stand for themselves. Also, did you notice the square brackets? They have <em>nothing</em> to do with how we use square brackets in R.</p>
<p>In addition we can specify how many times we want to match the atom:</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Quantifier</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>?</code></td>
<td>zero or one</td>
</tr>
<tr class="even">
<td><code>*</code></td>
<td>zero or more</td>
</tr>
<tr class="odd">
<td><code>+</code></td>
<td>at least one</td>
</tr>
</tbody>
</table>
<p>Then, we can “anchor” the pattern in our string:</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Anchor</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>^</code></td>
<td>start of the string</td>
</tr>
<tr class="even">
<td><code>$</code></td>
<td>end of the string</td>
</tr>
</tbody>
</table>
<p>Let us put that together. Say we have the following strings:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>strings <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"a"</span>, <span class="st">"ab"</span>, <span class="st">"ac"</span>, <span class="st">"ad"</span>, <span class="st">"bc"</span>, <span class="st">"abc"</span>, <span class="st">"bac"</span>, <span class="st">"cab"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To match these which contain an <code>a</code>, we can use the regular expression <code>a</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>strings[<span class="fu">str_detect</span>(strings, <span class="st">"a"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "a"   "ab"  "ac"  "ad"  "abc" "bac" "cab"</code></pre>
</div>
</div>
<p>To match these which <em>start</em> with an <code>a</code>, we can use the regular expression <code>^a</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>strings[<span class="fu">str_detect</span>(strings, <span class="st">"^a"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "a"   "ab"  "ac"  "ad"  "abc"</code></pre>
</div>
</div>
<p>To find these which start with an <code>a</code> followed by <code>b</code> or <code>c</code>, but not <code>d</code>, we can use the square brackets:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>strings[<span class="fu">str_detect</span>(strings, <span class="st">"^a[bc]"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ab"  "ac"  "abc"</code></pre>
</div>
</div>
<p>OK, that was a lot. Take some time to digest it. Regulare expressions are what you call in German <em>gewöhnungsbedürftig</em> – they require some getting used to, but they are worth it.</p>
</section>
<section id="usage-of-regular-expressions-in-r" class="level3 page-columns page-full" data-number="3.5.3">
<h3 data-number="3.5.3" class="anchored" data-anchor-id="usage-of-regular-expressions-in-r"><span class="header-section-number">3.5.3</span> Usage of regular expressions in R</h3>
<p>Most commonly you will use regular expressions in R for two things:</p>
<ul>
<li>Searching for something in a character vector (like we did above)</li>
<li>Replacing something in a character vector, for example using <code>str_replace_all</code></li>
</ul>
<p>What we did not tell you when we introduced <code>str_replace_all</code> is that it actually, its second argument is a regular expression. Therefore, we can use it to find a pattern and replace it with something else. For example, we can unify the following messed up vector denoting the gender of patients:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>gender <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"m"</span>, <span class="st">"f"</span>, <span class="st">"m"</span>, <span class="st">"w"</span>, <span class="st">"frau"</span>,</span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>            <span class="st">"female"</span>, <span class="st">"other"</span>, <span class="st">"male"</span>, </span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a>            <span class="st">"männlich"</span>, <span class="st">"x"</span>, <span class="st">"weiblich"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>German and English are mixed here, but we can see a pattern: if the strings starts with <code>m</code>, it is male, if it starts with either <code>f</code> or <code>w</code>, it is female. And in remaining cases it is “other”. We can clean up this mess with just three lines of code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>gender <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(gender, <span class="st">"^m.*"</span>, <span class="st">"male"</span>)</span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>gender <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(gender, <span class="st">"^[fw].*"</span>, <span class="st">"female"</span>)</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a>gender <span class="ot">&lt;-</span> <span class="fu">str_replace</span>(gender, <span class="st">"^[^mfw].*"</span>, <span class="st">"other"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that the <code>^</code> character has two different meanings on line 3 above. As the first character of a regular expression, it anchors it to the beginning of the string. However, inside the square brackets, it negates the selection. So, <code>[^mfw]</code> means “any character except <code>m</code>, <code>f</code> or <code>w</code>”.</p>
<p>The <code>.*</code> idiom is a common one in regular expressions. It means “zero or more of any character”. So, <code>^m.*</code> matches both the string “m” and the string “male”. And because we used <code>^</code>, it will only match the strings where m is the first letter (so it does not match “female”).</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-regex1" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.11 (Quick question)</strong></span> What would happen if we ommited the <code>^</code> at the beginning of the strings above? For example, if we used <code>[^mfw].*</code> instead of <code>^[^mfw].*</code>? Think first, then try it out.</p>
</div>
</div>
</div>
</div>
<p>But wait, if some characters have a special meaning, how can we replace them? For example, how can we replace a dot in a string? The following will not work as intended:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"5.6"</span>, <span class="st">"5.7"</span>, <span class="st">"5.8"</span>)</span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str_replace_all</span>(vec, <span class="st">"."</span>, <span class="st">","</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] ",,," ",,," ",,,"</code></pre>
</div>
</div>
<div class="page-columns page-full"><p>Since the character <code>.</code> means “any character”, <em>every character</em> will be replaced by a comma in the example above. In order to search or replace  special characters, we must <em>escape</em> them – which in R means putting two backslashes<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a> in front of them in the regular expression.</p><div class="no-row-height column-margin column-container"><span class="margin-aside">escaping characters in a regexp</span><div id="fn6"><p><sup>6</sup>&nbsp;In R, it is two backslashes. In other programming languages, it is usually a single backslash.</p></div></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"5.6"</span>, <span class="st">"5.7"</span>, <span class="st">"5.8"</span>)</span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str_replace_all</span>(vec, <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>, <span class="st">","</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "5,6" "5,7" "5,8"</code></pre>
</div>
</div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-regex2" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.12</strong></span> &nbsp;</p>
<ul>
<li>Use <code>str_replace_all()</code> to make the following uniform: <code>c("male", "Male ", "M", "F", "female", " Female")</code></li>
<li>Using <code>str_replace_all()</code> and <code>toupper()</code>, clean up the gene names such that they conform to the HGNC (all capital letters, no spaces, no dashes): <code>c("ankrd22", "ifng", "Nf-kb", " Cxcl 5", "CCL 6.", "ANK.r.d. 12")</code></li>
<li>What regular expression matches all of the ankyrin repeat genes (but not other genes) in the following vector: <code>c("ANKRD22", "ank.rep.d. 12", "ANKRD-33", "ankrd23", "ANKEN", "MAPK", "ifng-1", "ANKA-REP-6")</code>? Ankyrin repeat domain genes are the first 4 in the vector.</li>
</ul>
</div>
</div>
</div>
</div>
</section>
<section id="correcting-columns-in-iris_data-with-regular-expressions" class="level3 page-columns page-full" data-number="3.5.4">
<h3 data-number="3.5.4" class="anchored" data-anchor-id="correcting-columns-in-iris_data-with-regular-expressions"><span class="header-section-number">3.5.4</span> Correcting columns in <code>iris_data</code> with regular expressions</h3>
<p>Let us now turn to another column in the data frame, the petal lengths. Using the approach we have just learned, we can find the problematic values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>petal_length <span class="ot">&lt;-</span> iris_data<span class="sc">$</span>petal_length</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>problems <span class="ot">&lt;-</span> <span class="fu">is.na</span>(<span class="fu">as.numeric</span>(petal_length))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: NAs introduced by coercion</code></pre>
</div>
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(problems)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  12  13  14 115 149</code></pre>
</div>
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>petal_length[problems]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "1.6!" "1 .4" "1.1?" "5k.1" "5. 4"</code></pre>
</div>
</div>
<p>We <em>could</em> use for example <code>str_replace_all(sepal_length, "k", "")</code>, to change <code>5k.1</code> to <code>5.1</code>, but that would not be a general solution. What if it is <code>j</code> in the new data? We should actually remove <em>everything</em> that is not a number and not a decimal dot. To this end, we will use the regular expressions.</p>
<p>We need to remove anything <em>but</em> numbers and decimal dots. We can use the square brackets for that:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>petal_length <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(petal_length, <span class="st">"[^0-9.]"</span>, <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="page-columns page-full"><p>The <code>0-9</code> means anything between 0 and 9, and the <code>.</code> is a literal dot. As you can see, you don’t have to escape the dot if it is already in the  square brackets. The <code>^</code> inside the square brackets negates the selection, so <code>[^0-9.]</code> means exactly what we wanted.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>[^0-9]</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check the problems</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>petal_length[problems]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "1.6" "1.4" "1.1" "5.1" "5.4"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to numbers</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>petal_length <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(petal_length)</span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a><span class="co"># check for remaining NA's</span></span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">is.na</span>(petal_length))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co"># assign</span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>petal_length <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(petal_length)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Done! The iris data set is now clean.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="fu">skim</span>(iris_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td style="text-align: left;">Name</td>
<td style="text-align: left;">iris_data</td>
</tr>
<tr class="even">
<td style="text-align: left;">Number of rows</td>
<td style="text-align: left;">150</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Number of columns</td>
<td style="text-align: left;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">_______________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Column type frequency:</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">character</td>
<td style="text-align: left;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">________________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Group variables</td>
<td style="text-align: left;">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: character</strong></p>
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 19%">
<col style="width: 13%">
<col style="width: 19%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 8%">
<col style="width: 12%">
<col style="width: 15%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">min</th>
<th style="text-align: right;">max</th>
<th style="text-align: right;">empty</th>
<th style="text-align: right;">n_unique</th>
<th style="text-align: right;">whitespace</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">species</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: numeric</strong></p>
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 18%">
<col style="width: 13%">
<col style="width: 18%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 6%">
<col style="width: 5%">
<col style="width: 6%">
<col style="width: 7%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">mean</th>
<th style="text-align: right;">sd</th>
<th style="text-align: right;">p0</th>
<th style="text-align: right;">p25</th>
<th style="text-align: right;">p50</th>
<th style="text-align: right;">p75</th>
<th style="text-align: right;">p100</th>
<th style="text-align: left;">hist</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">sepal_length</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">5.84</td>
<td style="text-align: right;">0.83</td>
<td style="text-align: right;">4.3</td>
<td style="text-align: right;">5.1</td>
<td style="text-align: right;">5.80</td>
<td style="text-align: right;">6.4</td>
<td style="text-align: right;">7.9</td>
<td style="text-align: left;">▆▇▇▅▂</td>
</tr>
<tr class="even">
<td style="text-align: left;">sepal_width</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3.06</td>
<td style="text-align: right;">0.44</td>
<td style="text-align: right;">2.0</td>
<td style="text-align: right;">2.8</td>
<td style="text-align: right;">3.00</td>
<td style="text-align: right;">3.3</td>
<td style="text-align: right;">4.4</td>
<td style="text-align: left;">▁▆▇▂▁</td>
</tr>
<tr class="odd">
<td style="text-align: left;">petal_length</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">3.76</td>
<td style="text-align: right;">1.77</td>
<td style="text-align: right;">1.0</td>
<td style="text-align: right;">1.6</td>
<td style="text-align: right;">4.35</td>
<td style="text-align: right;">5.1</td>
<td style="text-align: right;">6.9</td>
<td style="text-align: left;">▇▁▆▇▂</td>
</tr>
<tr class="even">
<td style="text-align: left;">petal_width</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1.20</td>
<td style="text-align: right;">0.76</td>
<td style="text-align: right;">0.1</td>
<td style="text-align: right;">0.3</td>
<td style="text-align: right;">1.30</td>
<td style="text-align: right;">1.8</td>
<td style="text-align: right;">2.5</td>
<td style="text-align: left;">▇▁▇▅▃</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-day3-correcting" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.13 (Correcting metadata)</strong></span> In <a href="#exr-day3-diagnosing" class="quarto-xref">Exercise&nbsp;<span>3.7</span></a>, you have diagnosed the file <code>meta_data_botched.xlsx</code>. Now go ahead and correct the problems you have found.</p>
</div>
</div>
</div>
</div>
</section>
<section id="sec-quick-look" class="level3 page-columns page-full" data-number="3.5.5">
<h3 data-number="3.5.5" class="anchored" data-anchor-id="sec-quick-look"><span class="header-section-number">3.5.5</span> Just a quick look</h3>
<div class="page-columns page-full"><p>The newly cleaned up data set begs for a quick look. We already created a simple plot before using the <code>plot()</code> function, and today I will show you how to use the <code>ggplot2</code> package to create somewhat nicer plots. In this case, it is rather trivial. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>ggplot2</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(iris_data, <span class="fu">aes</span>(<span class="at">x=</span>sepal_length, <span class="at">y=</span>sepal_width, <span class="at">color=</span>species)) <span class="sc">+</span></span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day3-reading-your-data_files/figure-html/unnamed-chunk-56-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>As you can see, the <code>ggplot</code> function takes two arguments: the data frame and something which is calles <em>aesthetics</em> and which is produced by the <code>aes()</code> function. The <code>aes()</code> function serves as a kind of link between the data and the plot, showing how what should be plotted. In this case, we tell <code>aes()</code> that the x-axis should be the sepal length, the y-axis the sepal width, and the color should be determined by the species.</p>
<p>Then, a magical thing happens. Whatever is produced by <code>ggplot()</code> is not a number, but nonetheless can be added with a <code>+</code> sign to other stuff. In this case, we add a <code>geom_point()</code> function, which tells <code>ggplot</code> to plot points at the x, y coordinates (there are many others, and you will learn about them the day after tomorrow).</p>
<p>There, all done. See how <em>I. setosa</em> is different from the other two?</p>
</section>
</section>
<section id="writing-data" class="level2 page-columns page-full" data-number="3.6">
<h2 data-number="3.6" class="anchored" data-anchor-id="writing-data"><span class="header-section-number">3.6</span> Writing data</h2>
<section id="keep-your-data-organized" class="level3" data-number="3.6.1">
<h3 data-number="3.6.1" class="anchored" data-anchor-id="keep-your-data-organized"><span class="header-section-number">3.6.1</span> Keep your data organized</h3>
<p>Writing data is <em>way</em> easier than reading it. However, you always have to come up with a name for the file, and I would like you to follow a few points.</p>
<ul>
<li>Never overwrite the original file (we will discuss it in more detail shortly).</li>
<li>Create an output directory and save your files only there, do not mix the original and the output files.</li>
<li>Use a meaningful name, so not <code>output</code>, but <code>iris_cleaned</code>. No spaces or special characters (use an underscore <code>_</code> instead).</li>
<li>Thirdly, always version your files: add a date, your initials, a version number and a date, e.g.&nbsp;<code>iris_cleaned_2021-09-01_JW_v1.csv</code>.</li>
<li>Do not use the old Excel XLS format (file extension <code>.xls</code>). Use the newer XLSX format (file extension <code>.xlsx</code>).</li>
</ul>
<p>This latter point warrants an explanation. There are two main problems with that format. Firstly, it is less portable then XLSX (the new one) – so many programs can read XLSX, but not XLS (or, which is worse, they can, but read it incorrectly). Secondly, the number of rows and columns in an XLS file is severly limited (65536 rows and 256 columns). These limits are easily reached in modern bioinformatic datasets.</p>
</section>
<section id="functions-to-use-for-writing-data" class="level3 page-columns page-full" data-number="3.6.2">
<h3 data-number="3.6.2" class="anchored" data-anchor-id="functions-to-use-for-writing-data"><span class="header-section-number">3.6.2</span> Functions to use for writing data</h3>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>write_tsv()</code>, <code>write_csv()</code>,<code>write_xlsx()</code></span></div></div>
<table class="caption-top table">
<colgroup>
<col style="width: 24%">
<col style="width: 26%">
<col style="width: 26%">
<col style="width: 22%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Data type</th>
<th style="text-align: left;">Function</th>
<th style="text-align: left;">Package</th>
<th style="text-align: left;">Notes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">TSV / TAB separated values</td>
<td style="text-align: left;"><code>write_tsv()</code></td>
<td style="text-align: left;"><code>readr</code></td>
<td style="text-align: left;">No rownames!</td>
</tr>
<tr class="even">
<td style="text-align: left;">CSV / comma separated</td>
<td style="text-align: left;"><code>write_csv()</code></td>
<td style="text-align: left;"><code>readr</code></td>
<td style="text-align: left;">No rownames!</td>
</tr>
<tr class="odd">
<td style="text-align: left;">XLS (old Excel)</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">Just don’t use it. No, seriously, don’t.</td>
</tr>
<tr class="even">
<td style="text-align: left;">XLSX (new Excel)</td>
<td style="text-align: left;"><code>write_xlsx()</code></td>
<td style="text-align: left;"><code>writexl</code></td>
<td style="text-align: left;">No rownames!</td>
</tr>
</tbody>
</table>
<p>Just as in the case of reading data, there are several functions to write data. There are also functions in the base R that can be used for writing (e.g.&nbsp;<code>write.csv</code>), but they are not recommended. We advise you to only use the <code>readr</code> package for writing TSV and CSV files, and the <code>writexl</code> package for writing XLSX files.</p>
<p>However, keep in mind that row names are not exported with these packages. That is why we do not use row names in this course.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-writing-data" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 3.14 (Writing data)</strong></span> &nbsp;</p>
<ul>
<li>In your project directory, create the directory “Data_clean” (if you want, you can use the R function <code>dir.create()</code> for that).</li>
<li>Write the cleaned <code>iris_data</code> to a new file in the “Data_clean” directory. Use the <code>write_csv</code> function from the <code>readr</code> package. Use a meaningful name, version it and use a date.</li>
</ul>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="principles-of-data-management" class="level2 page-columns page-full" data-number="3.7">
<h2 data-number="3.7" class="anchored" data-anchor-id="principles-of-data-management"><span class="header-section-number">3.7</span> Principles of data management</h2>
<section id="keeping-originals" class="level3" data-number="3.7.1">
<h3 data-number="3.7.1" class="anchored" data-anchor-id="keeping-originals"><span class="header-section-number">3.7.1</span> Keeping originals</h3>
<p>As every day, we end the day with some food for thought. Today, we would like to introduce a few important principles that you should follow in your everyday work with data – not only in R, but in general.</p>
<p>The first rule is: <strong>always keep the originals</strong>. Whether you have received the data from someone, extracted them from a database such as RedCap or created it yourself, once you start working with R you should freeze your data, version it and not touch it anymore.</p>
<p>Let me reiterate: <strong>do not touch the original data</strong>. If you absolutely <em>have</em> to edit it by hand, make a copy and add a suffix like <code>_edited</code> or your initials to the file. However, it is better to perform all editing operations in R.</p>
<p><strong>Versioning.</strong> Always include a date, and possibly your initials and version in the file name. If you ever submitted a manuscript of a scientific paper, you know how quickly versions can accumulate. It is absolutely crucial to know which version of the data was used for which analysis.</p>
<p><strong>Data editing.</strong> Every edit operation on your data should be somehow recorded. This is not different from a lab notebook, where you write down every step of your experiment. When working with a spreadsheet this is hard to do, therefore…</p>
</section>
<section id="use-r-for-data-editing" class="level3 page-columns page-full" data-number="3.7.2">
<h3 data-number="3.7.2" class="anchored" data-anchor-id="use-r-for-data-editing"><span class="header-section-number">3.7.2</span> Use R for data editing</h3>
<p>Optimally, your mode of operation should never involve editing data directly in a spreadsheet. You should read your data in R and perform all operations here, recording them in a script. This has two huge advantages:</p>
<ul>
<li><strong>Reproducibility.</strong> You and others will be able to trace back all your operations. There will be no doubt whether and which data were edited and how.</li>
<li><strong>Automation.</strong> When you get updated data (and you will), for example when new samples arrived, you will be able to re-run your script (maybe with a few changes) and get the updated results with minimal effort. Otherwise, you will have to repeat every. Single. Edit. Again.</li>
</ul>
<p>Consider the example of the <code>iris.csv</code> file. We have seen that one of the values of sepal width seems to be missing a decimal dot. You <em>could</em> edit that in the Excel or CSV file directly. You might even make a note of that somewhere. However, chances are that you were wrong – maybe that <em>was</em> a flower wiht a 36 sepal width after all. This change influences your data and your analysis results and <em>must</em> be documented if you were to do science. If you did it in R, you probably have entered a comment in your script, something like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="co"># There seems to be a missing decimal dot in some of the sepal widths</span></span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Changing it to 1/10th of the value</span></span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a>sepal_width <span class="ot">&lt;-</span> iris_data[[<span class="st">"Sepal Width"</span>]]</span>
<span id="cb121-5"><a href="#cb121-5" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">&lt;-</span> <span class="fu">which</span>(sepal_width <span class="sc">&gt;</span> <span class="dv">10</span>)</span>
<span id="cb121-6"><a href="#cb121-6" aria-hidden="true" tabindex="-1"></a>sepal_width[sel] <span class="ot">&lt;-</span> sepal_width[sel] <span class="sc">/</span> <span class="dv">10</span></span>
<span id="cb121-7"><a href="#cb121-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-8"><a href="#cb121-8" aria-hidden="true" tabindex="-1"></a>iris_data[[<span class="st">"Sepal Width"</span>]] <span class="ot">&lt;-</span> sepal_width</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This concisely informs the reader that a change was made, how and why.</p>
<div class="page-columns page-full"><p>And what happens when Ronald Fisher raises from his grave and collects  another 150 flowers? You will get an updated data set with 300 flowers and import it <em>again</em>. In the best case scenario, you will find the problem again (or remember it). Then you will have to go into Excel and make the same edit again. And all the other edits as well. If you did it in R, you simply use your script again.</p><div class="no-row-height column-margin column-container"><span class="margin-aside">undead Ronald Fisher</span></div></div>
</section>
<section id="working-with-excel-and-other-spreadsheets" class="level3 page-columns page-full" data-number="3.7.3">
<h3 data-number="3.7.3" class="anchored" data-anchor-id="working-with-excel-and-other-spreadsheets"><span class="header-section-number">3.7.3</span> Working with Excel and other spreadsheets</h3>
<div class="page-columns page-full"><p> We all use Excel, LibreOffice Calc or Google Sheets. They are great tools, no doubt about that. However, by now you can see that certain operations make working with R harder, and some even that can mess up your data.</p><div class="no-row-height column-margin column-container"><span class="margin-aside">How to work with Excel</span></div></div>
<p>If possible, you should actually avoid working with Excel when working with scientific data. Excel does certain operations automatically (depending on the settings), so it can change your data <em>without leaving a trace of the changes</em>. This is the precise opposite of what you want in science.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Avoid working with Excel
</div>
</div>
<div class="callout-body-container callout-body">
<p>Excel can automatically replace strings with dates, like changing <code>MARCH9</code> into <code>9th of March</code>. However, <code>MARCH9</code> is a valid gene name. In fact, it turns out that a substantial proportion (about a third!) of all Excel files containing gene names and published as supplementary materials online contain gene names transformed to dates. Not only that, but even though that this has been discovered many years ago and even some genes were officialy renamed <em>because of Excel</em>, this is still a problem. And Excel is now able to recognize dates in many languages, exacerbating the problem <span class="citation" data-cites="abeysooriya2021gene">(<a href="references.html#ref-abeysooriya2021gene" role="doc-biblioref">Abeysooriya et al. 2021</a>)</span>.</p>
</div>
</div>
<p>However, sometimes it is unavoidable to work with Excel.</p>
<p><strong>Don’t’s:</strong></p>
<ul>
<li><strong>Using color and font to encode information</strong>, for example marking the different treatments with different background color or formatting controls in bold. While it is possible to read this information in R, it is not trivial. In addition, even outside of R formatting of cells can be lost when the data is passed around, and this would mean that the information is lost.</li>
<li><strong>Comments in the same cells as values</strong>, for example <code>50 (measured twice)</code>. This prevents the columns to be interpreted as numbers and has to be dealt with in R. Create a separate column for comments, even it is just a single comment in a row.</li>
<li><strong>Meta-data information in the header.</strong> Very often, column names contain additional information, for example units, dates, encoding (like “0 for male, 1 for female”). This makes it hard to both, import the data and actually use that information. Create a separate sheet in your spreadsheet for meta data, with one row per column name.</li>
<li><strong>Adding header and tail information</strong>. Just like in the example with <code>deaths.xlsx</code> file, additional lines in spreadsheets require you to carefully select what you wish to import. Avoid it if you can, use another sheet.</li>
<li><strong>More than one table on a spreadsheet</strong>. A spreadsheet should contain a single table. Otherwise, just like in case of header and tail information, you will have to spend some time cautiously selecting the areas that you want to import.</li>
<li><strong>Merging cells</strong>. Merged cells are a nightmare to work with in R. If you have to, unmerge them before importing the data.</li>
</ul>
<p><strong>Do’s:</strong></p>
<ul>
<li><strong>Use a single table per sheet</strong>. If you have more than one table, use more than one sheet.</li>
<li><strong>Use a single header row</strong>. If you have more than one header row, use create another sheet with meta-data on the column names.</li>
<li><strong>Switch off automatic conversion</strong>. Excel can automatically change your data in many ways. For example, it can change gene names to dates. It can also change the decimal separator, which can mess up your data. Newer versions of Excel allow to switch off this behaviour, so do it (<a href="https://techcommunity.microsoft.com/t5/microsoft-365-insider-blog/control-data-conversions-in-excel-for-windows-and-mac/ba-p/4215336">here are the instructions</a>).</li>
<li><strong>Control your import.</strong> When importing data from CSV files to TSV files, Excel allows you to control which fields are imported as text, which as dates etc. Use this feature to make sure that your data is correctly imported.</li>
</ul>
</section>
</section>
<section id="review" class="level2" data-number="3.8">
<h2 data-number="3.8" class="anchored" data-anchor-id="review"><span class="header-section-number">3.8</span> Review</h2>
<p>Things that you learned today:</p>
<ul>
<li>Reading and writing
<ul>
<li><code>read_csv</code>, <code>read_tsv</code>, <code>read_delim</code>, <code>read_xls</code>, <code>read_xlsx</code></li>
<li>reading a file from working directory</li>
<li>reading a file from a different directory using relative paths</li>
</ul></li>
<li>Paths
<ul>
<li>Get the current working directory with <code>getwd()</code></li>
<li>relative and absolute paths</li>
</ul></li>
<li>Diagnosing datasets
<ul>
<li>data types: <code>str</code>, <code>class</code>, <code>typeof</code></li>
<li>summaries: <code>summary</code>, <code>glimpse</code></li>
<li>categorical data: <code>unique</code>, <code>table</code></li>
<li><code>skim()</code> from the <code>skimr</code> package</li>
<li>diagnosis checklist</li>
</ul></li>
<li>Correcting datasets
<ul>
<li>converting character vectors to numbers with <code>as.numeric()</code></li>
<li>checking a vector with <code>is.numeric()</code></li>
<li>checking a logical vector with <code>any()</code></li>
<li>replacing values with <code>str_replace_all()</code></li>
</ul></li>
<li>Principles of data management
<ul>
<li>keeping originals</li>
<li>versioning</li>
<li>data editing</li>
<li>working with Excel</li>
</ul></li>
<li>Other
<ul>
<li>using help pages to find out about functions and their parameters</li>
</ul></li>
</ul>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-abeysooriya2021gene" class="csl-entry" role="listitem">
Abeysooriya, Mandhri, Megan Soria, Mary Sravya Kasu, and Mark Ziemann. 2021. <span>“Gene Name Errors: Lessons Not Learned.”</span> <em>PLoS Computational Biology</em> 17 (7): e1008984. <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008984">https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008984</a>.
</div>
</div>
</section>


</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./day2-complex-structures.html" class="pagination-link" aria-label="Complex data structures">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Complex data structures</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./day4-manipulating-your-data.html" class="pagination-link" aria-label="Manipulating data frames">
        <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Manipulating data frames</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>