<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.54">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>4&nbsp; Manipulating data frames – Five days of R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./day5-visualization-and-statistics.html" rel="next">
<link href="./day3-reading-your-data.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://hypothes.is/embed.js"></script>
<script>
  window.document.addEventListener("DOMContentLoaded", function (_event) {
    document.body.classList.add('hypothesis-enabled');
  });
</script>


<link rel="stylesheet" href="custom.css">
</head>

<body class="nav-sidebar floating nav-fixed slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Five days of R</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="./index.html" aria-current="page"> 
<span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day1-intro-to-r.html"> 
<span class="menu-text">One</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day2-complex-structures.html"> 
<span class="menu-text">Two</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day3-reading-your-data.html"> 
<span class="menu-text">Three</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day4-manipulating-your-data.html"> 
<span class="menu-text">Four</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day5-visualization-and-statistics.html"> 
<span class="menu-text">Five</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./solutions.html"> 
<span class="menu-text">Solutions</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/bihealth/RCrashcourse-book/" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./day4-manipulating-your-data.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Manipulating data frames</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day1-intro-to-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day2-complex-structures.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Complex data structures</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day3-reading-your-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Reading and Writing Files</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day4-manipulating-your-data.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Manipulating data frames</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day5-visualization-and-statistics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">R markdown, basic statistics and visualizations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix-wide-long.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendix: Wide vs long data</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix-more-stats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendix: More statistics and visualizations</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Solutions to Exercises</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#aims-for-today" id="toc-aims-for-today" class="nav-link active" data-scroll-target="#aims-for-today"><span class="header-section-number">4.1</span> Aims for today</a></li>
  <li><a href="#selecting-columns" id="toc-selecting-columns" class="nav-link" data-scroll-target="#selecting-columns"><span class="header-section-number">4.2</span> Selecting columns</a>
  <ul class="collapse">
  <li><a href="#selecting-columns-using-square-brackets" id="toc-selecting-columns-using-square-brackets" class="nav-link" data-scroll-target="#selecting-columns-using-square-brackets"><span class="header-section-number">4.2.1</span> Selecting columns using square brackets</a></li>
  <li><a href="#selecting-columns-using-tidyverse" id="toc-selecting-columns-using-tidyverse" class="nav-link" data-scroll-target="#selecting-columns-using-tidyverse"><span class="header-section-number">4.2.2</span> Selecting columns using tidyverse</a></li>
  <li><a href="#renaming-columns" id="toc-renaming-columns" class="nav-link" data-scroll-target="#renaming-columns"><span class="header-section-number">4.2.3</span> Renaming columns</a></li>
  </ul></li>
  <li><a href="#sorting-and-ordering" id="toc-sorting-and-ordering" class="nav-link" data-scroll-target="#sorting-and-ordering"><span class="header-section-number">4.3</span> Sorting and ordering</a>
  <ul class="collapse">
  <li><a href="#sorting-and-ordering-vectors" id="toc-sorting-and-ordering-vectors" class="nav-link" data-scroll-target="#sorting-and-ordering-vectors"><span class="header-section-number">4.3.1</span> Sorting and ordering vectors</a></li>
  <li><a href="#sorting-character-vectors" id="toc-sorting-character-vectors" class="nav-link" data-scroll-target="#sorting-character-vectors"><span class="header-section-number">4.3.2</span> Sorting character vectors</a></li>
  <li><a href="#sorting-data-frames-with-order" id="toc-sorting-data-frames-with-order" class="nav-link" data-scroll-target="#sorting-data-frames-with-order"><span class="header-section-number">4.3.3</span> Sorting data frames with <code>order()</code></a></li>
  <li><a href="#sorting-data-frames-with-tidyverse" id="toc-sorting-data-frames-with-tidyverse" class="nav-link" data-scroll-target="#sorting-data-frames-with-tidyverse"><span class="header-section-number">4.3.4</span> Sorting data frames with tidyverse</a></li>
  </ul></li>
  <li><a href="#modifying-and-filtering-data-frames-in-tidyverse" id="toc-modifying-and-filtering-data-frames-in-tidyverse" class="nav-link" data-scroll-target="#modifying-and-filtering-data-frames-in-tidyverse"><span class="header-section-number">4.4</span> Modifying and filtering data frames in tidyverse</a>
  <ul class="collapse">
  <li><a href="#modifying-with-mutate" id="toc-modifying-with-mutate" class="nav-link" data-scroll-target="#modifying-with-mutate"><span class="header-section-number">4.4.1</span> Modifying with <code>mutate()</code></a></li>
  <li><a href="#using-ifelse-with-mutate" id="toc-using-ifelse-with-mutate" class="nav-link" data-scroll-target="#using-ifelse-with-mutate"><span class="header-section-number">4.4.2</span> Using <code>ifelse()</code> with <code>mutate()</code></a></li>
  <li><a href="#filtering-with-logical-vectors-and-filter" id="toc-filtering-with-logical-vectors-and-filter" class="nav-link" data-scroll-target="#filtering-with-logical-vectors-and-filter"><span class="header-section-number">4.4.3</span> Filtering with logical vectors and <code>filter()</code></a></li>
  <li><a href="#using-the-in-operator" id="toc-using-the-in-operator" class="nav-link" data-scroll-target="#using-the-in-operator"><span class="header-section-number">4.4.4</span> Using the <code>%in%</code> operator</a></li>
  </ul></li>
  <li><a href="#combining-data-sets" id="toc-combining-data-sets" class="nav-link" data-scroll-target="#combining-data-sets"><span class="header-section-number">4.5</span> Combining data sets</a>
  <ul class="collapse">
  <li><a href="#binding-data-frames-with-rbind-and-removing-duplicates" id="toc-binding-data-frames-with-rbind-and-removing-duplicates" class="nav-link" data-scroll-target="#binding-data-frames-with-rbind-and-removing-duplicates"><span class="header-section-number">4.5.1</span> Binding data frames with <code>rbind()</code> and removing duplicates</a></li>
  <li><a href="#binding-data-frames-with-cbind" id="toc-binding-data-frames-with-cbind" class="nav-link" data-scroll-target="#binding-data-frames-with-cbind"><span class="header-section-number">4.5.2</span> Binding data frames with <code>cbind()</code></a></li>
  <li><a href="#merging-data-sets" id="toc-merging-data-sets" class="nav-link" data-scroll-target="#merging-data-sets"><span class="header-section-number">4.5.3</span> Merging data sets</a></li>
  <li><a href="#complex-merges" id="toc-complex-merges" class="nav-link" data-scroll-target="#complex-merges"><span class="header-section-number">4.5.4</span> Complex merges</a></li>
  <li><a href="#bringing-it-all-together" id="toc-bringing-it-all-together" class="nav-link" data-scroll-target="#bringing-it-all-together"><span class="header-section-number">4.5.5</span> Bringing it all together</a></li>
  </ul></li>
  <li><a href="#pipes-in-r" id="toc-pipes-in-r" class="nav-link" data-scroll-target="#pipes-in-r"><span class="header-section-number">4.6</span> Pipes in R</a>
  <ul class="collapse">
  <li><a href="#too-many-parentheses" id="toc-too-many-parentheses" class="nav-link" data-scroll-target="#too-many-parentheses"><span class="header-section-number">4.6.1</span> Too many parentheses</a></li>
  <li><a href="#pipes" id="toc-pipes" class="nav-link" data-scroll-target="#pipes"><span class="header-section-number">4.6.2</span> Pipes</a></li>
  </ul></li>
  <li><a href="#review" id="toc-review" class="nav-link" data-scroll-target="#review"><span class="header-section-number">4.7</span> Review</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Manipulating data frames</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="aims-for-today" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="aims-for-today"><span class="header-section-number">4.1</span> Aims for today</h2>
<ul>
<li>Searching, sorting and selecting</li>
<li>Matching and merging data</li>
<li>Pipes - writing readable code</li>
<li>Wide and long format</li>
</ul>
</section>
<section id="selecting-columns" class="level2 page-columns page-full" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="selecting-columns"><span class="header-section-number">4.2</span> Selecting columns</h2>
<section id="selecting-columns-using-square-brackets" class="level3" data-number="4.2.1">
<h3 data-number="4.2.1" class="anchored" data-anchor-id="selecting-columns-using-square-brackets"><span class="header-section-number">4.2.1</span> Selecting columns using square brackets</h3>
<p>As you know, if we want the actual column as vector, we use the <code>$</code> operator:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ara=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="at">bera=</span><span class="dv">6</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">cora=</span><span class="dv">11</span><span class="sc">:</span><span class="dv">15</span>, <span class="at">dora=</span><span class="dv">16</span><span class="sc">:</span><span class="dv">20</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>ara <span class="co"># same as df[["ara"]]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1 2 3 4 5</code></pre>
</div>
</div>
<p>Just a reminder: you don’t want to try to select a single column using square brackets (<code>df[, 1]</code>), because the behavior is different for data frames and tibbles. Better use <code>df$a</code> or <code>df[["a"]]</code>.</p>
<p>We will now discuss selecting and searching for more than one column. Quite often, you want to select only some columns from a data set, or maybe you want to change their order. As you learned on Day 2, you can select columns from a data frame using vectors. This is very similar to how you select elements from a matrix, or even a vector. You can use integers, negative integers, or column names.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select columns 1 to 2</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> df[ , <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># select anything but column 2</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> df[ , <span class="sc">-</span><span class="dv">2</span>]</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># select all columns in reverse order</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> df[ , <span class="fu">ncol</span>(df)<span class="sc">:</span><span class="dv">1</span>]</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># select columns ara and cora</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> df[ , <span class="fu">c</span>(<span class="st">"ara"</span>, <span class="st">"cora"</span>)]</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="co"># select all columns ara and cora, but in reverse order</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> df[ , <span class="fu">c</span>(<span class="st">"cora"</span>, <span class="st">"ara"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This is very similar to what we did when dealing with matrices, and actually similar to how we select elements from a vector. Note, however, that rather than using <code>4:1</code> in the line 8 above, we use <code>ncol(df):1</code>. This ensures that if data frame grows for some reason, we still get all the columns.</p>
</section>
<section id="selecting-columns-using-tidyverse" class="level3 page-columns page-full" data-number="4.2.2">
<h3 data-number="4.2.2" class="anchored" data-anchor-id="selecting-columns-using-tidyverse"><span class="header-section-number">4.2.2</span> Selecting columns using tidyverse</h3>
<div class="page-columns page-full"><p> Tidyverse has the <code>select()</code> function, which is a bit more explicit and readable. Most importantly, it also has extra features that make it easier to work with.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>select()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># select columns ara and cora</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> <span class="fu">select</span>(df, ara, cora)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># select columns ara to cora</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> <span class="fu">select</span>(df, ara<span class="sc">:</span>cora)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="co"># select anything but column bera</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> <span class="fu">select</span>(df, <span class="sc">-</span>bera)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Can you spot what is weird about the code above? Exactly! There are no quotes around the column names. One would expect that R should throw an error – after all, there is no variable “ara” defined yet. However, this is an extra feature of tidyverse. It can be confusing at first, but you will soon get to like it.</p>
<div class="page-columns page-full"><p> What’s more, you see the constructs like <code>ara:cora</code> or <code>-bera</code>. In the base R, <code>ara:cora</code> means “look up variables ara and cora and return all integers between these two values”. In Tidyverse, that means “select all columns from ara to cora”. Similarly, <code>-bera</code> means “select all columns except bera”. This is called <strong>tidy evaluation</strong> and it works only with some tidyverse functions.</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Tidy evaluation</span></div></div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Remember!
</div>
</div>
<div class="callout-body-container callout-body">
<p>Tidy evaluation only works with tidyverse functions!</p>
</div>
</div>
<div class="page-columns page-full"><p> Another nice feature of <code>select()</code> is that you can use the a few specialized helper functions, saving you tedious regular expression and column searches. For example, how would you select for columns that end with a particular suffix? This is a common enough task in data science – many clinical data sets have hundreds of columns and a systematic way of naming them, thus allowing to search for example all columns related to laboratory diagnostics. This can easily be done with the <code>ends_with()</code> function:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>ends_with()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> <span class="fu">select</span>(df, <span class="fu">ends_with</span>(<span class="st">"ora"</span>))</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(df2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "cora" "dora"</code></pre>
</div>
</div>
<div class="page-columns page-full"><p> There are many more such functions, like <code>starts_with()</code>, <code>contains()</code>, <code>matches()</code>, <code>one_of()</code>, <code>everything()</code> and even <code>last_col()</code> (for selecting the last column). You can find them all in the help page for <code>select()</code>.</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Other tidy selection functions</span></div></div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
To quote or not to quote?
</div>
</div>
<div class="callout-body-container callout-body">
<p>The fact that you don’t use quotes around column names in tidyverse is confusing for many. As a rule of thumb, at this stage you should use quotes around column names, unless:</p>
<ul>
<li>you are using <code>$</code> operator (e.g.&nbsp;<code>df$ara</code>)</li>
<li>you are using <code>select()</code>, <code>mutate()</code> or <code>filter()</code> from tidyverse</li>
<li>you are using another tidyverse function that uses tidy evaluation (look up the help page if unsure)</li>
</ul>
<p>Once you start to program your own functions, the tidy evaluation will be an additional source of confusion, but burn that bridge when you get to it.</p>
</div>
</div>
</section>
<section id="renaming-columns" class="level3 page-columns page-full" data-number="4.2.3">
<h3 data-number="4.2.3" class="anchored" data-anchor-id="renaming-columns"><span class="header-section-number">4.2.3</span> Renaming columns</h3>
<p>The select function has one more useful feature: you can directly rename the variables in the same call.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> <span class="fu">select</span>(df, <span class="at">Alpha=</span>ara, <span class="at">Gamma=</span>cora)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(df2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Alpha" "Gamma"</code></pre>
</div>
</div>
<div class="page-columns page-full"><p> However, there is another function which can be used for renaming specific columns, aptly named <code>rename()</code><a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. The way its syntax works it is well suitable for working with pipes (see below), and is a great deal easier than renaming the columns using <code>colnames()</code>.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>rename()</code></span><div id="fn1"><p><sup>1</sup>&nbsp;Unfortunately, there are many functions in R that are named <code>rename()</code>. Someone should really rename them (badum-tss).</p></div></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> <span class="fu">rename</span>(df, <span class="at">Alpha=</span>ara, <span class="at">Gamma=</span>cora)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(df2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Alpha" "bera"  "Gamma" "dora" </code></pre>
</div>
</div>
<p>As you can see, no selection was made here, only renaming. Again, no quotes are needed around the column names.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-day4-selecting-columns" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.1</strong></span> &nbsp;</p>
<ul>
<li>Read the file ‘Datasets/transcriptomics_results.csv’</li>
<li>What columns are in the file?</li>
<li>Select only the columns ‘GeneName’, ‘Description’, ‘logFC.F.D1’ and ‘qval.F.D1’</li>
<li>Rename the columns to ‘Gene’, ‘Description’, ‘LFC’ and ‘FDR’</li>
</ul>
<p>(<a href="solutions.html#sec-day4-selecting-columns">Solution</a>)</p>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="sorting-and-ordering" class="level2 page-columns page-full" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="sorting-and-ordering"><span class="header-section-number">4.3</span> Sorting and ordering</h2>
<section id="sorting-and-ordering-vectors" class="level3 page-columns page-full" data-number="4.3.1">
<h3 data-number="4.3.1" class="anchored" data-anchor-id="sorting-and-ordering-vectors"><span class="header-section-number">4.3.1</span> Sorting and ordering vectors</h3>
<p>Before we jump to sorting, filtering and ordering data frames, we should first discuss these operations on vectors. This is important, because although data frames can easily be sorted with functions such as <code>arrange()</code> from tidyverse, you will have situations in which you need to sort a vector, a list or another data type. Secondly, understanding the concepts of ordering will aid you also with your work with data frames.</p>
<div class="page-columns page-full"><p> The basic R function for sorting is, appropriately, <code>sort()</code>. It is as simple as it gets:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>sort()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>vec <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">rnorm</span>(<span class="dv">10</span>), <span class="dv">2</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>vec</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] -1.09  0.42  0.67  0.57  1.68 -0.57 -1.42 -0.82 -1.13  0.00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>(vec)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] -1.42 -1.13 -1.09 -0.82 -0.57  0.00  0.42  0.57  0.67  1.68</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>(vec, <span class="at">decreasing=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  1.68  0.67  0.57  0.42  0.00 -0.57 -0.82 -1.09 -1.13 -1.42</code></pre>
</div>
</div>
<p>The <code>rnorm()</code> function generates a vector of random numbers from a normal distribution (you met it on Day 2). The <code>round()</code> function with parameter <code>2</code> rounds the numbers to two decimal places. The <code>sort()</code> with the <code>decreasing=TRUE</code> argument sorts the vector in descending order.</p>
<div class="page-columns page-full"><p> Amazingly, it is hardly ever used. Mostly we need the <code>order()</code> function. Rather than returning the sorted vector, it returns the <em>indices</em> of the sorted vector:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>order()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>ord <span class="ot">&lt;-</span> <span class="fu">order</span>(vec)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>ord</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  7  9  1  8  6 10  2  4  3  5</code></pre>
</div>
</div>
<p>So, what does that mean? Take the first element of <code>ord</code>. It is 7. This means that the first value to appear in the sorted vector should be the element number 7 of the original vector, which is -1.42 and the smallest number in <code>vec</code> (check it!). Likewise, the second element of <code>ord</code> is 9, which points to -1.13, the second smallest element of <code>vec</code>.</p>
<p>So what happens if we use the vector <code>ord</code> to select element from the vector <code>vec</code>? We get the sorted vector, that’s what:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>vec[ord]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] -1.42 -1.13 -1.09 -0.82 -0.57  0.00  0.42  0.57  0.67  1.68</code></pre>
</div>
</div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-order" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.2 (Reverse sorting)</strong></span> How can you get the reverse order vector for vec?</p>
</div>
</div>
</div>
</div>
</section>
<section id="sorting-character-vectors" class="level3 page-columns page-full" data-number="4.3.2">
<h3 data-number="4.3.2" class="anchored" data-anchor-id="sorting-character-vectors"><span class="header-section-number">4.3.2</span> Sorting character vectors</h3>
<div class="page-columns page-full"><p> We can also sort character vectors, but the behavior might not be exactly what you expect. Take a look:</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Sorting character vectors</span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>chvec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"b"</span>, <span class="st">"a"</span>, <span class="st">"Zorro"</span>, <span class="st">"10"</span>, <span class="st">"Anton"</span>, <span class="st">"A"</span>, <span class="st">"2"</span>, <span class="st">"100"</span>, <span class="st">"zxy"</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>(chvec)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "10"    "100"   "2"     "a"     "A"     "Anton" "b"     "Zorro" "zxy"  </code></pre>
</div>
</div>
<p>OK, so what happened here? The <code>sort()</code> function sorts the vector in lexicographical order – just like you sort words in a dictionary (or, in the olden days, in a phone book). For the most part that seems quite intuitive, however take a look at the numbers: <code>100</code> comes before <code>2</code>.</p>
<p>Lexicographical order means that the words are sorted first by first letter, then by second etc. So a word starting with a <code>1</code> always comes before a word starting with a <code>2</code>, even if it <em>means</em> a number that is larger than 2.</p>
<p>This has one important implication for all of us who work with data. Identifiers, especially sample and patient identifiers are quite often a combination of letters and numbers, e.g.&nbsp;<code>ID1</code> or <code>Patient10</code>. This can cause problems when sorting, as <code>ID10</code> will come before <code>ID2</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>chvec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ID1"</span>, <span class="st">"ID2"</span>, <span class="st">"ID3"</span>, <span class="st">"ID4"</span>, <span class="st">"ID10"</span>, <span class="st">"ID20"</span>, <span class="st">"ID100"</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>(chvec)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ID1"   "ID10"  "ID100" "ID2"   "ID20"  "ID3"   "ID4"  </code></pre>
</div>
</div>
<p>There are several solutions for that, but here is one with the tools that you already know. First, we will extract the numbers from the strings, then we will sort the vector based on these numbers:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract numbers</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>chvec_n <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(chvec, <span class="st">"ID"</span>, <span class="st">""</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to numeric</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>chvec_n <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(chvec_n)</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="co"># get the order</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>ord <span class="ot">&lt;-</span> <span class="fu">order</span>(chvec_n)</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a><span class="co"># sort the original vector</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>chvec[ord]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ID1"   "ID2"   "ID3"   "ID4"   "ID10"  "ID20"  "ID100"</code></pre>
</div>
</div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-day4-lastname" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.3 (Sorting by last name)</strong></span> Here is a vector of names. Can you think of how to sort it by last name?</p>
<pre><code>persons &lt;- c("Henry Fonda", "Bob Marley", "Robert F. Kennedy", 
  "Bob Dylan", "Alan Rickman")</code></pre>
<p>(<a href="solutions.html#sec-day4-lastname">Solution</a>)</p>
</div>
</div>
</div>
</div>
</section>
<section id="sorting-data-frames-with-order" class="level3" data-number="4.3.3">
<h3 data-number="4.3.3" class="anchored" data-anchor-id="sorting-data-frames-with-order"><span class="header-section-number">4.3.3</span> Sorting data frames with <code>order()</code></h3>
<p>By now it should be clear that you can use this method to order data from data frames (or matrices, or other data types). We will show it on example of the <code>transcriptomics_results.csv</code> file that you have read in the <span class="quarto-unresolved-ref">?exr-selecting-columns</span>. I assume that you have read the file and selected the columns with something like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>tr_res <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"Datasets/transcriptomics_results.csv"</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>tr_res <span class="ot">&lt;-</span> <span class="fu">select</span>(tr_res, </span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Gene=</span>GeneName, Description,</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">logFC=</span>logFC.F.D1, <span class="at">FDR=</span>qval.F.D1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The transcriptomics data set comes from a vaccine study <span class="citation" data-cites="weiner2019characterization">(<a href="references.html#ref-weiner2019characterization" role="doc-biblioref">Weiner et al. 2019</a>)</span> and show the changes in gene expression after vaccination with a particular adjuvanted fluad vaccine. <code>logFC</code> stands for log fold change, and <code>FDR</code> is the false discovery rate (resulting from raw p-values being corrected using the Benjamini-Hochberg procedure). We will sort the data frame by decreasing <code>logFC</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>ord <span class="ot">&lt;-</span> <span class="fu">order</span>(tr_res<span class="sc">$</span>logFC, <span class="at">decreasing=</span><span class="cn">TRUE</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tr_res[ord, ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  Gene     Description                                            logFC      FDR
  &lt;chr&gt;    &lt;chr&gt;                                                  &lt;dbl&gt;    &lt;dbl&gt;
1 Q5D1D6   tc|Q5D1D6_CERAE (Q5D1D6) Guanylate binding protein 1,…  4.85 1.80e-14
2 ANKRD22  ref|Homo sapiens ankyrin repeat domain 22 (ANKRD22), …  4.81 5.55e-15
3 ETV7     ref|Homo sapiens ets variant 7 (ETV7), transcript var…  4.65 2.53e-15
4 SERPING1 ref|Homo sapiens serpin peptidase inhibitor, clade G …  4.52 5.95e-16
5 CXCL10   ref|Homo sapiens chemokine (C-X-C motif) ligand 10 (C…  4.38 1.51e-10
6 GBP1P1   ref|Homo sapiens guanylate binding protein 1, interfe…  4.11 2.33e-17</code></pre>
</div>
</div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-sorting" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.4 (Sorting data frames with order())</strong></span> &nbsp;</p>
<ul>
<li>Sort the data frame <code>tr_res</code> by increasing <code>FDR</code></li>
<li>Sort the data frame alphabetically by <code>Gene</code></li>
</ul>
</div>
</div>
</div>
</div>
</section>
<section id="sorting-data-frames-with-tidyverse" class="level3 page-columns page-full" data-number="4.3.4">
<h3 data-number="4.3.4" class="anchored" data-anchor-id="sorting-data-frames-with-tidyverse"><span class="header-section-number">4.3.4</span> Sorting data frames with tidyverse</h3>
<div class="page-columns page-full"><p> Tidyverse makes it simple to sort the data frames with the <code>arrange()</code> function:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>arrange()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>tr_res <span class="ot">&lt;-</span> <span class="fu">arrange</span>(tr_res, <span class="fu">desc</span>(logFC))</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tr_res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  Gene     Description                                            logFC      FDR
  &lt;chr&gt;    &lt;chr&gt;                                                  &lt;dbl&gt;    &lt;dbl&gt;
1 Q5D1D6   tc|Q5D1D6_CERAE (Q5D1D6) Guanylate binding protein 1,…  4.85 1.80e-14
2 ANKRD22  ref|Homo sapiens ankyrin repeat domain 22 (ANKRD22), …  4.81 5.55e-15
3 ETV7     ref|Homo sapiens ets variant 7 (ETV7), transcript var…  4.65 2.53e-15
4 SERPING1 ref|Homo sapiens serpin peptidase inhibitor, clade G …  4.52 5.95e-16
5 CXCL10   ref|Homo sapiens chemokine (C-X-C motif) ligand 10 (C…  4.38 1.51e-10
6 GBP1P1   ref|Homo sapiens guanylate binding protein 1, interfe…  4.11 2.33e-17</code></pre>
</div>
</div>
<div class="page-columns page-full"><p> Note the use of <code>desc()</code> function – rather than specifying an argument like <code>decreasing=TRUE</code>.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>desc()</code></span></div></div>
<p>Using <code>arrange()</code>, it is easy to sort by multiple columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>tr_res <span class="ot">&lt;-</span> <span class="fu">arrange</span>(tr_res, logFC, FDR)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tr_res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  Gene            Description                                      logFC     FDR
  &lt;chr&gt;           &lt;chr&gt;                                            &lt;dbl&gt;   &lt;dbl&gt;
1 DSC1            ref|Homo sapiens desmocollin 1 (DSC1), transcri… -1.97 3.33e-4
2 XLOC_010084     tc|Q6FUE3_CANGA (Q6FUE3) Similarity, partial (1… -1.76 7.59e-3
3 LRRC69          ref|Homo sapiens leucine rich repeat containing… -1.68 5.06e-2
4 ZMAT4           ref|Homo sapiens zinc finger, matrin-type 4 (ZM… -1.67 4.61e-2
5 PCSK6           ref|Homo sapiens proprotein convertase subtilis… -1.63 3.13e-2
6 ENST00000456460 Unknown                                          -1.59 2.53e-2</code></pre>
</div>
</div>
<p>It is also allowed to use the expressions like <code>logFC * FDR</code> to sort by the sum of the two columns (that doesn’t make much sense in this context, but you get the idea):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>tr_res <span class="ot">&lt;-</span> <span class="fu">arrange</span>(tr_res, logFC <span class="sc">*</span> FDR)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>abs()</code></span></div></div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-day4-sorting-tidyverse" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.5 (Sorting data frames with arrange())</strong></span> The <code>abs()</code> function returns the absolute value of a number, for example <code>abs(c(-3, 7))</code> returns <code>3, 7</code>. Use the <code>arrange()</code> function to sort the data frame <code>tr_res</code> by the decreasing absolute value of <code>logFC</code>.</p>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="modifying-and-filtering-data-frames-in-tidyverse" class="level2 page-columns page-full" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="modifying-and-filtering-data-frames-in-tidyverse"><span class="header-section-number">4.4</span> Modifying and filtering data frames in tidyverse</h2>
<section id="modifying-with-mutate" class="level3 page-columns page-full" data-number="4.4.1">
<h3 data-number="4.4.1" class="anchored" data-anchor-id="modifying-with-mutate"><span class="header-section-number">4.4.1</span> Modifying with <code>mutate()</code></h3>
<div class="page-columns page-full"><p> Yesteday we have been busy cleaning up data sets. To modify the columns of the data frame, we have been using simple assignment. This works, but there is another method in tidyverse. Just like with regular assignments, the <code>mutate()</code> function allows both, creating new columns and modifying existing ones.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>mutate()</code></span></div></div>
<p>Below we create a new column <code>logFC_abs</code> which contains the absolute value of the <code>logFC</code> column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>tr_res <span class="ot">&lt;-</span> <span class="fu">mutate</span>(tr_res, <span class="at">logFC_abs =</span> <span class="fu">abs</span>(logFC))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Arguably, this is not much simpler than using simple assignment:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>tr_res<span class="sc">$</span>logFC_abs <span class="ot">&lt;-</span> <span class="fu">abs</span>(tr_res<span class="sc">$</span>logFC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The main difference, as you see, is that <code>mutate()</code> operates on data frames – it takes a data frame as the first argument and returns a data frame. The advantages will be obvious later, when you start to use pipes (see below).</p>
</section>
<section id="using-ifelse-with-mutate" class="level3 page-columns page-full" data-number="4.4.2">
<h3 data-number="4.4.2" class="anchored" data-anchor-id="using-ifelse-with-mutate"><span class="header-section-number">4.4.2</span> Using <code>ifelse()</code> with <code>mutate()</code></h3>
<div class="page-columns page-full"><p> Say, you would like to create a new column that contains “up” for genes that are upregulated (i.e.&nbsp;<code>logFC &gt;= 0</code>) and “down” for genes that are downregulated (i.e.&nbsp;<code>logFC &lt; 0</code>). We <em>could</em> do it with simple assignment and logical vectors:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>ifelse()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># default value</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>tr_res<span class="sc">$</span>direction <span class="ot">&lt;-</span> <span class="st">"up"</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="co"># create logical vector</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>negative <span class="ot">&lt;-</span> tr_res<span class="sc">$</span>logFC <span class="sc">&lt;</span> <span class="dv">0</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a><span class="co"># change "up" to "down" where logFC &lt; 0</span></span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>tr_res<span class="sc">$</span>direction[negative] <span class="ot">&lt;-</span> <span class="st">"down"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>However, this is a frequent operation and a much more convenient way exists. It is often used in combination with mutate, although it also can work in a normal assignment:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>tr_res <span class="ot">&lt;-</span> <span class="fu">mutate</span>(tr_res, </span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">direction =</span> <span class="fu">ifelse</span>(logFC <span class="sc">&lt;</span> <span class="dv">0</span>, <span class="st">"down"</span>, <span class="st">"up"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>ifelse()</code> function takes three arguments: a logical vector, a value to return if the logical vector is <code>TRUE</code>, and a value to return if the logical vector is <code>FALSE</code>. In other words, it “converts” a logical vector replacing <code>TRUE</code> with the second argument and <code>FALSE</code> with the third.</p>
</section>
<section id="filtering-with-logical-vectors-and-filter" class="level3 page-columns page-full" data-number="4.4.3">
<h3 data-number="4.4.3" class="anchored" data-anchor-id="filtering-with-logical-vectors-and-filter"><span class="header-section-number">4.4.3</span> Filtering with logical vectors and <code>filter()</code></h3>
<p>As you have learned, you can filter data frames using logical vectors. This is relatively straightforward. For example, we might want to create a table with only genes that are significant, that is have an FDR &lt; 0.05:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>small_fdr <span class="ot">&lt;-</span> tr_res<span class="sc">$</span>FDR <span class="sc">&lt;</span> <span class="fl">0.05</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>significant <span class="ot">&lt;-</span> tr_res[small_fdr, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you use <code>nrow(significant)</code>, you will find that there are 1478 significant genes in the data set.</p>
<div class="page-columns page-full"><p> The same can be achieved using the <code>filter()</code> function from tidyverse:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>filter()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>significant <span class="ot">&lt;-</span> <span class="fu">filter</span>(tr_res, FDR <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Simlarly, we can use the <code>str_detect()</code> function (which you learned about yesterday) to select only genes that contain the word “interferon” in the description:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>interferon <span class="ot">&lt;-</span> <span class="fu">filter</span>(tr_res, </span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>                     <span class="fu">str_detect</span>(Description, <span class="st">"interferon"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are 80 such genes.</p>
<p>But what if we want to set <em>two</em> conditions? In statistics it is common to set a threshold not only for the p-value or FDR, but also for an effect size. In our case, the measure of the effect size is the log2 fold change. Thus, we might want to select only genes that are at leas 2-fold upregulated or down-regulated. Twofold upregulation corresponds to log2-fold change greater than 1, and 2-fold downregulation corresponds to log2-fold change less than -1.</p>
<p>One way is to use <code>filter()</code> with two parameters:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>significant <span class="ot">&lt;-</span> <span class="fu">filter</span>(tr_res, FDR <span class="sc">&lt;</span> <span class="fl">0.05</span>, <span class="fu">abs</span>(logFC) <span class="sc">&gt;</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="page-columns page-full"><p> However, there is a more elegant and versatile way. We can achieve the same effect by combining two logical vectors with the <code>&amp;</code> operator:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>&amp;</code> (logical operator)</span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>large_fc <span class="ot">&lt;-</span> <span class="fu">abs</span>(tr_res<span class="sc">$</span>logFC) <span class="sc">&gt;</span> <span class="dv">1</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>significant <span class="ot">&lt;-</span> tr_res[small_fdr <span class="sc">&amp;</span> large_fc, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Or, in tidyverse,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>significant <span class="ot">&lt;-</span> <span class="fu">filter</span>(tr_res, FDR <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> <span class="fu">abs</span>(logFC) <span class="sc">&gt;</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>&amp;</code> operator is the logical AND operator. It combines two logical vectors, returning (at the given position) <code>TRUE</code> only if both vectors are <code>TRUE</code>. Take a look:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>vec1 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>vec2 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>vec1 <span class="sc">&amp;</span> vec2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  TRUE FALSE FALSE FALSE</code></pre>
</div>
</div>
<p>At the first position both vectors are <code>TRUE</code>, so the result is <code>TRUE</code>. At position 2, the first vector is <code>TRUE</code> and the second is <code>FALSE</code>, so the result is <code>FALSE</code>. And so on. Try it.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-day4-logical-vectors" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.6 (Logical vectors)</strong></span> &nbsp;</p>
<ul>
<li>Create a logical vector that selects genes that are both significant (FDR &lt; 0.05) and contain the string “interferon” in the Description. How many are there? How many genes are significant and do not contain “interferon” in the Description column?</li>
<li>How many of the genes containing “interferon” in the Description are are guanylate binding proteins? You can recognize the GBPs by their Gene symbol – it starts with “GBP”. (<a href="solutions.html#sec-day4-logical-vectors">Solution</a>) gical</li>
</ul>
</div>
</div>
</div>
</div>
</section>
<section id="using-the-in-operator" class="level3 page-columns page-full" data-number="4.4.4">
<h3 data-number="4.4.4" class="anchored" data-anchor-id="using-the-in-operator"><span class="header-section-number">4.4.4</span> Using the <code>%in%</code> operator</h3>
<div class="page-columns page-full"><p> Another useful utility in searching data frames is the <code>%in%</code> operator. It compares two vectors; for each element of the first vector, it returns <code>TRUE</code> if that element is in the second vector, and <code>FALSE</code> otherwise.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>%in%</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>vec1 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"a"</span>, <span class="st">"b"</span>, <span class="st">"c"</span>)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>vec2 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"a"</span>, <span class="st">"b"</span>, <span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"z"</span>)</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>vec1 <span class="sc">%in%</span> vec2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  TRUE  TRUE FALSE</code></pre>
</div>
</div>
<p>The elements 1, of the result are <code>TRUE</code>, because both <code>a</code> and <code>b</code> are in <code>vec2</code>. The last element, however, is <code>FALSE</code>, because <code>c</code> is not in <code>vec2</code>.</p>
<p>With <code>%in%</code> we can for example select a specified subset of genes that we are interested in.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GBP1"</span>, <span class="st">"GBP2"</span>, <span class="st">"GBP3"</span>, <span class="st">"GBP4"</span>, <span class="st">"GBP5"</span>, <span class="st">"ANKRD22"</span>)</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>(tr_res, Gene <span class="sc">%in%</span> genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 7 × 6
  Gene    Description                         logFC      FDR logFC_abs direction
  &lt;chr&gt;   &lt;chr&gt;                               &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;    
1 GBP4    ref|Homo sapiens guanylate binding…  2.25 2.59e-21      2.25 up       
2 GBP5    ref|Homo sapiens guanylate binding…  2.92 1.39e-20      2.92 up       
3 GBP3    ref|Homo sapiens guanylate binding…  2.63 1.62e-15      2.63 up       
4 ANKRD22 ref|Homo sapiens ankyrin repeat do…  4.81 5.55e-15      4.81 up       
5 GBP2    ref|Homo sapiens guanylate binding…  1.8  1.17e-13      1.8  up       
6 GBP1    ref|Homo sapiens guanylate binding…  2.76 7.38e-12      2.76 up       
7 GBP3    ref|Homo sapiens guanylate binding…  1.27 3.40e- 3      1.27 up       </code></pre>
</div>
</div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-in" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.7</strong></span> Which of these genes are significant? Use the <code>&amp;</code> operator to combine the result from <code>%in%</code> with the result of <code>FDR &lt; 0.05</code>.</p>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="combining-data-sets" class="level2 page-columns page-full" data-number="4.5">
<h2 data-number="4.5" class="anchored" data-anchor-id="combining-data-sets"><span class="header-section-number">4.5</span> Combining data sets</h2>
<section id="binding-data-frames-with-rbind-and-removing-duplicates" class="level3 page-columns page-full" data-number="4.5.1">
<h3 data-number="4.5.1" class="anchored" data-anchor-id="binding-data-frames-with-rbind-and-removing-duplicates"><span class="header-section-number">4.5.1</span> Binding data frames with <code>rbind()</code> and removing duplicates</h3>
<p>The simplest way to combine two data frames is to bind them. This can be done horizontally (when you have the same number of observations – rows) or vertically, by stacking them on top of each other, when you have the same columns. The two functions are, respectively, <code>cbind()</code> and <code>rbind()</code>, and they work for both, matrices and data frames.</p>
<div class="page-columns page-full"><p> Let’s start with stacking them on top of each other. For example, we might have generated two data frames with interesting results: one with interferones and the other with guanylate binding proteins, and at some point we decide to have them in one data frame. Stacking them can be done with the <code>rbind()</code> (“row bind”) function:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>rbind()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>interferon <span class="ot">&lt;-</span> <span class="fu">filter</span>(tr_res, </span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>                     <span class="fu">str_detect</span>(Description, <span class="st">"interferon"</span>))</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>gbp <span class="ot">&lt;-</span> <span class="fu">filter</span>(tr_res,</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>              <span class="fu">str_detect</span>(Gene, <span class="st">"GBP"</span>))</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">rbind</span>(interferon, gbp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that this works <em>only</em> if the data frames have the same columns. If you were to try to bind two data frames with different columns, you would get an error:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">a=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="at">b=</span><span class="dv">6</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">a=</span><span class="dv">11</span><span class="sc">:</span><span class="dv">15</span>, <span class="at">c=</span><span class="dv">16</span><span class="sc">:</span><span class="dv">20</span>)</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rbind</span>(a, b)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in match.names(clabs, names(xi)): names do not match previous names</code></pre>
</div>
</div>
<p>OK, but if you have checked the <code>interferon</code> and <code>gbp</code> data frames, you would have noticed that there are some genes that are both interferones and guanylate binding proteins. You did not check, because I haven’t told you yet how to do it. Here is how:</p>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>intersect()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">intersect</span>(interferon<span class="sc">$</span>Gene, gbp<span class="sc">$</span>Gene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GBP1P1" "GBP2"   "GBP1"  </code></pre>
</div>
</div>
<div class="page-columns page-full"><p> The <code>intersect()</code> function returns the elements that are common to two vectors. As you can see, there are three: GBP1P1, GBP2 and GBP1. We can get rid of them from the <code>results</code> table thanks to the function <code>duplicated()</code>. For every element of a vector, <code>duplicated()</code> returns <code>TRUE</code> if the element occured previously in the vector. So for the vector <code>c(1, 2, 1)</code> the result will be <code>FALSE, FALSE, TRUE</code> - because the second <code>1</code> is duplicated.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>duplicated()</code></span></div></div>
<p>To remove the elements which are duplicated, however, we need to use the negation – <code>!</code> to get <code>TRUE</code> for every element which is <em>not</em> duplicated. Here it is in action:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># count the number of the duplicated results</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">duplicated</span>(results<span class="sc">$</span>Gene))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 19</code></pre>
</div>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># number of rows before removing dups</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(results)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 96</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove duplicates</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">filter</span>(results, <span class="sc">!</span><span class="fu">duplicated</span>(Gene))</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a><span class="co"># how many are left?</span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(results)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 77</code></pre>
</div>
</div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-dups" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.8</strong></span> You might have been expecting that there are only three duplicates; after all, only four genes were in common between <code>interferon</code> and <code>gbp</code>- So where do these other duplicates come from? What are they?</p>
</div>
</div>
</div>
</div>
</section>
<section id="binding-data-frames-with-cbind" class="level3 page-columns page-full" data-number="4.5.2">
<h3 data-number="4.5.2" class="anchored" data-anchor-id="binding-data-frames-with-cbind"><span class="header-section-number">4.5.2</span> Binding data frames with <code>cbind()</code></h3>
<p>Combining data frames (or matrices) horizontally is much easier, because you do not have to worry about column names. You just need to make sure that the number of rows is identical<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>.</p>
<div class="no-row-height column-margin column-container"><div id="fn2"><p><sup>2</sup>&nbsp;Strictly speaking, the recycling rules apply. See Day 1. This can be sometimes useful, but usually it is better to be more explicit.</p></div></div><div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>cbind()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>df1 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">a =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="at">b=</span><span class="dv">11</span><span class="sc">:</span><span class="dv">14</span>)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">c=</span><span class="dv">21</span><span class="sc">:</span><span class="dv">24</span>)</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cbind</span>(df1, df2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  a  b  c
1 1 11 21
2 2 12 22
3 3 13 23
4 4 14 24</code></pre>
</div>
</div>
<p>We will use <code>cbind()</code> tomorrow when running a principal component analysis (PCA).</p>
</section>
<section id="merging-data-sets" class="level3 page-columns page-full" data-number="4.5.3">
<h3 data-number="4.5.3" class="anchored" data-anchor-id="merging-data-sets"><span class="header-section-number">4.5.3</span> Merging data sets</h3>
<p>Quite often you will have to merge two data sets. For example, you might have one Excel file containing the covariates for the patients (like BMI, age, gender etc.) and another file containing the gene expression data. Both data frames must have some identifiers present that allow us to match the two data frames.</p>
<p>This is <em>not</em> a simple binding operation, because you do not have the guarantee that (a) the identifiers between these two data sets are in the same order or even that (b) both data frames have exactly the same sets of identifiers. Actually, the reverse is more common that not.</p>
<p>In order to merge two data frames we need first to identify by what to merge them. In the code below we will concoct two data frames that share some of the identifiers.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>ids_x <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"ID"</span>, <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>df_x <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ID=</span>ids_x, <span class="at">val_x=</span><span class="fu">rnorm</span>(<span class="dv">6</span>))</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>df_x</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   ID      val_x
1 ID7 -1.9189798
2 ID4 -0.0566305
3 ID1  0.9386393
4 ID5 -1.0920805
5 ID2 -0.1552096
6 ID3  0.1936370</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>ids_y <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"ID"</span>, <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>, <span class="dv">6</span>))</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>df_y <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ID=</span>ids_y, <span class="at">val_y=</span><span class="fu">rnorm</span>(<span class="dv">6</span>, <span class="dv">10</span>, <span class="dv">1</span>))</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>df_y</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   ID     val_y
1 ID7 10.272775
2 ID4 11.101763
3 ID6 10.181991
4 ID3 10.217116
5 ID8  8.330966
6 ID5  9.536251</code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">intersect</span>(ids_x, ids_y)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ID7" "ID4" "ID5" "ID3"</code></pre>
</div>
</div>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>sample()</code></span></div></div>
<p>The <code>sample()</code> function is very useful – it generates a random sample by choosing elements from a vector. This is what is called <em>sampling</em> and comes in handy in many situations. Here we use it to create two sets of IDs that are similar, but not identical. As you can see from the output of <code>intersect()</code>, only 4 elements out of the total 6 are common between the two data frames.</p>
<p>Merging can be done either with the base R function <code>merge()</code>, or with corresponding <code>*_join</code> functions from the tidyverse. Whether you prefer one or the other is a matter of preference; I like <code>merge()</code> and we will be using it in the examples below.</p>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside">Inner join, full join</span></div></div>
<p>First, however, let us think <em>how</em> we want to merge, or, more specifically, what to do with the identifiers that are present in one data frame, but not in the other? That depends only on what we want to achieve. Possibly we are not interested in any elements that are not present in <em>both</em> data frames. This is what is called in database lingo an <strong>inner join</strong> Alternatively, we might want to keep all possible elements, whether they are found in <code>df_x</code> or <code>df_y</code> only. This is called an <strong>full join</strong>.</p>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>merge()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># inner join by ID</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>joined <span class="ot">&lt;-</span> <span class="fu">merge</span>(df_x, df_y, <span class="at">by=</span><span class="st">"ID"</span>)</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>joined</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   ID      val_x     val_y
1 ID3  0.1936370 10.217116
2 ID4 -0.0566305 11.101763
3 ID5 -1.0920805  9.536251
4 ID7 -1.9189798 10.272775</code></pre>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># full join by ID</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>joined <span class="ot">&lt;-</span> <span class="fu">merge</span>(df_x, df_y, <span class="at">by=</span><span class="st">"ID"</span>, <span class="at">all=</span><span class="cn">TRUE</span>)</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>joined</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   ID      val_x     val_y
1 ID1  0.9386393        NA
2 ID2 -0.1552096        NA
3 ID3  0.1936370 10.217116
4 ID4 -0.0566305 11.101763
5 ID5 -1.0920805  9.536251
6 ID6         NA 10.181991
7 ID7 -1.9189798 10.272775
8 ID8         NA  8.330966</code></pre>
</div>
</div>
<p>As you can see, in the first case there are just 4 rows – as many as there are common elements between the ID columns of the two data frames. In the second case, we get all 8 possible IDs. Where the ID was present in one, but not the other data frame, the values were filled up with <code>NA</code>’s. To get the full join we used the parameter <code>all=TRUE</code>, standing for “use <em>all</em> IDs”.</p>
<p>Note: it is possible to omit the <code>by=</code> parameter specifying by what column to join the data frames. However, do yourself a favor and <em>always</em> specify the column to join on explicitely.</p>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside">Left join, right join</span></div></div>
<p>There are two more situations, allthough less common in practice. We might want to keep all elements of <code>df_x</code>, but discard any elements of <code>df_y</code> which are not present in <code>df_x</code>. This is called a <strong>left join</strong>. Or, vice versa, we will discard only these elements which are present in <code>df_x</code>, but not in <code>df_y</code>. This is called, you guessed it, a <strong>right join</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co"># left join by ID</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>joined <span class="ot">&lt;-</span> <span class="fu">merge</span>(df_x, df_y, <span class="at">by=</span><span class="st">"ID"</span>, <span class="at">all.x =</span> <span class="cn">TRUE</span>)</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>joined</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   ID      val_x     val_y
1 ID1  0.9386393        NA
2 ID2 -0.1552096        NA
3 ID3  0.1936370 10.217116
4 ID4 -0.0566305 11.101763
5 ID5 -1.0920805  9.536251
6 ID7 -1.9189798 10.272775</code></pre>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># right join by ID</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>joined <span class="ot">&lt;-</span> <span class="fu">merge</span>(df_x, df_y, <span class="at">by=</span><span class="st">"ID"</span>, <span class="at">all.y =</span> <span class="cn">TRUE</span>)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>joined</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   ID      val_x     val_y
1 ID3  0.1936370 10.217116
2 ID4 -0.0566305 11.101763
3 ID5 -1.0920805  9.536251
4 ID6         NA 10.181991
5 ID7 -1.9189798 10.272775
6 ID8         NA  8.330966</code></pre>
</div>
</div>
<p>As you can see, in case of the <em>left</em> join, there are missing values in the <em>right</em> column only, and in case of the <em>right</em> join – only in the <em>left</em> column.</p>
<p>Let us summarize it in a table:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 16%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 33%">
<col style="width: 36%">
</colgroup>
<thead>
<tr class="header">
<th>Type of join</th>
<th>all x</th>
<th>all y</th>
<th>merge options</th>
<th>tidyverse alternative</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>inner join</td>
<td>FALSE</td>
<td>FALSE</td>
<td><code>merge(x, y)</code></td>
<td><code>inner_join(x, y)</code></td>
</tr>
<tr class="even">
<td>left join</td>
<td>TRUE</td>
<td>FALSE</td>
<td><code>merge(x, y, all.x=TRUE)</code></td>
<td><code>left_join(x, y)</code></td>
</tr>
<tr class="odd">
<td>right join</td>
<td>FALSE</td>
<td>TRUE</td>
<td><code>merge(x, y, all.y=TRUE)</code></td>
<td><code>right_join(x, y)</code></td>
</tr>
<tr class="even">
<td>full join</td>
<td>TRUE</td>
<td>TRUE</td>
<td><code>merge(x, y, all=TRUE)</code></td>
<td><code>full_join(x, y)</code></td>
</tr>
</tbody>
</table>
<p>You don’t have to remember the names (inner, left, right, full) – just remember the principle: sometimes you need the common elements, sometimes you need all elements, and sometimes you need all elements from x or y only.</p>
<p><strong>Different identifier column names</strong>. There is one other thing that you should know, because it happens quite often. The two data frames might have identifiers in columns that have different names, for example <code>ID</code> in one and <code>PatientID</code> in the other. You can, of course, rename the columns, but you can also specify the columns to join on explicitly:</p>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>merge()</code> with different column names</span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>df_x <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ID=</span>ids_x, <span class="at">val_x=</span><span class="fu">rnorm</span>(<span class="dv">6</span>))</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>df_y <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">PatientID=</span>ids_y, <span class="at">val_y=</span><span class="fu">rnorm</span>(<span class="dv">6</span>, <span class="dv">10</span>, <span class="dv">1</span>))</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a><span class="co"># inner join by ID</span></span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>joined <span class="ot">&lt;-</span> <span class="fu">merge</span>(df_x, df_y, <span class="at">by.x=</span><span class="st">"ID"</span>, <span class="at">by.y=</span><span class="st">"PatientID"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The resulting data frame has the identifiers in the column name specified in the “x” data frame, that is – in this case – in the <code>ID</code> column.</p>
</section>
<section id="complex-merges" class="level3 page-columns page-full" data-number="4.5.4">
<h3 data-number="4.5.4" class="anchored" data-anchor-id="complex-merges"><span class="header-section-number">4.5.4</span> Complex merges</h3>
<p>The situation above is quite trivial. In real world, unfortunately, the situations can be quite complex.</p>
<p>First, there can be duplicates. For example, one data frame contains the meta data on a group of lab animals, while the second contains the results of the experiments at two time points. This is not a problematic situation, but you should understand what happens here:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>meta_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ID=</span><span class="fu">paste0</span>(<span class="st">"ID"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>),</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">age=</span><span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">4</span>, <span class="at">replace=</span><span class="cn">TRUE</span>),</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">group=</span><span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"control"</span>, <span class="st">"treated"</span>), <span class="at">each=</span><span class="dv">2</span>))</span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>meta_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   ID age   group
1 ID1   1 control
2 ID2   3 control
3 ID3   2 treated
4 ID4   1 treated</code></pre>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>crp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ID=</span><span class="fu">rep</span>(<span class="fu">paste0</span>(<span class="st">"ID"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>), <span class="at">each=</span><span class="dv">2</span>),</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">time=</span><span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"day1"</span>, <span class="st">"day2"</span>), <span class="dv">4</span>),</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">CRP=</span><span class="fu">rnorm</span>(<span class="dv">8</span>))</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>crp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   ID time         CRP
1 ID1 day1 -2.04128669
2 ID1 day2  0.73964228
3 ID2 day1  0.02292169
4 ID2 day2 -2.17568505
5 ID3 day1 -0.50429276
6 ID3 day2  0.24247957
7 ID4 day1  1.32033448
8 ID4 day2 -1.06346436</code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>merged_data <span class="ot">&lt;-</span> <span class="fu">merge</span>(meta_data, crp, <span class="at">by=</span><span class="st">"ID"</span>)</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>merged_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   ID age   group time         CRP
1 ID1   1 control day1 -2.04128669
2 ID1   1 control day2  0.73964228
3 ID2   3 control day1  0.02292169
4 ID2   3 control day2 -2.17568505
5 ID3   2 treated day1 -0.50429276
6 ID3   2 treated day2  0.24247957
7 ID4   1 treated day1  1.32033448
8 ID4   1 treated day2 -1.06346436</code></pre>
</div>
</div>
<p>As you can see, each line in the <code>meta_data</code> data frame is duplicated to fill up the information matching the <code>ID</code> in the <code>crp</code> data frame.</p>
<p>However, chances are the situation is more complex. Imagine that apart from the above measurements, you have another set of measurements, say, of the albumin (ALB) levels. You would like to merge the two data frames. Naturally, you expect that you will get a data frame with 8 rows. The following will not work as expected:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>albumin <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ID=</span><span class="fu">rep</span>(<span class="fu">paste0</span>(<span class="st">"ID"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>), <span class="at">each=</span><span class="dv">2</span>),</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">time=</span><span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"day1"</span>, <span class="st">"day2"</span>), <span class="dv">4</span>),</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">ALB=</span><span class="fu">rnorm</span>(<span class="dv">8</span>))</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a><span class="co"># incorrect code!!!</span></span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>crp_alb <span class="ot">&lt;-</span> <span class="fu">merge</span>(crp, albumin, <span class="at">by=</span><span class="st">"ID"</span>)</span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a>crp_alb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    ID time.x         CRP time.y        ALB
1  ID1   day1 -2.04128669   day1 -0.6402011
2  ID1   day1 -2.04128669   day2 -0.9585395
3  ID1   day2  0.73964228   day1 -0.6402011
4  ID1   day2  0.73964228   day2 -0.9585395
5  ID2   day1  0.02292169   day1 -0.3241016
6  ID2   day1  0.02292169   day2  2.5869671
7  ID2   day2 -2.17568505   day1 -0.3241016
8  ID2   day2 -2.17568505   day2  2.5869671
9  ID3   day1 -0.50429276   day1 -1.2320729
10 ID3   day1 -0.50429276   day2 -2.6556808
11 ID3   day2  0.24247957   day1 -1.2320729
12 ID3   day2  0.24247957   day2 -2.6556808
13 ID4   day1  1.32033448   day1 -0.4802608
14 ID4   day1  1.32033448   day2 -1.3147089
15 ID4   day2 -1.06346436   day1 -0.4802608
16 ID4   day2 -1.06346436   day2 -1.3147089</code></pre>
</div>
</div>
<p>Uh-oh, what happened here?</p>
<p>You see, the problem is that the <code>merge()</code> function merges the data frames by the column <code>ID</code>, but it does not take into account the <code>time</code> column. Since it observes that ID’s are duplicated in <em>both</em> data frames, it creates all possible pairs of the rows: CRP from day1 with ALB day2; but also CRP from day1 and ALB from day1, CRP from day2 and ALB from day1 and so on. For each identifier. That is why we are getting <code>4</code> (and not <code>2</code>) rows for each identifier.</p>
<div class="page-columns page-full"><p> This is a common problem, and the solution is to merge by more than one identifier. The important question is: <em>what combination of identifiers <strong>uniquely</strong> identifies a row in the data frame?</em> In our case, it is the combination of <code>ID</code> and <code>time</code>. We can specify that in the <code>by=</code> parameter:</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Merging by more than one ID</span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>crp_alb <span class="ot">&lt;-</span> <span class="fu">merge</span>(crp, albumin, <span class="at">by=</span><span class="fu">c</span>(<span class="st">"ID"</span>, <span class="st">"time"</span>))</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>crp_alb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   ID time         CRP        ALB
1 ID1 day1 -2.04128669 -0.6402011
2 ID1 day2  0.73964228 -0.9585395
3 ID2 day1  0.02292169 -0.3241016
4 ID2 day2 -2.17568505  2.5869671
5 ID3 day1 -0.50429276 -1.2320729
6 ID3 day2  0.24247957 -2.6556808
7 ID4 day1  1.32033448 -0.4802608
8 ID4 day2 -1.06346436 -1.3147089</code></pre>
</div>
</div>
</section>
<section id="bringing-it-all-together" class="level3" data-number="4.5.5">
<h3 data-number="4.5.5" class="anchored" data-anchor-id="bringing-it-all-together"><span class="header-section-number">4.5.5</span> Bringing it all together</h3>
<p>We are now coming to the final exercise in this section. This exercise is based on data from the vaccination study <span class="citation" data-cites="weiner2019characterization">(<a href="references.html#ref-weiner2019characterization" role="doc-biblioref">Weiner et al. 2019</a>)</span>. One file contains meta-data for the samples used in transcriptomic analysis – it only includes the subjects and time points for which the gene expression data is present. The other file is quite large and contains the results of several laboratory tests taken for many patients at many time points.</p>
<p>Note that the data is not real, just based on real data.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-day4-merging" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.9 (Merging large data sets)</strong></span> The files <code>expression_data_vaccination_example.xlsx</code> and <code>labresults_full.csv</code> contain data from the same study. The first contains some meta-data for samples for which gene expression has been measured, and the latter contains the results of laboratory tests for many patients in the trial. The goal here is to connect the RNA expression data with the laboratory values, so that one can analyze the relationship between the expression of certain genes and, say, the observed levels of inflammation.</p>
<ol type="1">
<li>Read CSV file and the first sheet (“targets”) from the XLSX file.</li>
<li>Which columns contain the ID the subjects? Are there any subjects in common? (Hint: use the <code>intersect()</code> function)</li>
<li>What other columns are common between the two data frames? Is the column with the ID of the subject sufficient to identify each row?</li>
<li>We are interested only in the following information: SUBJ, ARM (group), time point, sex, age, test name and the actual measurement. Are the measurements numeric? Remember, you can use expressions like <code>[ , c("ARM", "sex") ]</code> or <code>select(df, ARM, sex)</code> to select the desired columns from a data set.</li>
<li>Use the subjects and / or other information to merge the two data frames however you see fit. Note that there are multiple time points per subject and multiple measurements per subject and time point.</li>
</ol>
<p>See <a href="solutions.html#sec-exercise-merging">Solution</a></p>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="pipes-in-r" class="level2 page-columns page-full" data-number="4.6">
<h2 data-number="4.6" class="anchored" data-anchor-id="pipes-in-r"><span class="header-section-number">4.6</span> Pipes in R</h2>
<section id="too-many-parentheses" class="level3 page-columns page-full" data-number="4.6.1">
<h3 data-number="4.6.1" class="anchored" data-anchor-id="too-many-parentheses"><span class="header-section-number">4.6.1</span> Too many parentheses</h3>
<p>Remember how functions work? Each function takes some arguments and returns a value. And we can use that value to plug in back into another function. This allows us to write terse code, for example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="cn">NA</span>, <span class="dv">3</span>, <span class="dv">10</span>)</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(<span class="fu">is.na</span>(<span class="fu">as.numeric</span>(vec)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
</div>
<p>This is quite easy to understand. Unfortunately, this can quickly become unreadable. Let us consider cleaning up the data from yesterday. We can do it one by one, like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(janitor)</span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>iris_data <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"Datasets/iris.csv"</span>)</span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>iris_data <span class="ot">&lt;-</span> <span class="fu">clean_names</span>(iris_data)</span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>species <span class="ot">&lt;-</span> <span class="fu">tolower</span>(iris_data<span class="sc">$</span>species)</span>
<span id="cb92-6"><a href="#cb92-6" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>petal_length <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(iris_data<span class="sc">$</span>petal_length ,<span class="st">"[^0-9.]"</span>, <span class="st">""</span>) </span>
<span id="cb92-7"><a href="#cb92-7" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>petal_length <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(iris_data<span class="sc">$</span>petal_length)</span>
<span id="cb92-8"><a href="#cb92-8" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>sepal_length <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(iris_data<span class="sc">$</span>sepal_length, <span class="st">","</span>, <span class="st">"."</span>)</span>
<span id="cb92-9"><a href="#cb92-9" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>sepal_length <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(iris_data<span class="sc">$</span>sepal_length ,<span class="st">"&gt; "</span>, <span class="st">""</span>) </span>
<span id="cb92-10"><a href="#cb92-10" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>sepal_length <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(iris_data<span class="sc">$</span>sepal_length)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>That is a lot of code. However, we <em>could</em> do it all on one line:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>iris_data <span class="ot">&lt;-</span> </span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">mutate</span>(<span class="fu">mutate</span>(<span class="fu">clean_names</span>(<span class="fu">read_csv</span>(<span class="st">"Datasets/iris.csv"</span>)),</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">petal_length =</span> <span class="fu">as.numeric</span>(</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_replace_all</span>(petal_length, <span class="st">"[^0-9.]"</span>, <span class="st">""</span>))),</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">sepal_length =</span> <span class="fu">as.numeric</span>(<span class="fu">str_replace_all</span>(</span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_replace_all</span>(sepal_length, <span class="st">","</span>, <span class="st">"."</span>), <span class="st">"&gt; "</span>, <span class="st">""</span>))),</span>
<span id="cb93-7"><a href="#cb93-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">species =</span> <span class="fu">tolower</span>(species))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you spend some time deciphering this, you will see that it does precisely the same operations as the previous code fragment. However, it is completely unreadable and unmaintainable<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>.</p>
<div class="no-row-height column-margin column-container"><div id="fn3"><p><sup>3</sup>&nbsp;I got a headache just trying to figure out the correct number of parentheses in this code.</p></div></div></section>
<section id="pipes" class="level3 page-columns page-full" data-number="4.6.2">
<h3 data-number="4.6.2" class="anchored" data-anchor-id="pipes"><span class="header-section-number">4.6.2</span> Pipes</h3>
<div class="page-columns page-full"><p> Fortunately, there is a <em>dirty</em> trick that results in clean and readable code: the pipe operator <code>|&gt;</code> (pronounced “pipe”; in older code you might see “%&gt;%” instead<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>).</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Pipe operator <code>|&gt;</code></span><div id="fn4"><p><sup>4</sup>&nbsp;The idea of the pipe operator was popularized by the <code>magrittr</code> package, which is a part of Tidyverse, as <code>%&gt;%</code>. R users found it so cool that with R version 4.1.0, the pipe operator <code>|&gt;</code> was included in the base R. There are some differences between <code>|&gt;</code> and <code>%&gt;%</code>, but for our purposes, they are equivalent.</p></div></div></div>
<p>Pipe operator allows for writing very clean, very convenient and very readable code. The basic idea behind the pipe operator is as follows: <code>a |&gt; f(b)</code> is the same as <code>f(a, b)</code>. We can say it <em>pipes</em> the value of <code>a</code> to the function <code>f</code>. And if a is already the result of a function, say <code>g(a)</code>, then <code>g(a) |&gt; f(b)</code> is the same as <code>f(g(a), b)</code>. We build a pipe from the output of <code>g()</code> to the input of <code>f()</code>.</p>
<p>So, for example, the following two lines are equivalent:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="cn">NA</span>, <span class="dv">3</span>, <span class="dv">10</span>)</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as.numeric</span>(vec)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  1 NA  3 10</code></pre>
</div>
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>vec <span class="sc">|&gt;</span> <span class="fu">as.numeric</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  1 NA  3 10</code></pre>
</div>
</div>
<p>Pipe takes whatever is on its left side and inserts it in the parentheses of the function on its right side. Now think for a moment what it does to constructs like <code>f1(f2(f3(f4(f5(x)))))</code>. Now compare this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(<span class="fu">is.na</span>(<span class="fu">as.numeric</span>(vec)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>vec <span class="sc">|&gt;</span> <span class="fu">as.numeric</span>() <span class="sc">|&gt;</span> <span class="fu">is.na</span>() <span class="sc">|&gt;</span> <span class="fu">which</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
</div>
<p>The two lines are equivalent. However, the latter is much more readable: it shows how the value goes into <code>as.numeric()</code>, then the output of <code>as.numeric()</code> goes into <code>is.na()</code>, and the resulting logical vector goes into <code>which()</code>.</p>
<p>See how that works for our iris dataset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>iris_data <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"Datasets/iris.csv"</span>) <span class="sc">|&gt;</span></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>        <span class="fu">clean_names</span>() <span class="sc">|&gt;</span></span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">species =</span> <span class="fu">tolower</span>(species)) <span class="sc">|&gt;</span></span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">petal_length =</span> <span class="fu">str_replace_all</span>(petal_length ,<span class="st">"[^0-9.]"</span>, <span class="st">""</span>)) <span class="sc">|&gt;</span></span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">petal_length =</span> <span class="fu">as.numeric</span>(petal_length)) <span class="sc">|&gt;</span></span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">sepal_length =</span> <span class="fu">str_replace_all</span>(sepal_length, <span class="st">","</span>, <span class="st">"."</span>)) <span class="sc">|&gt;</span></span>
<span id="cb102-7"><a href="#cb102-7" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">sepal_length =</span> <span class="fu">str_replace_all</span>(sepal_length ,<span class="st">"&gt; "</span>, <span class="st">""</span>)) <span class="sc">|&gt;</span></span>
<span id="cb102-8"><a href="#cb102-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">sepal_length =</span> <span class="fu">as.numeric</span>(sepal_length))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It is just as clear and readable as the first example, but with some additional benefits.</p>
<p>First, think what happens if you want to change the name of the variable, say, from <code>iris_data</code> to <code>flowers</code>. In the first example, you would have to change every single occurence of <code>iris_data</code> to <code>flowers</code>. Sure, it would not be all too bad if you can do it automatically with a search-and-replace, but what if the variable name was <code>a</code>?</p>
<p>Second, the pipe facilitates development. You build a pipe line by line, and each time you execute the code, all operations are repeated. This is a good thing – you avoid the situation where you forget to execute one line and the code does not work.</p>
<p>There is an alternative to the above code which uses pipes, but not all on the same line:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>iris_data <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"Datasets/iris.csv"</span>) <span class="sc">|&gt;</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>        <span class="fu">clean_names</span>() <span class="sc">|&gt;</span></span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">species =</span> <span class="fu">tolower</span>(species))</span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>petal_length <span class="ot">&lt;-</span> iris_data<span class="sc">$</span>petal_length <span class="sc">|&gt;</span> </span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">str_replace_all</span>(<span class="st">"[^0-9.]"</span>, <span class="st">""</span>) <span class="sc">|&gt;</span></span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as.numeric</span>()</span>
<span id="cb103-8"><a href="#cb103-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-9"><a href="#cb103-9" aria-hidden="true" tabindex="-1"></a>iris_data<span class="sc">$</span>sepal_length <span class="ot">&lt;-</span> iris_data<span class="sc">$</span>sepal_length <span class="sc">|&gt;</span></span>
<span id="cb103-10"><a href="#cb103-10" aria-hidden="true" tabindex="-1"></a>        <span class="fu">str_replace_all</span>(<span class="st">","</span>, <span class="st">"."</span>) <span class="sc">|&gt;</span></span>
<span id="cb103-11"><a href="#cb103-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">str_replace_all</span>(<span class="st">"&gt; "</span>, <span class="st">""</span>) <span class="sc">|&gt;</span></span>
<span id="cb103-12"><a href="#cb103-12" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as.numeric</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Whether you prefer this one or the previous is a matter of both, taste and context. The second is more appropriate if for each column you have to do a lot of manipulations.</p>
<p>One limitation is that in the form that is used here<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>, the value passed on from function to function is always the first argument. This is why we are doing the substitutions with the Tidyverse function <code>str_replace_all()</code> (which takes the character vector as the first argument) rather than the base R function <code>gsub()</code> (which takes the pattern as the first argument).</p>
<div class="no-row-height column-margin column-container"><div id="fn5"><p><sup>5</sup>&nbsp;This limitation can be circumvented either by using only named arguments, or by using a placeholder, <code>_</code>. See <a href="https://stackoverflow.com/questions/70598384/piping-second-or-higher-argument-in-native-pipe">here</a> for an explanation.</p></div></div><div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-pipes" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 4.10 (Botched meta data)</strong></span> Go back to the code that you have used yesterday for cleaning up the dataset from file <code>meta_data_botched.xlsx</code>. Use pipes to make it more readable.</p>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="review" class="level2" data-number="4.7">
<h2 data-number="4.7" class="anchored" data-anchor-id="review"><span class="header-section-number">4.7</span> Review</h2>
<p>Things that you learned today:</p>
<ul>
<li>Selecting columns in data frames:
<ul>
<li>using square brackets</li>
<li>using <code>select()</code> from tidyverse</li>
<li>tidy evaluation</li>
<li>renaming columns with <code>rename()</code></li>
</ul></li>
<li>Sorting, ordering, filtering
<ul>
<li><code>sort()</code> to sort a vector</li>
<li><code>order()</code> to get the order of a vector</li>
<li><code>order()</code> to sort a data frame</li>
<li><code>arrange()</code> from tidyverse to sort a data frame</li>
<li>filtering with logical vectors</li>
<li>filtering with <code>filter()</code> from tidyverse</li>
<li>handling duplicates with <code>duplicated()</code></li>
<li>combining logical vectors with <code>&amp;</code></li>
<li>using <code>%in%</code> operator</li>
</ul></li>
<li>Merging data frames
<ul>
<li><code>rbind()</code> and <code>cbind()</code></li>
<li><code>merge()</code> inner, left, right and full joins</li>
<li>merging by more than one column</li>
</ul></li>
<li>Pipes in R
<ul>
<li>using <code>|&gt;</code> operator</li>
<li>building pipelines</li>
</ul></li>
<li>Other
<ul>
<li>using <code>rnorm()</code> to generate random numbers</li>
<li>using <code>round()</code> to round numbers</li>
<li><code>sample()</code> for generating random samples</li>
</ul></li>
</ul>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-weiner2019characterization" class="csl-entry" role="listitem">
Weiner, January, David JM Lewis, Jeroen Maertzdorf, Hans-Joachim Mollenkopf, Caroline Bodinham, Kat Pizzoferro, Catherine Linley, et al. 2019. <span>“Characterization of Potential Biomarkers of Reactogenicity of Licensed Antiviral Vaccines: Randomized Controlled Clinical Trials Conducted by the BIOVACSAFE Consortium.”</span> <em>Scientific Reports</em> 9 (1): 20362. <a href="https://www.nature.com/articles/s41598-019-56994-8">https://www.nature.com/articles/s41598-019-56994-8</a>.
</div>
</div>
</section>


</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./day3-reading-your-data.html" class="pagination-link" aria-label="Reading and Writing Files">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Reading and Writing Files</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./day5-visualization-and-statistics.html" class="pagination-link" aria-label="R markdown, basic statistics and visualizations">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">R markdown, basic statistics and visualizations</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>