<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.54">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>5&nbsp; R markdown, basic statistics and visualizations – Five days of R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./appendix-wide-long.html" rel="next">
<link href="./day4-manipulating-your-data.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://hypothes.is/embed.js"></script>
<script>
  window.document.addEventListener("DOMContentLoaded", function (_event) {
    document.body.classList.add('hypothesis-enabled');
  });
</script>
<script src="site_libs/quarto-diagram/mermaid.min.js"></script>
<script src="site_libs/quarto-diagram/mermaid-init.js"></script>
<link href="site_libs/quarto-diagram/mermaid.css" rel="stylesheet">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="custom.css">
</head>

<body class="nav-sidebar floating nav-fixed slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Five days of R</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="./index.html" aria-current="page"> 
<span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day1-intro-to-r.html"> 
<span class="menu-text">One</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day2-complex-structures.html"> 
<span class="menu-text">Two</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day3-reading-your-data.html"> 
<span class="menu-text">Three</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day4-manipulating-your-data.html"> 
<span class="menu-text">Four</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./day5-visualization-and-statistics.html"> 
<span class="menu-text">Five</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./solutions.html"> 
<span class="menu-text">Solutions</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/bihealth/RCrashcourse-book/" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./day5-visualization-and-statistics.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">R markdown, basic statistics and visualizations</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day1-intro-to-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">First day of R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day2-complex-structures.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Complex data structures</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day3-reading-your-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Reading and Writing Files</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day4-manipulating-your-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Manipulating data frames</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day5-visualization-and-statistics.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">R markdown, basic statistics and visualizations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix-wide-long.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendix: Wide vs long data</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./appendix-more-stats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendix: More statistics and visualizations</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Solutions to Exercises</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#aims-for-today" id="toc-aims-for-today" class="nav-link active" data-scroll-target="#aims-for-today"><span class="header-section-number">5.1</span> Aims for today</a></li>
  <li><a href="#r-markdown" id="toc-r-markdown" class="nav-link" data-scroll-target="#r-markdown"><span class="header-section-number">5.2</span> R markdown</a>
  <ul class="collapse">
  <li><a href="#what-is-r-markdown" id="toc-what-is-r-markdown" class="nav-link" data-scroll-target="#what-is-r-markdown"><span class="header-section-number">5.2.1</span> What is R markdown?</a></li>
  <li><a href="#markdown-basics" id="toc-markdown-basics" class="nav-link" data-scroll-target="#markdown-basics"><span class="header-section-number">5.2.2</span> Markdown basics</a></li>
  <li><a href="#r-markdown-header" id="toc-r-markdown-header" class="nav-link" data-scroll-target="#r-markdown-header"><span class="header-section-number">5.2.3</span> R markdown header</a></li>
  <li><a href="#r-code-chunks" id="toc-r-code-chunks" class="nav-link" data-scroll-target="#r-code-chunks"><span class="header-section-number">5.2.4</span> R code chunks</a></li>
  </ul></li>
  <li><a href="#visualizations-with-r" id="toc-visualizations-with-r" class="nav-link" data-scroll-target="#visualizations-with-r"><span class="header-section-number">5.3</span> Visualizations with R</a>
  <ul class="collapse">
  <li><a href="#basic-principles-of-visualization" id="toc-basic-principles-of-visualization" class="nav-link" data-scroll-target="#basic-principles-of-visualization"><span class="header-section-number">5.3.1</span> Basic principles of visualization</a></li>
  <li><a href="#base-r-vs-ggplot2" id="toc-base-r-vs-ggplot2" class="nav-link" data-scroll-target="#base-r-vs-ggplot2"><span class="header-section-number">5.3.2</span> Base R vs <code>ggplot2</code></a></li>
  <li><a href="#esthetics-and-information-channels" id="toc-esthetics-and-information-channels" class="nav-link" data-scroll-target="#esthetics-and-information-channels"><span class="header-section-number">5.3.3</span> Esthetics and information channels</a></li>
  <li><a href="#boxplots-and-violin-plots" id="toc-boxplots-and-violin-plots" class="nav-link" data-scroll-target="#boxplots-and-violin-plots"><span class="header-section-number">5.3.4</span> Boxplots and violin plots</a></li>
  <li><a href="#heatmaps" id="toc-heatmaps" class="nav-link" data-scroll-target="#heatmaps"><span class="header-section-number">5.3.5</span> Heatmaps</a></li>
  <li><a href="#sec-output-formats" id="toc-sec-output-formats" class="nav-link" data-scroll-target="#sec-output-formats"><span class="header-section-number">5.3.6</span> Output formats</a></li>
  </ul></li>
  <li><a href="#basic-statistics-with-r" id="toc-basic-statistics-with-r" class="nav-link" data-scroll-target="#basic-statistics-with-r"><span class="header-section-number">5.4</span> Basic statistics with R</a>
  <ul class="collapse">
  <li><a href="#statistics-with-r" id="toc-statistics-with-r" class="nav-link" data-scroll-target="#statistics-with-r"><span class="header-section-number">5.4.1</span> Statistics with R</a></li>
  <li><a href="#descriptive-statistics" id="toc-descriptive-statistics" class="nav-link" data-scroll-target="#descriptive-statistics"><span class="header-section-number">5.4.2</span> Descriptive statistics</a></li>
  <li><a href="#simple-tests" id="toc-simple-tests" class="nav-link" data-scroll-target="#simple-tests"><span class="header-section-number">5.4.3</span> Simple tests</a></li>
  </ul></li>
  <li><a href="#pca-and-scatter-plots" id="toc-pca-and-scatter-plots" class="nav-link" data-scroll-target="#pca-and-scatter-plots"><span class="header-section-number">5.5</span> PCA and scatter plots</a>
  <ul class="collapse">
  <li><a href="#principal-component-analysis" id="toc-principal-component-analysis" class="nav-link" data-scroll-target="#principal-component-analysis"><span class="header-section-number">5.5.1</span> Principal component analysis</a></li>
  <li><a href="#the-data" id="toc-the-data" class="nav-link" data-scroll-target="#the-data"><span class="header-section-number">5.5.2</span> The data</a></li>
  <li><a href="#running-the-pca" id="toc-running-the-pca" class="nav-link" data-scroll-target="#running-the-pca"><span class="header-section-number">5.5.3</span> Running the PCA</a></li>
  <li><a href="#interpreting-the-pca" id="toc-interpreting-the-pca" class="nav-link" data-scroll-target="#interpreting-the-pca"><span class="header-section-number">5.5.4</span> Interpreting the PCA</a></li>
  </ul></li>
  <li><a href="#afterword" id="toc-afterword" class="nav-link" data-scroll-target="#afterword"><span class="header-section-number">5.6</span> Afterword</a>
  <ul class="collapse">
  <li><a href="#where-to-get-r-packages" id="toc-where-to-get-r-packages" class="nav-link" data-scroll-target="#where-to-get-r-packages"><span class="header-section-number">5.6.1</span> Where to get R packages</a></li>
  <li><a href="#where-to-go-from-here" id="toc-where-to-go-from-here" class="nav-link" data-scroll-target="#where-to-go-from-here"><span class="header-section-number">5.6.2</span> Where to go from here</a></li>
  <li><a href="#famous-last-words" id="toc-famous-last-words" class="nav-link" data-scroll-target="#famous-last-words"><span class="header-section-number">5.6.3</span> Famous last words</a></li>
  </ul></li>
  <li><a href="#review" id="toc-review" class="nav-link" data-scroll-target="#review"><span class="header-section-number">5.7</span> Review</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">R markdown, basic statistics and visualizations</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="aims-for-today" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="aims-for-today"><span class="header-section-number">5.1</span> Aims for today</h2>
<ul>
<li>working R markdown / Quarto</li>
<li>basic visualisations with R and <code>ggplot2</code></li>
<li>simple statistics with R</li>
<li>final example: principal component analysis</li>
</ul>
</section>
<section id="r-markdown" class="level2 page-columns page-full" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="r-markdown"><span class="header-section-number">5.2</span> R markdown</h2>
<section id="what-is-r-markdown" class="level3 page-columns page-full" data-number="5.2.1">
<h3 data-number="5.2.1" class="anchored" data-anchor-id="what-is-r-markdown"><span class="header-section-number">5.2.1</span> What is R markdown?</h3>
<p>R markdown is a remarkably simple to use and yet powerful tool. The basic idea is as follows: you write a document in a simple text format – for example, a manuscript, a report, or even a thesis. In that document, you include R “code chunks”. These code chunks are executed by R, and the output is included in the document – including figures, numbers or tables. And then, from this single simple text document you can create a PDF, a Word document, an HTML page (or a whole website!), a book, a slide presentation and much, much more.</p>
<div class="cell" data-layout-align="default">
<div class="cell-output-display">
<div>
<p></p><figure class="figure"><p></p>
<div>
<pre class="mermaid mermaid-js">flowchart LR
    A(R markdown / Quarto\ndocument\nwith R code) --&gt;|knitr| B(Markdown\nwith R code output)
    B --&gt; C[LaTeX]
    C --&gt; CC[PDF]
    B --&gt; D[Word]
    B --&gt; E[HTML]
    B --&gt; F[Presentation]
    B --&gt; G[Book]
</pre>
</div>
<p></p></figure><p></p>
</div>
</div>
</div>
<p>In fact, this book you are reading now has been entirely written in R markdown<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. You can go to the <a href="https://github.com/bihealth/RCrashcourse-book/">github repository</a> for this book and see the R markdown/Quarto files that generated it – for example, <a href="https://github.com/bihealth/RCrashcourse-book/blob/main/day5-visualization-and-statistics.qmd">here</a> is the page you are reading right now. It is also possible to create a scientific paper completely in R markdown (here is an example of such a paper: <span class="citation" data-cites="weiner2022venn">Weiner 3rd, Obermayer, and Beule (<a href="references.html#ref-weiner2022venn" role="doc-biblioref">2022</a>)</span>).</p>
<div class="no-row-height column-margin column-container"><div id="fn1"><p><sup>1</sup>&nbsp;Actually, in its successor called “Quarto”.</p></div></div><p>Why is this so great?</p>
<ul>
<li><strong>Reproducibility:</strong> It allows you to keep the methods, analysis itself and the analysis results in one place.</li>
<li><strong>Avoiding errors:</strong> Whenever you re-run your analysis, R automatically updates all your figures, tables, results, etc. in your document. This prevents you from forgetting to update a figure or a number in your manuscript.</li>
<li><strong>Flexibility:</strong> You can easily change the output format of your document; you can update your figures or your tables easily without having to resort to manual editing.</li>
<li><strong>Bibliography</strong>. As you have seen in this book, it is not hard to include bibliography in your document, in any style you desire. You can use a free package manager such as Zotero or Mendeley to manage your bibliography, and produce the bibliography in the format that R markdown uses.</li>
<li><strong>Focus on content</strong>. The minimal formatting required in R markdown allows you to focus on the content of your document, not on the formatting.</li>
</ul>
<p>What are the disadvantages?</p>
<ul>
<li><strong>Steeper learning curve.</strong> You have to learn all the stuff first. Luckily, when it comes to R markdown, it is quite easy and there are plenty of resources, and after today you will know all the most important stuff.</li>
<li><strong>No fine control over the layout.</strong> While you can easily use simple formatting commands, you will have to resort to more complex tools to control things like font size of the chapter headers<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>. Many people consider it to be a good thing – this is the boring part that the computers should take care of, you shouldn’t worry about formatting but about your text, but sometimes it <em>is</em> annoying. However, you can always generate a Word file with R markdown that uses a Word template, and then fine tune it in Word.</li>
<li><strong>Collaboration.</strong> Collaborating with markdown is not as easy as with a Word document, because most people can’t use markdown. Often that means that you need to communicate with your co-authors using Word, and then patiently type in their changes into your markdown document.</li>
</ul>
<div class="no-row-height column-margin column-container"><div id="fn2"><p><sup>2</sup>&nbsp;Actually, you can control the format for PDF and HTML output very precisely, but then you have to learn LaTeX and CSS, respectively. For Word, the only option is to use a Word file with pre-defined styles as a template.</p></div></div><div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
R markdown vs Quarto
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>R markdown is older and more widely supported</li>
<li>Quarto is newer, slightly more complex, with additional features and generally better looking</li>
</ul>
<p>Documents created in Quarto can largely be processed with R markdown and vice versa, only some visual bells &amp; whistles might be lacking.</p>
<p>Both Quarto and R markdown are available if you have installed RStudio. Standalone R installations without RStudio may require additional packages (e.g.&nbsp;<code>rmarkdown</code> for R markdown) or programs (<code>quarto</code> for Quarto).</p>
</div>
</div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-basic-markdown" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 5.1 (Create an R markdown document)</strong></span> Go to the “File” menu in Rstudio and choose “New File” -&gt; “R markdown”. Enter a title and click on “OK”. Save the file in your project directory and take a look at it. Rstudio has created a simple R markdown file for you so that you might get an idea of how it works.</p>
<p>Click on the “knit” icon in the toolbar to render the document. What do you see? How does it relate to the content of the document? Try changing a few words in the document and click on “knit” again. What happens?</p>
</div>
</div>
</div>
</div>
</section>
<section id="markdown-basics" class="level3 page-columns page-full" data-number="5.2.2">
<h3 data-number="5.2.2" class="anchored" data-anchor-id="markdown-basics"><span class="header-section-number">5.2.2</span> Markdown basics</h3>
<p>The “R” in R markdown stands for “R”, the programming language, combined with markdown. But what <em>is</em> markdown?</p>
<div class="page-columns page-full"><p>Markdown is a very lightweight formatting system, much like what many of us are using in emails or messengers, stressing words by surrounding them with stars etc., but with a few extra features. The idea is that you can write the text in a very simple way and it remains readable <em>even with the formatting marks</em> (<a href="https://github.com/bihealth/RCrashcourse-book/blob/main/day5-visualization-and-statistics.qmd">take a look!</a>). </p><div class="no-row-height column-margin column-container"><span class="margin-aside">Basic markdown formatting</span></div></div>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Code</th>
<th>Output</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>**bold**</code></td>
<td><strong>bold</strong></td>
</tr>
<tr class="even">
<td><code>*italic*</code></td>
<td><em>italic</em></td>
</tr>
<tr class="odd">
<td><code>3^2^</code></td>
<td>3<sup>2</sup></td>
</tr>
<tr class="even">
<td><code>log~2~</code></td>
<td>log<sub>2</sub></td>
</tr>
<tr class="odd">
<td><code>`code`</code></td>
<td><code>code</code></td>
</tr>
<tr class="even">
<td><code>[link to website](https://cubi.bihealth.org)</code></td>
<td><a href="https://cubi.bihealth.org">link to website</a></td>
</tr>
</tbody>
</table>
<hr>
<div class="page-columns page-full"><p>Here is another feature: lists. </p><div class="no-row-height column-margin column-container"><span class="margin-aside">Lists in markdown</span></div></div>
<div class="columns">
<div class="column" style="width:50%;">
<p>Code:</p>
<pre><code>- item 1
- item 2
  - subitem 1
  - subitem 2
- item 3</code></pre>
</div><div class="column" style="width:50%;">
<p>Result:</p>
<ul>
<li>item 1</li>
<li>item 2
<ul>
<li>subitem 1</li>
<li>subitem 2</li>
</ul></li>
<li>item 3</li>
</ul>
</div>
</div>
<hr>
<p>There is much more to it (look up for example <a href="https://rmarkdown.rstudio.com/lesson-1.html">markdown guide</a> from Rstudio or the <a href="https://quarto.org/docs/authoring/markdown-basics.html">Quarto markdown documentation</a>), but you don’t need it right now. Just keep in mind that you can always take a look at the markdown source of this document to see how things can be done.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Mathematical formulas
</div>
</div>
<div class="callout-body-container callout-body">
<p>Using a special syntax, it is possible to include virtually any mathematical formula in R, both inline variant (like <span class="math inline">\(\sigma=\sqrt{\frac{(x_i-\bar{x})^2}{n}}\)</span>) or as a stand-alone block:</p>
<p><span class="math display">\[\sigma=\sqrt{\frac{(x_i-\bar{x})^2}{n}}\]</span></p>
<p>When you convert the R markdown document to Word, you will even be able to edit the formulas natively in Word.</p>
<p>There are several formulas in this book, if you are interested, look up the book quarto sources on <a href="https://github.com/bihealth/RcrashCourse-book/">github</a> or check <a href="https://rpruim.github.io/s341/S19/from-class/MathinRmd.html">this guide</a>.</p>
</div>
</div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-formatting-markdown" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 5.2 (Formatting markdown)</strong></span> &nbsp;</p>
<ul>
<li>Try some formatting commands in markdown. For example, try to make a word <strong>bold</strong> or <em>italic</em>. Try to create a list. Insert a link to a website.</li>
<li>This one is a bit harder, because I did not show it to you (on purpose). First download any kind of image from the internet and save it in your project directory. Then, using whatever means necessary (Google, the markdown guide, the Rstudio help), try to figure out how to insert an image in your document.</li>
</ul>
</div>
</div>
</div>
</div>
</section>
<section id="r-markdown-header" class="level3 page-columns page-full" data-number="5.2.3">
<h3 data-number="5.2.3" class="anchored" data-anchor-id="r-markdown-header"><span class="header-section-number">5.2.3</span> R markdown header</h3>
<p>You might have noticed that at the top of the first R markdown file you created there is a block of text that might look something like this:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Untitled"</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "JW"</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "`​r Sys.Date()`"</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="an">output:</span><span class="co"> html_document</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="page-columns page-full"><p>This is a block with meta-information about your document. For example, it specifies the title, author and date. Note that date is updated automatically every time the document is rendered by executing the r command <code>Sys.Date()</code> (you will learn about the inline chunks in a moment). Naturally, you can run the command in your console or script. </p><div class="no-row-height column-margin column-container"><span class="margin-aside">`Sys.Date()´</span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.Date</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2024-10-14"</code></pre>
</div>
</div>
<p>The <code>output:</code> specifies the type of output you would like to have.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-output-options" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 5.3 (Choosing output format)</strong></span> When editing your R markdown document, click on the little ▼ arrow next to the “Knit” icon in RStudio. Choose the PDF format. Observe what happens to the header of your document.</p>
</div>
</div>
</div>
</div>
</section>
<section id="r-code-chunks" class="level3 page-columns page-full" data-number="5.2.4">
<h3 data-number="5.2.4" class="anchored" data-anchor-id="r-code-chunks"><span class="header-section-number">5.2.4</span> R code chunks</h3>
<p>In between the markdown text, you can include R code chunks. These are executed consicutively by R and the output is included in the document. Every single fragment of code in this book is a code chunk.</p>
<div class="page-columns page-full"><p>Each chunk starts with a <code>```{r}</code> and ends with a <code>```</code> (these should  be placed on separate lines). You can also add the chunks by clicking on the little green plus sign in the Rstudio toolbar and choosing “R chunk”. Here is an example of a code chunk:</p><div class="no-row-height column-margin column-container"><span class="margin-aside">R chunks</span></div></div>
<div class="columns">
<div class="column" style="width:45%;">
<p>R markdown:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="in">x &lt;- 1:10</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="in">print(x)</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="in">plot(x)</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:50%;">
<p>Output:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  1  2  3  4  5  6  7  8  9 10</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-4-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<div class="page-columns page-full"><p>By default, the code itself is also shown in the document. You can configure this by clicking on the little cogwheel ⚙ icon to the right of the chunk start. You will notice that configuring the options means essentially adding stuff like <code>echo=FALSE</code> to the chunk start (<code>echo=FALSE</code>  means that the code itself is not shown in the document).</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>echo=FALSE</code></span></div></div>
<div class="columns">
<div class="column" style="width:45%;">
<p>R markdown:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r echo=FALSE}</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="in">x &lt;- 1:10</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="in">print(x)</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="in">plot(x)</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:50%;">
<p>Output:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code> [1]  1  2  3  4  5  6  7  8  9 10</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-5-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Code chunks run in one environment
</div>
</div>
<div class="callout-body-container callout-body">
<p>Code chunks share their environment. That is, if you define a variable in one chunk, you will be able to use it in the subsequent chunks.</p>
<p>Think of all the chunks as consecutive fragments of your script file!</p>
</div>
</div>
<p>The code chunks can also be <em>inline</em>, that is, you can put directly a code chunk in your sentence. For example, when I write that the <span class="math inline">\(\pi\)</span> constant is equal to 3.1415927, I am using an inline code chunk:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>For example, when I write that the $\pi$ constant is </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>equal to <span class="in">`​r pi`</span>, I am using an inline code chunk:</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The point is not to save typing. The point is to <em>update</em> your document whenever you <em>update</em> your analysis. For example, if you add samples to your data and re-run the analysis, the number of samples that you have written in your method section must reflect the change. If you use an inline code chunk, it will do so automatically, without you needing to painstakingly update each number, each figure, each result in your manuscript.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-basic-markdown" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 5.4 (Put your code in your markdown)</strong></span> &nbsp;</p>
<ul>
<li>Go back to Day 1, the water lillies example in <a href="day1-intro-to-r.html#sec-lilies" class="quarto-xref"><span>Section 1.6.1</span></a>. Copy the code to the R markdown, including the plot and <a href="day1-intro-to-r.html#exr-lilies" class="quarto-xref">Exercise&nbsp;<span>1.11</span></a>.</li>
<li>Write a brief summary including inline code chunks that show the number of days of the simulation used.</li>
</ul>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="visualizations-with-r" class="level2 page-columns page-full" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="visualizations-with-r"><span class="header-section-number">5.3</span> Visualizations with R</h2>
<section id="basic-principles-of-visualization" class="level3 page-columns page-full" data-number="5.3.1">
<h3 data-number="5.3.1" class="anchored" data-anchor-id="basic-principles-of-visualization"><span class="header-section-number">5.3.1</span> Basic principles of visualization</h3>
<p>Before we launch into creating our own plots, let me give you a few tips about visualizations in general.</p>
<p>Each plot is a way of communicating with your reader. You want to tell them something with that plot. Therefore, before you code anything, <em>think</em> what you want to tell the reader with the plot. What is the message, what is the story? How to best get your point across? What <em>is</em> that point?</p>
<p>This is really helpful, because we (and I do mean <em>we all</em>) have the tendency to want to document, to “show our data”: show everything that we got, all the results, all the measurements, all the experiments. Very often this results in cluttered, unreadable plots that do not really communicate anything apart from your prowess in R. While this may (or may not) be needed for things like supplementary data, usually it is not very effective.</p>
<p>The next step is to think <em>how</em> to communicate the message. Again, we all have a knee-jerk reflex to use the plots that we know how to do. This is why I would like to urge you to first take a piece of paper and a pencil and sketch a plot that you think would best communicate your message. Don’t worry about the details or how to implement it, just think how you would like it to look.</p>
<p>In addition, you can always look at how other people have visualized similar data and get inspired. Not to copy, but to see which way of visualizing is convincing <em>for you</em>, which plots are clear and fantastically communicative, and which are hard to read and to understand.</p>
<p>Finally, when creating the plot, think about accessibility. Thus, ask yourself. Are your fonts and symbols large enough to be read by a person with normal vision even if the plot is scaled down to fit a multi-panel illustration? Are the colors color-blind friendly<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>?</p>
<div class="no-row-height column-margin column-container"><div id="fn3"><p><sup>3</sup>&nbsp;Around 5%-8% of all males and 0.5%-1% of all femals are color-blind. When I was a young post-doc I created an interface to single-nucleotide polymorphism data in a sequencing project. I showed synonymous mutations with green arrows and non-synonymous mutations with red arrows. Very proud, I presented it to the twenty or so members of the consortium. One of them stood up and said: “I am color-blind. I can’t see shit on your plot”.</p></div></div><div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Visualization principles
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><strong>Message</strong>: think what the message of your plot is</li>
<li><strong>Design</strong>: make a sketch on how to best communicate the message</li>
<li><strong>Inspiration</strong>: look at other people’s plots for inspiration</li>
<li><strong>Simplicity</strong>: less is more</li>
<li><strong>Accessibility</strong>: make your fonts large and colors color-blind friendly</li>
</ul>
</div>
</div>
</section>
<section id="base-r-vs-ggplot2" class="level3 page-columns page-full" data-number="5.3.2">
<h3 data-number="5.3.2" class="anchored" data-anchor-id="base-r-vs-ggplot2"><span class="header-section-number">5.3.2</span> Base R vs <code>ggplot2</code></h3>
<p>We already had simple examples of plots in R (for example on day one, in the Water Lily example - <a href="day1-intro-to-r.html#sec-lilies" class="quarto-xref"><span>Section 1.6.1</span></a>). We even had a look at the <code>ggplot2</code> package on Day 3 (<a href="day3-reading-your-data.html#sec-quick-look" class="quarto-xref"><span>Section 3.5.5</span></a>). Why are there so many plotting systems in R?</p>
<p>Actully, the situation is even more complex than you might imagine. Partly that is because there are so many different types of plots that it is hard to make a single system that would be good for all of them. Partly because when R was first created, many modern graphical formats and features did not exist yet. But the main problem, as usual, is that anybody can write their own. And so they did.</p>
<p>Fortunately for us, the two graphic systems – base R and <code>ggplot2</code> – are sufficient even for the most sophisticated plots. It is useful to know them both, however. Base R is simple and allows very quick plotting of simple tasks. Moreover, many base R statistical functions have built-in plotting capabilities (e.g.&nbsp;you can simply call <code>plot(model)</code> for a linear regression model to get all relevant plots). <code>ggplot2</code>, on the other hand, is more complex, but it is also working on a much higher level and takes care of many things automatically.</p>
<p>Here is a table comparing the basic features of the two systems.</p>
<table class="caption-top table">
<colgroup>
<col style="width: 50%">
<col style="width: 50%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Base R</th>
<th style="text-align: left;">ggplot2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Simple to use for simple task, gets very complex for complex tasks</td>
<td style="text-align: left;">More complex, but more powerful and makes complex tasks easier</td>
</tr>
<tr class="even">
<td style="text-align: left;">Many packages</td>
<td style="text-align: left;">Many packages</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Plots need many lines to be customized</td>
<td style="text-align: left;">Plots are customized in one line</td>
</tr>
<tr class="even">
<td style="text-align: left;">Low-level, with absolute feature control</td>
<td style="text-align: left;">High-level, with automatic feature control</td>
</tr>
</tbody>
</table>
<p>Most of what can be done with <code>ggplot2</code> can be done with base R, but it often requires many more lines of code. On the other hand, it is easier to develop a <em>de novo</em> approach with base R, because programming new features in ggplot is not for the faint of heart.</p>
<p>Let us plot a simple scatter plot with both systems and show how customization will look like in both. We will use the <code>iris</code> data set<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>.</p>
<div class="no-row-height column-margin column-container"><div id="fn4"><p><sup>4</sup>&nbsp;Not the doctored version we have been loading in the previous days, but the built-in original data set. If you have created an <code>iris</code> variable in your environment, type <code>rm(iris)</code> to remove it.</p></div></div><p><strong>Base R:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>colors_map <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">setosa=</span><span class="st">"red"</span>, <span class="at">versicolor=</span><span class="st">"green"</span>, <span class="at">virginica=</span><span class="st">"blue"</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> colors_map[iris<span class="sc">$</span>Species]</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Sepal.Length, iris<span class="sc">$</span>Sepal.Width, </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">col=</span>colors, <span class="at">pch=</span><span class="dv">19</span>, <span class="at">xlab=</span><span class="st">"Sepal length"</span>, </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">cex=</span><span class="fl">1.5</span>,</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">"Sepal width"</span>, <span class="at">main=</span><span class="st">"The iris data set"</span>)</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="sc">-</span><span class="fl">2.8</span>, <span class="fl">1.1</span>, <span class="at">col=</span><span class="st">"red"</span>)</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend=</span><span class="fu">unique</span>(iris<span class="sc">$</span>Species), </span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">col=</span>colors_map[<span class="fu">unique</span>(iris<span class="sc">$</span>Species)], <span class="at">pch=</span><span class="dv">19</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-7-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Note that in the code above the <code>col=</code> parameter simply takes a vector describing colors. Color management must be done by the user – you cannot simply say “choose the viridis palette” (or similar). You also need to remember some weird things, like that the <code>pch=19</code> means “filled circles” (in contrast, <code>pch=1</code> means “empty circles”, and if you need triangles, you have to use <code>pch=2</code>)<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>. And <code>cex</code> is the size of the points.</p>
<div class="no-row-height column-margin column-container"><div id="fn5"><p><sup>5</sup>&nbsp;For many years I had a piece of paper hanging over my desk with all the <code>pch</code> values written on it and symbols scrawled with pencil next to them.</p></div></div><p>Legend drawing is a completely separate function that slaps the legend on top of the existing plot, whatever it is. The <code>abline()</code> function draws a line on the plot, also not really caring what was drawn before.</p>
<p>Ggplot2:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(iris, <span class="fu">aes</span>(<span class="at">x =</span> Sepal.Length, <span class="at">y =</span> Sepal.Width, <span class="at">color =</span> Species)) <span class="sc">+</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="sc">-</span><span class="fl">2.8</span>, <span class="at">slope =</span> <span class="fl">1.1</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Sepal length"</span>, <span class="at">y =</span> <span class="st">"Sepal width"</span>, </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">"The iris data set"</span>) <span class="sc">+</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-8-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="page-columns page-full"><p>In this example you see why <code>ggplot2</code> is so popular. The code is much cleaner and does not require you calling separate functions for each task. The parameters have explicit names, easy to remember. Legend and abline are directly connected to the plot, and the color scale is chosen with a single command, <code>scale_color_viridis_d()</code>. There is no need to separately change colors for the legend and the plot. The <code>theme_minimal()</code> function changes a whole range of parameters such as the font face used or the background color to achieve a “minimal look”. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>scale_color_viridis_d()</code>, <code>theme_minimal()</code></span></div></div>
<div class="page-columns page-full"><p>Noteworthy is also the <code>labs()</code> function, which allows you to change the labels of the axes and the title of the plot in one go (you can also use individual functions like <code>xlab()</code>, <code>ylab()</code> and <code>ggtitle()</code>). </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>labs()</code>, <code>xlab()</code>, <code>ylab()</code>, <code>ggtitle()</code></span></div></div>
<p>However, some operations can be challenging. If you want your legend placed directly on top of the drawing in the right upper corner, you have to specify the coordinates manually. And if you want triangles… well, you still have to use the numbers, for example using <code>geom_point(shape=2)</code> for triangles<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a>.</p>
<div class="no-row-height column-margin column-container"><div id="fn6"><p><sup>6</sup>&nbsp;And here the punchline: I still have this piece of paper with symbols scrawled on it.</p></div><div id="fn7"><p><sup>7</sup>&nbsp;You basically have to calculate the LOESS lines manually and then manually plot them with <code>lines()</code> and / or <code>polygon()</code>.</p></div></div><p>Then again, some complex operations are easy. For example, to add per-group LOESS lines to the plot with confidence intervals, you would simply add <code>geom_smooth(method="loess")</code>. In base R, the same operation requires about a dozen lines of code<a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a>.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-basic-statistics" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 5.5 (Ggplot2)</strong></span> Take the code for the <code>ggplot2</code> plot above and make the following modifications:</p>
<ul>
<li>Try adding a <code>geom_smooth()</code> layer</li>
<li>Remove <code>scale_color_viridis_d()</code> layer. What happens?</li>
<li>Add the following layer: <code>scale_color_manual(values=colors_map)</code>. What happens?</li>
<li>Change the color of all points to red (use the <code>color</code> parameter in <code>geom_point()</code>)</li>
<li>If you wanted to show different shapes for different species rather than different colors, where would you change the plot? Hint: where is the species mentioned in the code?</li>
</ul>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(iris, <span class="fu">aes</span>(<span class="at">x =</span> Sepal.Length, <span class="at">y =</span> Sepal.Width, <span class="at">color =</span> Species)) <span class="sc">+</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span>colors_map) <span class="sc">+</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># LOESS smoothing, separate for each Species</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>() <span class="sc">+</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="sc">-</span><span class="fl">2.8</span>, <span class="at">slope =</span> <span class="fl">1.1</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Sepal length"</span>, <span class="at">y =</span> <span class="st">"Sepal width"</span>, </span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">"The iris data set"</span>) <span class="sc">+</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using method = 'loess' and formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-9-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="esthetics-and-information-channels" class="level3 page-columns page-full" data-number="5.3.3">
<h3 data-number="5.3.3" class="anchored" data-anchor-id="esthetics-and-information-channels"><span class="header-section-number">5.3.3</span> Esthetics and information channels</h3>
<p>The last point in the exercise was sneaky. The answer is: you would need to change the <code>aes()</code> function call. Rather than use <code>color=Species</code>, you would have to use <code>shape=Species</code>.</p>
<p>The <code>aes()</code> function – and the concept of plot esthetics – is at the core of <code>ggplot2</code>. The idea is that a plot has different information channels, such as x position, y position, color, shape, size, etc. Each of these can be mapped or linked to a column in the input data frame. This is the job of <code>aes()</code>. The <code>geom</code> functions such as <code>geom_point()</code> then use these mapping to actually draw the plot.</p>
<pre><code>                    +------------+
Petal Width ------&gt; |            | -------&gt; x position
Petal Length -----&gt; |  aes()     | -------&gt; y position
Species ----------&gt; |            | -------&gt; color
                    +------------+</code></pre>
<p>This is why the layers (<code>geom_point()</code>, <code>scale_color_viridis_d()</code>, etc.) are being <em>added</em> to the plot using the <code>+</code> operator. Each layer adds another piece of information on how to display the plot. Then, the plot is displayed all in one go.</p>
<p>The assumption is also that this information is mapped automatically. You do not need (or should not need) to specify which precisely color which groups get, or which symbols are used for plotting. Rather, you chose a particular information channel (“use different shapes depending on species”) and <code>ggplot2</code> takes care of the rest.</p>
<p>Of course, it is still possible to manually specify the colors, shapes etc. For example, and as you have seen in the previous exercise, you can use the same manual mapping of colors as above by using <code>scale_color_manual(values=colors_map)</code><a href="#fn8" class="footnote-ref" id="fnref8" role="doc-noteref"><sup>8</sup></a>.</p>
<div class="no-row-height column-margin column-container"><div id="fn8"><p><sup>8</sup>&nbsp;I do not really show here how to use the various <code>scale_color_....</code> and <code>scale_fill_...</code> functions. There are so many use cases, and so many options, that it is hard to give a general overview. Use the google, Luke, or take a look at the <a href="https://r-graphics.org/">R Graphics Cookbook</a>.</p></div></div></section>
<section id="boxplots-and-violin-plots" class="level3 page-columns page-full" data-number="5.3.4">
<h3 data-number="5.3.4" class="anchored" data-anchor-id="boxplots-and-violin-plots"><span class="header-section-number">5.3.4</span> Boxplots and violin plots</h3>
<div class="page-columns page-full"><p>For continuous data, violin plots combined with boxplots or boxplots alone are the method of choice. In the following, we will use the sepal lengths of the iris dataset<a href="#fn9" class="footnote-ref" id="fnref9" role="doc-noteref"><sup>9</sup></a>. First, a boxplot. We have created a boxplot a while ago on day 2 using the R base function <code>boxplot()</code> (<a href="day2-complex-structures.html#sec-lists-as-return-values" class="quarto-xref"><span>Section 2.4.3</span></a>). Now we will do the same with <code>ggplot2</code>. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>geom_boxplot()</code></span></div></div>
<div class="no-row-height column-margin column-container"><div id="fn9"><p><sup>9</sup>&nbsp;Not the doctored version we have been loading in the previous days, but the built-in original data set. If you have created an <code>iris</code> variable in your environment, type <code>rm(iris)</code> to remove it.</p></div></div><div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(iris, <span class="fu">aes</span>(<span class="at">x=</span>Species, <span class="at">y=</span>Sepal.Length)) <span class="sc">+</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-10-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The thick line in the middle of the box is the median, the box itself goes from the lower quartile to the upper quartile (so its vertical size is the interquartile range), and the whiskers show the range of the data (excepting the outliers). The points shown are outliers. All in all, it is a non-parametric representation – that is, we do not assume that the data is normally distributed and we can use it for any kind of continuous data.</p>
<div class="page-columns page-full"><p>However, boxplots are still not perfect, as they do not show the actual distribution of the data. Here is a better method – we create a so-called violin plot which extrapolates a distribution from the data points.  In addition, we overlay that with a boxplot to show the quartiles and outliers.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>geom_violin()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(iris, <span class="fu">aes</span>(<span class="at">x=</span>Species, <span class="at">y=</span>Sepal.Length)) <span class="sc">+</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>() <span class="sc">+</span> </span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width=</span><span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-11-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Finally, in cases where we do not have too many data points we might want to show them directly on the plot (we can combine it with a boxplot or a violin plot). Without loading any additional packages, you can use <code>geom_jitter()</code> to show the points:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(iris, <span class="fu">aes</span>(<span class="at">x=</span>Species, <span class="at">y=</span>Sepal.Length)) <span class="sc">+</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.shape=</span><span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">alpha=</span>.<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-12-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="page-columns page-full"><p>However, this is not the optimal way to show the data points, because it is hard to see the actual distribution. A particularly fine way of showing the actual points is implemented by the <code>ggbeeswarm</code> package and its <code>geom_beeswarm()</code> function. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>geom_beeswarm()</code>, <code>ggbeeswarm</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggbeeswarm)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(iris, <span class="fu">aes</span>(<span class="at">x=</span>Species, <span class="at">y=</span>Sepal.Length)) <span class="sc">+</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.shape=</span><span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>(<span class="at">alpha=</span>.<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-13-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>OK, but what if we want to show how the different variables (Sepal.Length, Sepal.Width, Petal.Length, Petal.Width) differ between the species? One possibility would be to produce four figures like the one above, one after another. But maybe we would like to have all box plots on one figure? Yes, that is totally doable!</p>
<div class="page-columns page-full"><p>First, however, we need to prepare a data frame which contains the values of all the variables in one column. We could use the <code>pivot_longer()</code> function from the tidyverse here, but I haven’t shown you this one yet, so if you would like to learn more about it, read the <a href="appendix-wide-long.html">Appendix “Wide vs long data”</a>.  Instead, we will do it by hand.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>pivot_longer()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">species =</span> <span class="fu">rep</span>(iris<span class="sc">$</span>Species, <span class="dv">4</span>),</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">measurement =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Sepal.Length"</span>, <span class="st">"Sepal.Width"</span>,</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"Petal.Length"</span>, <span class="st">"Petal.Width"</span>), <span class="at">each=</span><span class="dv">150</span>),</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">value =</span> <span class="fu">c</span>(iris<span class="sc">$</span>Sepal.Length, iris<span class="sc">$</span>Sepal.Width,</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>            iris<span class="sc">$</span>Petal.Length, iris<span class="sc">$</span>Petal.Width)</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 600   3</code></pre>
</div>
</div>
<p>Make sure that you understand what is happening above. We are creating a data frame which has three columns: <code>species</code>, <code>measurement</code> and <code>value</code>; in the <code>value</code> column, we first put all the sepal lengths, then the sepal width etc. The resulting data frame is therefore four times longer than the initial one (with <span class="math inline">\(4 \times 150 = 600\)</span> rows), which is why we need to repeat the Species column four times<a href="#fn10" class="footnote-ref" id="fnref10" role="doc-noteref"><sup>10</sup></a>.</p>
<div class="no-row-height column-margin column-container"><div id="fn10"><p><sup>10</sup>&nbsp;If you have followed the “recycling” part on day 1 closely, you might think that it is not necessary to use <code>rep()</code> here – and you are absolutely right! However, I urge you to do such operations explicitely until you are very comfortable with R, because it is easy to make mistakes with recycling otherwise.</p></div></div><p>One possible solution is to assign the measurement to the <code>x position</code> channel, and assign the species to another channel, for example the <code>color</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>measurement, <span class="at">y=</span>value, <span class="at">color=</span>species)) <span class="sc">+</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/boxplot-all-variables-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="page-columns page-full"><p>OK, that is already not bad. But now let me now to introduce you to the <code>facet_wrap()</code> function. You add it to the plot like another layer, but what it does is very clever: it separates the plot into subplots based on an additional variable – in this case, we use the <code>measurement</code> column. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>facet_wrap()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>species, <span class="at">y=</span>value)) <span class="sc">+</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="st">"measurement"</span>, <span class="at">ncol=</span><span class="dv">4</span>, <span class="at">scales=</span><span class="st">"free_y"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/boxplot-all-variables-facet-wrap-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="page-columns page-full"><p>There is <em>a lot</em> one can do with <code>facet_wrap()</code> (and its cousin, <code>facet_grid()</code>).  Here we see two additional parameters. The <code>ncol</code> parameter specifies the number of columns – setting it to two would generate a <span class="math inline">\(2 \times 2\)</span> grid of subplots. The <code>scales="free_y"</code> parameter means that the y axis is not identical on all subplots – it makes sense here, because the different types of measurements are in different ranges; but in some other cases, you might want to omit it, or set the <span class="math inline">\(x\)</span> scales to be free instead.</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>facet_grid()</code></span></div></div>
<div class="page-columns page-full"><p>One more thing. You might be tempted to create a plot with the so-called bar plots. This might look something like this: </p><div class="no-row-height column-margin column-container"><span class="margin-aside">Bar plots</span></div></div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-17-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>I am sure you have seen a plot like that many times, and that you might be tempted to recreate it. <strong>Don’t.</strong> This plot is a terrible way to show continuous data. Bar plots in general should only be used in very special cases, like when you show proportions or absolute counts (say, number of votes). In all other cases you should use a boxplot combined with a violin plot or actual data points. If you want to know more about this, check <a href="appendix.more.stats.html#sec-barplots">this section in the Appendix</a>.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Show actual data
</div>
</div>
<div class="callout-body-container callout-body">
<p>If possible, always strive to show the actual data points (with <code>geom_beeswarm()</code> or <code>geom_jitter()</code>) in addition to the summary statistics. If not, at least show the distribution (with <code>geom_violin()</code>). Never use a bar plot.</p>
</div>
</div>
</section>
<section id="heatmaps" class="level3 page-columns page-full" data-number="5.3.5">
<h3 data-number="5.3.5" class="anchored" data-anchor-id="heatmaps"><span class="header-section-number">5.3.5</span> Heatmaps</h3>
<div class="page-columns page-full"><p>A very common type of figure in high throughput data setting and one which is hard to achieve with other tools is the heatmap. There are numerous packages and functions (including the base R function <code>heatmap()</code>), however we will use the <code>pheatmap()</code> function from the <code>pheatmap</code> package. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>pheatmap()</code></span></div></div>
<p>First, however, we need some data. For starters, we take our beloved iris data set; however we will chose only 10 flowers from each species. You haven’t learned the following idiom yet, but here is how to do it efficiently in tidyverse:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>iris_sel <span class="ot">&lt;-</span> iris <span class="sc">|&gt;</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>            <span class="fu">group_by</span>(Species) <span class="sc">|&gt;</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>            <span class="fu">slice_sample</span>(<span class="at">n=</span><span class="dv">10</span>) <span class="sc">|&gt;</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>            <span class="fu">ungroup</span>()</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(iris_sel<span class="sc">$</span>Species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    setosa versicolor  virginica 
        10         10         10 </code></pre>
</div>
</div>
<div class="page-columns page-full"><p>The <code>group_by()</code> function groups the data by the <code>Species</code> column, which means that the subsequent functions will affect each group separately. The <code>slice_sample()</code> is a specialized function only for this purpose – it randomly selects rows from a data frame. Finally we remove the grouping with the <code>ungroup()</code> function. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>group_by()</code>, <code>slice_sample()</code>, <code>ungroup()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>iris_mtx <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">as.matrix</span>(iris_sel[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]))</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(iris_mtx)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/pheatmap-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="page-columns page-full"><p>As you can see, we have first converted the first four columns of the matrix into a data frame, then transposed it (“flipped” so the rows become columns and vice versa) and finally plotted it with <code>pheatmap()</code>. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>pheatmap()</code>,<code>t()</code></span></div></div>
<p>However, we do not see the species information on the plot. We can add it using a special data frame that contains the species information. We will also add a few more parameters to the <code>pheatmap()</code> function to make the plot more readable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>iris_species <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">species=</span>iris_sel<span class="sc">$</span>Species)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>foo <span class="ot">&lt;-</span> <span class="fu">pheatmap</span>(iris_mtx, <span class="at">labels_col=</span>iris_sel<span class="sc">$</span>Species,</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">color =</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))(<span class="dv">100</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/pheatmap-2-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-output-formats" class="level3 page-columns page-full" data-number="5.3.6">
<h3 data-number="5.3.6" class="anchored" data-anchor-id="sec-output-formats"><span class="header-section-number">5.3.6</span> Output formats</h3>
<p>We will not be spending time here on the details of how to create different output formats for plots; however, there is one thing that we want to mention.</p>
<p>In general, there are two types of graphical files: raster (bitmap) and vector graphics. Raster graphics are made of pixels, like photos, and are good for complex images. Vector graphics are made of lines and shapes, like drawings, and are good for plots.</p>
<p>The two graphics below look identical, however, in the HTML version of this book, the left one is a raster image, and the right one is a vector graphics.</p>
<div class="columns">
<div class="column" style="width:45%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:45%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-23-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<p>Try magnifying these images – right click, select “open image in a new tab” (or similar) and then zoom in. You will see that the left image gets pixelated, while the right-hand one remains sharp:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/vector_vs_raster.png" class="img-fluid figure-img"></p>
<figcaption>Raster vs vector graphics</figcaption>
</figure>
</div>
<p>The advantage of vector graphics is not only that you can zoom in as much as you want and the image will not get pixelated. First and foremost, you can edit them using an appropriate tool. Many people use the commercial Adobe Illustrator for this task, but there are also free and open source tools like <a href="https://inkscape.org/">Inkscape</a>. Both these latter tools will do the job in a scientific setting – they will allow you to change the colors, the fonts, the line thickness, the text shown etc. of your plots.</p>
<p>Therefore, if possible, you should always save your plots in a vector format. You can always convert them to a raster format later, but once a plot is in a raster format, converting it to a vector format is practically impossible. The only exception is when you have an image with millions of data points – sometimes this may challenge your computer’s memory (vector graphics is more computationally intensive than raster graphics).</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Vector graphics
</div>
</div>
<div class="callout-body-container callout-body">
<p>If possible, use a vector graphic format for your plots. You can always convert them to a raster format later, while the opposite is not true.</p>
</div>
</div>
<p>R can produce vector graphics in the PDF and in the SVG format. Both formats can be edited in Inkscape or Illustrator, but the SVG format is also suitable for HTML pages, because SVG is a standard related to HTML – in fact, all the plots in this book are in SVG format.</p>
<p>To choose SVG or PDF format, you have the following options:</p>
<p><strong>Use Rstudio.</strong> In the right hand “Plots” tab on the lower right panel, you can click on “Export” and choose one of the available formats, including PDF or SVG.</p>
<p><strong>Using R markdown or Quarto.</strong> You can either use global document options or per-chunk options.</p>
<p>For global options, include the following code at the beginning of your R markdown file, just after the header:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r include=FALSE}</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="in">knitr::opts_chunk$set(dev="svg")</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Alternatively, insert this into the header:</p>
<pre><code>knitr:
  opts_chunk:
    dev: svg</code></pre>
<p>If you want to make sure that one specific chunk produces SVG, you can always set the option in the given chunk:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r dev="svg"}</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="in">plot(1:10)</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The only problem with this method is that if you create a Word document, the output will invariably be a raster image in your document.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Word and R plots
</div>
</div>
<div class="callout-body-container callout-body">
<p>When you create a Word document with Rmarkdown, <strong>never</strong> copy your plots from that file! Either create a standalone file (see below for instructions or use the “Export” button in Rstudio), or copy the SVG graphics from the HTML file.</p>
</div>
</div>
<p><strong>Directly creating graphics in your script.</strong> If you want your script to produce a file with the plot, you can do that in one of the many ways. Two of them are shown below, one using <code>svg()</code><a href="#fn11" class="footnote-ref" id="fnref11" role="doc-noteref"><sup>11</sup></a>, and the other one using <code>pdf()</code> or <code>cairo_pdf()</code><a href="#fn12" class="footnote-ref" id="fnref12" role="doc-noteref"><sup>12</sup></a>. Both commands need to be finalized with <code>dev.off()</code>:</p>
<div class="no-row-height column-margin column-container"><div id="fn11"><p><sup>11</sup>&nbsp;If you experience problems with SVG generated in R, you can try to use the <code>svglite()</code> function from the <code>svglite</code> package which produces a more standard-compliant SVG files.</p></div><div id="fn12"><p><sup>12</sup>&nbsp;The <code>cairo_pdf()</code> supports a wider range of characters, including Unicode characters, and allows font embedding. Long story short, use it if your fonts are garbled.</p></div></div><div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># producing an SVG image file</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">svg</span>(<span class="st">"test.svg"</span>, <span class="at">width=</span><span class="dv">14</span>, <span class="at">height=</span><span class="dv">7</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This produces an SVG file “test.svg” with the nominal size 14 x 7 inches (read more in the <a href="appendix-more-stats.html#sec-image-size">Appendix</a> regarding image and font sizes).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># produce a PDF file</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">"test.pdf"</span>, <span class="at">width=</span><span class="dv">14</span>, <span class="at">height=</span><span class="dv">7</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This produces a PDF file “test.pdf” with the nominal size 14 x 7 inches.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Vector graphics with Microsoft Office
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you rely heavily on Microsoft Office for your work, you might find another vector graphic format more useful – the Windows Meta File (WMF). The function is called <code>win.metafile()</code> and used in the same way as the <code>pdf()</code> function. The problems are that (a) you can only use it on Windows and (b) it does not support transparency.</p>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Remember!
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use SVG format for your plots whenever possible.</p>
</div>
</div>
</section>
</section>
<section id="basic-statistics-with-r" class="level2 page-columns page-full" data-number="5.4">
<h2 data-number="5.4" class="anchored" data-anchor-id="basic-statistics-with-r"><span class="header-section-number">5.4</span> Basic statistics with R</h2>
<section id="statistics-with-r" class="level3 page-columns page-full" data-number="5.4.1">
<h3 data-number="5.4.1" class="anchored" data-anchor-id="statistics-with-r"><span class="header-section-number">5.4.1</span> Statistics with R</h3>
<p>R is a powerful tool for statistics. It has a staggering number of packages that can be used for statistical analysis. I would venture the guess that if someone somewhere came up with a statistical method, then there is an R package for it<a href="#fn13" class="footnote-ref" id="fnref13" role="doc-noteref"><sup>13</sup></a>.</p>
<div class="no-row-height column-margin column-container"><div id="fn13"><p><sup>13</sup>&nbsp;Almost the same goes for bioinformatics and visualization.</p></div></div><p>One interesting fact about statistics is that is is <em>way</em> harder than programming in R. In fact, after some initial learning curve, you will find that it is much harder to understand, say, which statistical test you should use than how to actually run it in R. You will also see that it is harder to understand the plots that you have produced then to actually produce them.</p>
<p>Nonetheless, it takes a while to get into the “R mindset” of doing statistics. It is done very much differently than in UI-based software like Graphpad Prism or Excel.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-basic-statistics" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 5.6 (P values)</strong></span> P-values are one of the most basic concepts in statistics, and part of the language of science – it is hard to find a scientific paper without any p-values.</p>
<ul>
<li>Take a piece of paper or open a text editor and write down a one- or two-sentence explanation of what a p-value is.</li>
<li>Only when you have done this, read the section “Clarifications about p-values” in <a href="https://en.wikipedia.org/wiki/Misuse_of_p-values">this</a> Wikipedia article. Did you get it right? If yes, you are in the minority of scientists.</li>
</ul>
</div>
</div>
</div>
</div>
</section>
<section id="descriptive-statistics" class="level3 page-columns page-full" data-number="5.4.2">
<h3 data-number="5.4.2" class="anchored" data-anchor-id="descriptive-statistics"><span class="header-section-number">5.4.2</span> Descriptive statistics</h3>
<div class="page-columns page-full"><p>You have already seen some basic statiscs in R (see Day 1, for example  <a href="day1-intro-to-r.html#tip-functions" class="quarto-xref">Tip&nbsp;<span>1.1</span></a>). Here is a quick reminder of how to get some basic statistics for a vector of numbers:</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Descriptive statistics</span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">1000</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="co"># mean</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -0.03872947</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># median</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">median</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -0.04420835</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># standard deviation</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sd</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.048951</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Some basic statistics in one go</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
-3.49291 -0.76969 -0.04421 -0.03873  0.68598  3.19717 </code></pre>
</div>
</div>
<p>For a vector of numbers, the <code>summary()</code> function returns a vector with six numbers. Four of them should be self-explanatory: the minimum, median, mean, and maximum. The other two are the first and the third quartile.</p>
<div class="page-columns page-full"><p>The first number is the 1st quartile, also called “the lower quartile” or  25th percentile, which means that 25% of the data are below that number. The second is the 75th percentile (or the upper quartile), which means that 75% of the data are below the number. Together, between the first and the second number, there are 50% of the data. You can check it for yourself:</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Quartiles</span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>s <span class="ot">&lt;-</span> <span class="fu">summary</span>(x)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(x <span class="sc">&lt;</span> s[<span class="dv">2</span>]) <span class="sc">/</span> <span class="fu">length</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.25</code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(x <span class="sc">&gt;</span> s[<span class="dv">5</span>]) <span class="sc">/</span> <span class="fu">length</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.25</code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(x <span class="sc">&gt;</span> s[<span class="dv">2</span>] <span class="sc">&amp;</span> x <span class="sc">&lt;</span> s[<span class="dv">5</span>]) <span class="sc">/</span> <span class="fu">length</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.5</code></pre>
</div>
</div>
<div class="page-columns page-full"><p>Of particular interest is the so called interquartile range (IQR), which is the difference between the upper and lower quartile.  Just like the median is a robust, non-parametric measure of the “center” of the data, the IQR is a robust, non-parametric measure of the “spread” of the data. While median corresponds to the parametric mean, IQR corresponds to the standard deviation.</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Interquartile range</span></div></div>
<p>In non-normally distributed data, for example count data (when we count things) median and IQR are often more informative than mean and standard deviation. We will see that later when we come to visualizing data with bar plots.</p>
<div class="page-columns page-full"><p>You can get the IQR using the <code>IQR()</code> function: </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>IQR()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">IQR</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.45567</code></pre>
</div>
</div>
<p>For data frames, you can either use the <code>summary()</code> function or use the <code>skim()</code> function from the <code>skimr</code> package (see <a href="day3-reading-your-data.html#sec-diagnosing" class="quarto-xref"><span>Section 3.3.1</span></a> and <a href="day3-reading-your-data.html#sec-skim" class="quarto-xref"><span>Section 3.3.2</span></a>).</p>
</section>
<section id="simple-tests" class="level3 page-columns page-full" data-number="5.4.3">
<h3 data-number="5.4.3" class="anchored" data-anchor-id="simple-tests"><span class="header-section-number">5.4.3</span> Simple tests</h3>
<p>Most everyday statistical tests are available in base R without the need for loading any additional packages. Here we will show you some of them.</p>
<div class="page-columns page-full"><p><strong>t-test.</strong> The Student’s test is one of the most common tests in all statistics. It compares two groups of data. In its simplest form, the function <code>t.test()</code> takes two vectors of numbers. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>t.test()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># simulate two vectors with different means</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">15</span>, <span class="at">mean=</span><span class="dv">1</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">15</span>, <span class="at">mean=</span><span class="dv">3</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a><span class="co"># perform the t-test</span></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(a, b)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  a and b
t = -6.2053, df = 27.836, p-value = 1.082e-06
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -2.768422 -1.394013
sample estimates:
mean of x mean of y 
0.9100471 2.9912649 </code></pre>
</div>
</div>
<p>The actual output of the <code>t.test()</code> function is a list with a lot of information, but when printed to the console it shows a human-readable summary.</p>
<p>It is often useful to extract and format the p-value from the output of the <code>t.test()</code> function. You can save the t.test output to a variable and then access the element <code>p.value</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">t.test</span>(a, b)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>p.value</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.082117e-06</code></pre>
</div>
</div>
<div class="page-columns page-full"><p>To show the p-value in a more readable format, you can use the <code>format.pval</code>  function which converts the number to a string with the correct number of significant digits:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>format.pval(..., digits=2)</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">format.pval</span>(res<span class="sc">$</span>p.value, <span class="at">digits=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "1.1e-06"</code></pre>
</div>
</div>
<p>This allows us to include the p-value in a sentence, for example like this:</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>"The p-value of the t-test </span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>was <span class="in">`​r format.pval(res$p.value, digits=2)`</span>."</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This will be rendered as: “The p-value of the t-test was 1.1e-06.”</p>
<div class="page-columns page-full"><p>If we doo many tests or want to save the test results in a spreadsheet, it might be useful to use the <code>tidy()</code> function from the <code>broom</code> package. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>tidy()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(broom)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tidy</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 10
  estimate estimate1 estimate2 statistic    p.value parameter conf.low conf.high
     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;
1    -2.08     0.910      2.99     -6.21 0.00000108      27.8    -2.77     -1.39
# ℹ 2 more variables: method &lt;chr&gt;, alternative &lt;chr&gt;</code></pre>
</div>
</div>
<p>The <code>tidy()</code> function returns a data frame with the test results. It understands many different statistical tests, not only the t-test.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-t-test" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 5.7 (T-test)</strong></span> Use either the builtin iris dataset or the cleaned-up version of the doctored data set from Day 3. Perform a t-test to compare the sepal length between <em>Iris setosa</em> and <em>Iris versicolor</em>, as well as between <em>Iris versicolor</em> and <em>Iris virginica</em>.</p>
</div>
</div>
</div>
</div>
<div class="page-columns page-full"><p><strong>Wilcoxon test.</strong> The Wilcoxon test is a non-parametric version of the t-test, and almost as powerful. Use it via <code>wilcox.test()</code>: </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>wilcox.test()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wilcox.test</span>(a, b)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum exact test

data:  a and b
W = 9, p-value = 1.251e-06
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
</div>
<p>As you can see, the p-value is quite similar to the t-test, but we did not have to worry whether the data is normally distributed or not<a href="#fn14" class="footnote-ref" id="fnref14" role="doc-noteref"><sup>14</sup></a>.</p>
<div class="no-row-height column-margin column-container"><div id="fn14"><p><sup>14</sup>&nbsp;The t-test assumes normality, while the Wilcoxon test does not. However, despite that, the t-test is quite robust to non-normality, which means that it will work quite well even if the data is not quite normally distributed.</p></div></div><div class="page-columns page-full"><p><strong>Paired tests.</strong> Both the t-test and the Wilcoxon test can be used in a paired version. The paired variant is a special case of the regular t-test or Wilcoxon test where  the two groups are not independent. For example, you might have measured the same individuals before and after a treatment. <strong>You are not allowed to use a regular test if the data is paired</strong>, because one of the most fundamental and important assumptions is not met – the assumption of independence. Also, in many cases, using a paired test will give you more statistical power, as we will see shortly.</p><div class="no-row-height column-margin column-container"><span class="margin-aside">Paired t-test with <code>paired=TRUE</code></span></div></div>
<p>Consider this example. We first start with randomized vector <code>a</code>, then build vector <code>b</code> by adding a fixed effect and an error term:</p>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>boxplot()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">15</span>, <span class="at">mean=</span><span class="dv">1</span>, <span class="at">sd=</span><span class="dv">1</span>)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> a <span class="sc">+</span> <span class="fl">0.5</span> <span class="sc">+</span> <span class="fu">rnorm</span>(<span class="dv">15</span>, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span>.<span class="dv">5</span>)</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(a, b)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-34-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We see on the boxplot that the groups, if treated independently, are not really different. However, we can visualize it much better using a paired plot, in which a line is connecting the two values for each individual. We will use ggplot2 for that purpose:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">value=</span><span class="fu">c</span>(a, b),</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">group=</span><span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>), <span class="at">each=</span><span class="dv">15</span>),</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">id=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>)</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>group, <span class="at">y=</span>value, <span class="at">group=</span>id)) <span class="sc">+</span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-35-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="page-columns page-full"><p>The new parameter here is the group aesthetic, <code>group</code>, which tells ggplot2 that the points which have identical <code>id</code> belong together. The geoms, such as <code>geom_line()</code> which connects data points with lines, work group-by-group, and therefore lines connect only points within one group. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>geom_line()</code>, <code>group</code> aesthetic</span></div></div>
<p>On this plot we can clearly see that for the majority of individuals, the value in the second measurement (group B) is higher than in the first measurement (group A). We can confirm this with a paired t-test:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(a, b, <span class="at">paired=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Paired t-test

data:  a and b
t = -4.1849, df = 14, p-value = 0.0009172
alternative hypothesis: true mean difference is not equal to 0
95 percent confidence interval:
 -0.7638627 -0.2461963
sample estimates:
mean difference 
     -0.5050295 </code></pre>
</div>
</div>
<p>The <code>paired=TRUE</code> argument tells R that the two vectors are paired. We get a p-value of 0.00092, whereas with regular t-test we would have gotten a p-value of 0.21. Similarly, if you were to run the Wilcoxon paired test, the p value would be 0.0034, whereas in the regular Wilcoxon test it would be 0.17.</p>
<div class="page-columns page-full"><p><strong><span class="math inline">\(\chi^2\)</span> (Chi-square) test.</strong> If we have two categorical vectors, we can  use the <span class="math inline">\(\chi^2\)</span> test. Consider the results of the gene expression analysis we have looked at yesterday. We can define two vectors: “significant/non-significant” and “interferon/noninterferon”. These can be logical or character vectors, it doesn’t matter:</p><div class="no-row-height column-margin column-container"><span class="margin-aside"><span class="math inline">\(\chi^2\)</span> test</span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>tr_res <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"Datasets/transcriptomics_results.csv"</span>)</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>tr_res <span class="ot">&lt;-</span> <span class="fu">select</span>(tr_res, </span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">Gene=</span>GeneName, Description,</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">logFC=</span>logFC.F.D1, <span class="at">FDR=</span>qval.F.D1)</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>interferon <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"interferon"</span>, tr_res<span class="sc">$</span>Description)</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>significant <span class="ot">&lt;-</span> tr_res<span class="sc">$</span>FDR <span class="sc">&lt;</span> <span class="fl">0.01</span></span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(significant, interferon)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           interferon
significant FALSE  TRUE
      FALSE 40725    59
      TRUE    834    21</code></pre>
</div>
</div>
<div class="page-columns page-full"><p>Above we constructed a <em>contingency</em> table. Rows show the two  values of <code>significant</code> vector, columns – same for the <code>interferon</code> vector. In other words, most (40725) genes are neither significant nor interferons. However, out of 855 significant genes, as much as 21 have “interferon” in their description. And vice verse, out of the 80 interferon genes, more than a third are significant. But is this <em>statistically</em> significant? To answer this, we can use the <code>chisq.test()</code>. </p><div class="no-row-height column-margin column-container"><span class="margin-aside">Contingency table with <code>table(x, y)</code></span><span class="margin-aside"><code>chisq.test()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">chisq.test</span>(significant, interferon)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Pearson's Chi-squared test with Yates' continuity correction

data:  significant and interferon
X-squared = 221.44, df = 1, p-value &lt; 2.2e-16</code></pre>
</div>
</div>
<p>Yep, looks like it. In fact, the above was a simple case of gene set enrichment analysis.</p>
<p>With this, we conclude this simple statistics part. If you are interested in more, take a look at the <a href="appendix-more-stats.html">Appendix: more stats and visualizations</a> – I have included a few quite common statistical applications there.</p>
</section>
</section>
<section id="pca-and-scatter-plots" class="level2 page-columns page-full" data-number="5.5">
<h2 data-number="5.5" class="anchored" data-anchor-id="pca-and-scatter-plots"><span class="header-section-number">5.5</span> PCA and scatter plots</h2>
<section id="principal-component-analysis" class="level3 page-columns page-full" data-number="5.5.1">
<h3 data-number="5.5.1" class="anchored" data-anchor-id="principal-component-analysis"><span class="header-section-number">5.5.1</span> Principal component analysis</h3>
<p>In this last section we will combine <em>a lot</em> of things (and introduce a few new concepts, despite it being Friday afternoon<a href="#fn15" class="footnote-ref" id="fnref15" role="doc-noteref"><sup>15</sup></a>).</p>
<div class="no-row-height column-margin column-container"><div id="fn15"><p><sup>15</sup>&nbsp;Assuming, of course, that you started this course on a Monday…</p></div></div><div class="page-columns page-full"><p>One of the basic plots in statistics is the scatter plot, with two continuous variables. They are easy enough to generate in R, but we will do it with a twist, introducing you to a new statistical concept: <em>principal component analysis</em> (PCA). </p><div class="no-row-height column-margin column-container"><span class="margin-aside">PCA, Principal component analysis</span></div></div>
<p>Rember how we plotted the iris data set two days ago (<a href="day3-reading-your-data.html#sec-quick-look" class="quarto-xref"><span>Section 3.5.5</span></a>)? It was easy to see the differences between the groups, and if we wanted, we could have plotted all four variables on two plots. But what if we had thousands of variables? For example, expression of thousands of genes for hundreds of samples?</p>
<p>One of the possible approaches is to use a technique like PCA. PCA converts the data replacing the original variables with a new, smaller set, that however covers all the variance in the data. While the old variables are clearly defined, the new ones, called “principal components”, result from combining them, so they do not correspond to something specific<a href="#fn16" class="footnote-ref" id="fnref16" role="doc-noteref"><sup>16</sup></a>. However, they have a few nice properties:</p>
<div class="no-row-height column-margin column-container"><div id="fn16"><p><sup>16</sup>&nbsp;Components are the linear combinations of the original variables. So for a given sample, we basically add up expression of all genes, but each gene has a different weight.</p></div></div><ul>
<li>Most of the variance of the samples sits in the first components, then the next biggest share sits in the second etc. That means that by looking at the first few we are likely to get a very good idea of the overall differences in the data.</li>
<li>The components are orthogonal, that is, they are not correlated with each other. This means that if, say, two groups correspond to one component, then it is quite likely that they will <em>only</em> correspond to that component. And thus we are able to say: “this component explains differences between treatments, and this explains gender”.</li>
<li>The analysis is unsupervised, that is, we do not need to tell the algorithm which samples belong to which group. It will find the differences on its own (if there are any).</li>
</ul>
</section>
<section id="the-data" class="level3 page-columns page-full" data-number="5.5.2">
<h3 data-number="5.5.2" class="anchored" data-anchor-id="the-data"><span class="header-section-number">5.5.2</span> The data</h3>
<p>In the following, we will be using a set of measurements of laboratory values (such as white blood cell count, hemoglobin, etc.) from a clinical vaccination study. There were three treatments - placebo and two vaccines, one with an adjuvant (Fluad) and one without (Agrippal), and samples were taken on different time points, starting with a screening time point and base line, then on to D0, D1 etc. The data is in the <em>wide</em> format, that is, each row corresponds to a single sample – that is, a single patient and a single time point, and each column corresponds to a single measurement.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>labdata <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"Datasets/labresults_wide.csv"</span>)</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(labdata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1732   31</code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(labdata[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 10
  SUBJ.TP       ACA   ALB   ALP   ALT  BILI    CA CREAT   CRP   GGT
  &lt;chr&gt;       &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 4081.SCREEN  2.24  46.1  79.4  25.7  14.6  2.37  83.0 0.224  27.9
2 4081.BL      2.27  43.8  77.7  24.0  10.1  2.36  77.0 0.352  26.3
3 4081.D0      2.28  44.1  76.2  21.0  15.4  2.36  80.4 0.208  24.8
4 4081.D1      2.25  42.0  74.2  22.4  23.6  2.31  74.5 0.452  24.7
5 4081.D2      2.18  39.9  65.9  22.2  20.3  2.18  75.0 0.342  21.6
6 4081.D3      2.15  37.8  62.8  20.5  18.3  2.12  69.7 0.333  20.4</code></pre>
</div>
</div>
<p>The last line above uses <code>head()</code> to show the first few lines of the data frame, and with <code>[,1:10]</code> we limit the output to the first ten columns<a href="#fn17" class="footnote-ref" id="fnref17" role="doc-noteref"><sup>17</sup></a>.</p>
<div class="no-row-height column-margin column-container"><div id="fn17"><p><sup>17</sup>&nbsp;If you want to know more on what the abbreviations mean, take a look at the <code>labresults_full.csv</code> file. This is the down-side of the wide format: no place to store extra information about the columns!</p></div></div><div class="page-columns page-full"><p>As you can see, the first column identifies the sample by giving both the subject and the time point. In the following, we only want one time point (D1), and we could filter the data set by combining <code>filter()</code> with a suitable <code>grepl()</code> call, for example, <code>filter(labdata, grepl("D1$", TP))</code>. Instead, we will use this as an opportunity to show you the <code>separate()</code> function. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>separate()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>labdata <span class="ot">&lt;-</span> labdata <span class="sc">|&gt;</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(SUBJ.TP, <span class="at">into=</span><span class="fu">c</span>(<span class="st">"SUBJ"</span>, <span class="st">"TP"</span>), <span class="at">sep=</span><span class="st">"</span><span class="sc">\\</span><span class="st">."</span>) <span class="sc">|&gt;</span></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(TP <span class="sc">==</span> <span class="st">"D1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>separate()</code> function takes a column name, the names of the new columns and a separator. The separator is a regular expression, and since the dot normally matches any character, we need to escape it with two backslashes. The resulting data frame has two columns named <code>SUBJ</code> and <code>TP</code> instead of one called <code>SUBJ.TP</code>, and we could directly filter the data frame for the desired time point.</p>
<div class="page-columns page-full"><p>There is one more thing that we need to take care of, unfortunately. There are some <code>NA</code> values in the data set, and we need to remove them. Rather than figure out how to handle them, we will simply remove all samples that have a missing value anywhere in the data set with one function: </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>drop_na()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>labdata <span class="ot">&lt;-</span> labdata <span class="sc">|&gt;</span> <span class="fu">drop_na</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The problem we are facing <em>now</em> is that the data set does not contain any interesting meta-data, like any information about the actual treatment group! Lucky for us, we have that information in another file. The <code>expression_data_vaccination_example.xlsx</code> file contains the meta-data (as well as matching RNA expression data, but we will not use that here for now).</p>
<p>We will combine both data sets using an inner join with <code>merge()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read the meta-data</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="st">"Datasets/expression_data_vaccination_example.xlsx"</span>,</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">sheet=</span><span class="st">"targets"</span>) <span class="sc">|&gt;</span></span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">filter</span>(Timepoint <span class="sc">==</span> <span class="st">"D1"</span>)</span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a>                   </span>
<span id="cb75-7"><a href="#cb75-7" aria-hidden="true" tabindex="-1"></a>combined <span class="ot">&lt;-</span> <span class="fu">merge</span>(meta, labdata, <span class="at">by=</span><span class="st">"SUBJ"</span>)</span>
<span id="cb75-8"><a href="#cb75-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(combined[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  SUBJ          USUBJID Batch       ID Timepoint ARMCD      ARM AGE SEX PLACEBO
1 4023  CRC987X-4023-17     C F4023.D1        D1     F    FLUAD  22   F   FALSE
2 4028 CRC987X-4028-120     C A4028.D1        D1     A AGRIPPAL  29   F   FALSE
3 4034 CRC987X-4034-128     C A4034.D1        D1     A AGRIPPAL  17   M   FALSE
4 4066 CRC987X-4066-222     A P4066.D1        D1     P  PLACEBO  28   M    TRUE
5 4119 CRC987X-4119-117     C F4119.D1        D1     F    FLUAD  27   F   FALSE
6 4122 CRC987X-4122-185     A P4122.D1        D1     P  PLACEBO  26   F    TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(combined)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 60 41</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Check your results!
</div>
</div>
<div class="callout-body-container callout-body">
<p>Check your results frequently. Do you get the number of rows you expect? How does the data look like? Use <code>dim()</code>, <code>head()</code>, <code>tail()</code>, <code>summary()</code>, <code>View()</code> all the time.</p>
</div>
</div>
<p>To run the PCA, we need, however, only the numeric columns. If you take a look at the <code>combined</code> data frame with, for example, <code>View</code>, you will see that we need the columns from <code>ACA</code> (adjusted calcium) to <code>WBC</code> (white blood cell count). We will select these columns and convert the resulting data frame to a matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>combined_mtx <span class="ot">&lt;-</span> <span class="fu">select</span>(combined, ACA<span class="sc">:</span>WBC) <span class="sc">|&gt;</span> </span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>()</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(combined_mtx[,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          ACA      ALB      ALP      ALT      BILI
[1,] 2.049224 40.83864 72.74828 18.60445  9.781505
[2,] 2.124255 41.17438 61.13931 15.05414  6.878661
[3,] 2.203683 41.05199 56.41416 19.88788 11.611140
[4,] 2.160399 39.91228 40.65232 32.52709 12.058742
[5,] 2.100984 38.20202 69.14200 35.36937  9.516149
[6,] 2.198283 38.35306 81.23705 10.83461  8.167515</code></pre>
</div>
</div>
</section>
<section id="running-the-pca" class="level3 page-columns page-full" data-number="5.5.3">
<h3 data-number="5.5.3" class="anchored" data-anchor-id="running-the-pca"><span class="header-section-number">5.5.3</span> Running the PCA</h3>
<div class="page-columns page-full"><p>The PCA is a built-in function in R, and it is called <code>prcomp()</code>. We will only use a single parameter, <code>scale.</code> (mind the dot!) to tell PCA to scale the gene expression before working its magic: </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>prcomp()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co"># the actual PCA is just one line!</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(combined_mtx, <span class="at">scale.=</span><span class="cn">TRUE</span>)</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.list</span>(pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "sdev"     "rotation" "center"   "scale"    "x"       </code></pre>
</div>
</div>
<p>The object returned by <code>prcomp()</code> is a list, and the element that we are particularly interested in is called <code>x</code>. This is a matrix of the principal components, PC1, PC2 etc. There are quite a few principal components, and each one has as many elements as there are samples in the data set:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(pca<span class="sc">$</span>x)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "PC1"  "PC2"  "PC3"  "PC4"  "PC5"  "PC6"  "PC7"  "PC8"  "PC9"  "PC10"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(pca<span class="sc">$</span>x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 60 30</code></pre>
</div>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(combined)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 60 41</code></pre>
</div>
</div>
<p>As you can see, the number of rows of <code>pca$x</code> is the same as the number of samples in the <code>combined</code> data frame. In other words, for each sample, we have a bunch of new values that are the principal components.</p>
<div class="page-columns page-full"><p>We won’t be interested in more than the first few components, but we want to see them in connection with the covariates. Therefore, we will use <code>cbind()</code> to combine the <code>combined</code> data frame with the first few columns of the <code>pca$x</code> matrix. We will use the <code>ggplot2</code> package to plot the data. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>cbind()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>pca_df <span class="ot">&lt;-</span> <span class="fu">cbind</span>(combined, pca<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pca_df, <span class="fu">aes</span>(<span class="at">x=</span>PC1, <span class="at">y=</span>PC2, <span class="at">color=</span>SEX)) <span class="sc">+</span></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">cex=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-47-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="page-columns page-full"><p>On the plot above, we have plotted the first two principal components, PC1 and PC2. You can clearly see that the first component separates the samples by sex: males have a different value of PC1 then females. We can see it clearly on a boxplot: </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>geom_boxplot()</code></span></div></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pca_df, <span class="fu">aes</span>(<span class="at">x=</span>SEX, <span class="at">y=</span>PC1)) <span class="sc">+</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.shape=</span><span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-48-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div class="page-columns page-full"><p>However, because <code>geom_boxplot()</code> <em>also</em> puts a dot where outliers are, we use <code>outlier.shape=NA</code> to suppress them. </p><div class="no-row-height column-margin column-container"><span class="margin-aside"><code>geom_jitter()</code></span></div></div>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-pca" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 5.8 (Vaccines)</strong></span> Repeat the plot above, but instead of <code>SEX</code>, use the <code>ARM</code> column which corresponds to the study arm, i.e.&nbsp;which vaccine (or placebo) was administered. Can you tell the groups apart? Try it with other components: PC3 vs PC4, PC5 vs PC6 etc. Any luck?</p>
<div class="callout callout-style-default callout-caution no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-12-contents" aria-controls="callout-12" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-12" class="callout-12-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Yes, the PC4 seems to show differences between the three groups, with Fluad having the lowest values, and Placebo the highest.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pca_df, <span class="fu">aes</span>(<span class="at">x=</span>PC3, <span class="at">y=</span>PC4, <span class="at">color=</span>ARM)) <span class="sc">+</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">cex=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-49-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pca_df, <span class="fu">aes</span>(<span class="at">x=</span>ARM, <span class="at">y=</span>PC4)) <span class="sc">+</span></span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.shape=</span><span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-49-2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="interpreting-the-pca" class="level3" data-number="5.5.4">
<h3 data-number="5.5.4" class="anchored" data-anchor-id="interpreting-the-pca"><span class="header-section-number">5.5.4</span> Interpreting the PCA</h3>
<p>OK, but <em>what does that mean?</em> How does the PCA “know” who is male, and who is female? Where does this information come from?</p>
<p>As mentioned before, every principal component is derived from the actual values of the different variables. The values are summed up for each component, but not every variable contributes equally. We can wee how each value contributes to the component by directly looking at another member of the <code>pca</code> object, <code>rotation</code>. Again, it is a matrix, but now each row corresponds to a variable, and each column to a principal component:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pca<span class="sc">$</span>rotation[,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             PC1          PC2         PC3         PC4          PC5
ACA  -0.17887595  0.330789811 -0.12168346  0.08619332 -0.017461193
ALB  -0.26645587  0.054470599 -0.18559187 -0.20262134 -0.061104850
ALP  -0.08216838  0.139728838  0.06991512 -0.16908259 -0.006265994
ALT  -0.12070502 -0.008253212  0.16125243  0.11857842  0.390570215
BILI -0.16847944 -0.233759742 -0.01457699 -0.26068360 -0.042323962
CA   -0.26605009  0.280096277 -0.17364117 -0.02687325 -0.047787207</code></pre>
</div>
</div>
<p>These numbers are <em>loadings</em>, that is, the weights of each variable in the calculation of the given principal component. If the loading is close to <code>0</code>, then the given variable does not contribute much to the component. If, however, it is very large, or very small (negative), then the variable contributes a lot.</p>
<p>To understand why the male samples are separated from the female samples on the plot above, we will focus on PC1 and sort it by the decreasing absolute loadings.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>pc1 <span class="ot">&lt;-</span> pca<span class="sc">$</span>rotation[,<span class="dv">1</span>]</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>ord <span class="ot">&lt;-</span> <span class="fu">order</span>(<span class="fu">abs</span>(pc1), <span class="at">decreasing=</span><span class="cn">TRUE</span>)</span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pc1[ord])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       HGB        HCT        RBC        ALB         CA      CREAT 
-0.3645930 -0.3362286 -0.3076459 -0.2664559 -0.2660501 -0.2550704 </code></pre>
</div>
</div>
<p>First value that we find is HGB with a value of -0.36. The fact that it is negative means that when the actual value of the variable is high, the resulting PC1 will be low and vice versa. That is, samples with a high value of this variable will be likely to be shown on the left side of the plot, and samples with a low value on the right side.</p>
<p>Now, male samples are on the left side of the plot, and female samples on the right side. What is HGB? It turns out that it corresponds to hemoglobin. Since the loading is low, it means that males should have a low value of hemoglobin compared to females. We can check it with a boxplot:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day5-visualization-and-statistics_files/figure-html/unnamed-chunk-53-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Indeed, that seems to be the case! And, of course, it makes perfect sense biologically.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-pca" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 5.9 (Other top loadings)</strong></span> In the table below, you will find three more variables. Check their loadings. Make corresponding box plots and check if the values correspond to what you know about human biology.</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Variable</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ESR</td>
<td>Erythrocyte sedimentation rate</td>
</tr>
<tr class="even">
<td>CREAT</td>
<td>Creatinine</td>
</tr>
<tr class="odd">
<td>CA</td>
<td>Calcium</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<p>The purpose of this exercise was not only to show you the mechanics of PCA. This <em>is</em> a useful method, and quite often it is worthwile to run it before you do anything else with the data. PCA shows you the possible groupings even before you even <em>start</em> to think about the actual analysis.</p>
<p>However, another point of this whole story was to demonstrate an important aspect of bioinformatic analyses. Think how much work it was to pummel the data into a shape which was suitable for the PCA, and how much work to actually figure out what it means – compared with the actual analysis, which was just a single line of code. This is a common situation in bioinformatics: the actual analysis is easy, but the data preparation and interpretation are hard and time consuming.</p>
<div class="callout callout-style-simple callout-caution no-icon">
<div class="callout-body d-flex">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-body-container">
<div id="exr-pca" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 5.10 (Top loadings for vaccines)</strong></span> Which PC is the most important for separating the vaccine groups? Which variables contribute the most to this component? Can you explain why?</p>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="afterword" class="level2" data-number="5.6">
<h2 data-number="5.6" class="anchored" data-anchor-id="afterword"><span class="header-section-number">5.6</span> Afterword</h2>
<section id="where-to-get-r-packages" class="level3" data-number="5.6.1">
<h3 data-number="5.6.1" class="anchored" data-anchor-id="where-to-get-r-packages"><span class="header-section-number">5.6.1</span> Where to get R packages</h3>
<p>There are three main sources of R packages: <a href="https://cran.r-project.org">CRAN</a>, <a href="https://bioconductor.org">Bioconductor</a> and <a href="https://github.com">github</a>.</p>
<p><strong>CRAN</strong> is the resource you are using when using <code>install.packages()</code>. It features more than 20,000 packages from all possible areas of science, statistics and even entertainment. Packages in CRAN are rigorously tested for formal requirements: they must have a certain structure, all functions and data sets must be documented, and the package must pass a series of tests, including testing whether the packages can be installed on a variety of operating systems. However, the packages are never tested for their actual correctness, i.e.&nbsp;whether what they claim to do is what they actually do. There is no manual review process involved.</p>
<p>Here are a few packages on CRAN that might be of interest to you at this stage:</p>
<ul>
<li><a href="https://cran.r-project.org/web/packages/skimr/vignettes/skimr.html">colorDF</a>) – a package that I wrote for displaying colorized data frames in the console. It has also a summary function that provides information similar to that of <code>skimr::skim()</code> (<code>colorDF::summary_colorDF()</code>).</li>
<li><a href="https://vincentarelbundock.github.io/tinytable/">tinytable</a> – package for generating awesome tables with R and R markdown.</li>
<li><a href="https://github.com/rstudio/gt">gt</a> – another package for creating tables in R markdown; they can be really complex and have a professional look. Good for manuscripts and publications.</li>
<li><a href="https://github.com/elbersb/tidylog">tidylog</a> – a package for logging your data manipulation steps; makes functions like <code>filter()</code> more explicit.</li>
<li><a href="https://github.com/jalvesaq/colorout">colorout</a> – adds colors to your console.</li>
<li><a href="https://cran.r-project.org/web/packages/formatdown/vignettes/format_numbers.html">formatdown</a> – a bunch of functions to properly format numbers and more in Rmarkdown/Quarto, similar to the <code>format.pval()</code> base R function but better looking (e.g.&nbsp;<span class="math inline">\(6.022 \times 10^{23}\)</span>).</li>
<li><a href="https://claudiozandonella.github.io/trackdown/">trackdown</a> is an attempt to allow multi-user collaboration on R markdown documents. You work in R markdown / Quarto on your computer, but then upload a document to Google Docs, which is then edited by your collaborators; and then you can download the changes and apply them to your R markdown document automatically. Unfortunately, the hardest part is to set up the authentication with Google.</li>
<li><a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> – a package that helps you manage your R environment, including the packages you have installed. The idea is that rather than using a single global library on your computer, you install necessary packages into a local, project-specific library. This has the advantage that for a given project, you always work with the same versions of the packages, and helps to ensure reproducibility of your code.</li>
<li><a href="https://r-graph-gallery.com/38-rcolorbrewers-palettes.html">RColorBrewer</a> – a very popular set of color palettes for R graphics. You can use them directly in <code>ggplot2</code> by calling <code>scale_fill_brewer()</code> or <code>scale_color_brewer()</code>.</li>
</ul>
<p><strong>Bioconductor</strong> is both a framework and repository for packages which are mostly related to bioinformatics and biomedicine. The framework differs somewhat from, say, <code>tidyverse</code>, so that at times you will run into problems; for example, the <code>filter()</code> function from <code>tidyverse</code> will not work on the <code>DataFrame</code> object from <code>Bioconductor</code>; and vice versa, loading a Bioconductor package may mask the <code>filter()</code> function from tidyverse with its own variant (that works in a different way), causing infuriating errors in your code.</p>
<p>Bioconductor packages are rigorously checked and manually reviewed, and thus provide a higher level of quality assurance than CRAN packages.</p>
<p>To install a Bioconductor package, you need to install the <code>BiocManager</code> package from CRAN, and then use the <code>BiocManager::install()</code> function to actually install your packages.</p>
<p><strong>Github</strong> is a general repository for code, not only R but also many other languages. Anyone can create a repository on Github and publish their R package there. In fact, for many packages on CRAN or in Bioconductor there exists a Github repository that contains also development versions of the packages. Github is simply very convenient for developing software – and not only software. If you are reading this book, chances are that you are doing so over Github.</p>
<p>Installing github packages may depend on the specific package, but usually you can use the <code>install_github()</code> available when you load the <code>devtools</code> package. For example, to install the development version of <code>colorDF</code>, you can type <code>devtools::install_github("january3/colorDF")</code> (in this case, it is the same version as the one on CRAN).</p>
<p>There is no formalized check for the quality of the packages on Github. Individual developers may or may not include some QC, but there is no formal requirement for it. Therefore, the software quality may vary considerably.</p>
</section>
<section id="where-to-go-from-here" class="level3" data-number="5.6.2">
<h3 data-number="5.6.2" class="anchored" data-anchor-id="where-to-go-from-here"><span class="header-section-number">5.6.2</span> Where to go from here</h3>
<p>There are numerous resources for R on the web, allow me to recommend a few.</p>
<ul>
<li><a href="https://stackoverflow.com">StackOverflow</a> – this is a general Q &amp; A forum for programmers and other like them. Most of the questions you have someone has previously asked here, including many by yours truly.</li>
<li>Online R books
<ul>
<li><a href="https://rstudio-education.github.io/hopr/">Hands-On Programming with R</a> by Garrett Grolemund. A very different approach to teachin R to beginners than the one that I prefer, but maybe you will find it better. Worth checking in any case.</li>
<li><a href="https://r4ds.hadley.nz/">R for Data Science (tidyverse)</a> by Hadley Wickham. The tidyverse book, written by the main tidyverse developer. This book also starts with introducing R.</li>
<li><a href="https://r-graphics.org/">R Graphics Cookbook</a> by Winston Chang. The cookbook format splits the information in small “recipies” for common problems, both in base R and in ggplot2.</li>
<li><a href="https://bookdown.org/yihui/rmarkdown/">R markdown: the definitive guide</a> by Yihui Xie, J. J. Allaire and Garrett Grolemund. A comprehensive guide to R markdown written by the people who created it.</li>
<li><a href="https://hbiostat.org/rflow/">R Workflow</a> by Frank Harrell is somewhat hermetic, but there are many good ideas in it and a lot about reproducible scientific work with Quarto.</li>
<li><a href="https://cran.r-project.org/manuals.html">R manuals</a> from CRAN. Be</li>
<li>warned – they tend to be rather comprehensive.</li>
</ul></li>
<li>Other books:
<ul>
<li><a href="https://www.google.com/search?q=the+r+book">The R Book</a> by Michael J. Crawley – this is a magnificent book on statistics. It uses very conservative R language (no tidyverse at all), but it discusses at length even the more complex statistical issues. I recommend this book to every person willing to learn statistics with R.</li>
<li><a href="https://compgenomr.github.io/book/">Computational Genomics with R</a> by Altuna Akalin (available online) – this is a book on bioinformatics with R and BioConductor. It features both an introduction to genomics and a terse introduction to R. The R flavor presented is somewhat different from the one I prefer (for example, the author uses <code>=</code> instead of <code>&lt;-</code>), but the main focus of the book is elsewhere – on specific bioinformatics tasks.</li>
<li><a href="https://avehtari.github.io/ROS-Examples/">Regression and other stories</a> by Andrew Gelman, Jennifer Hill and Aki Vehtari. This book focuses on doing bayesian statistics with the RStan package. It is both, a very good introduction to R and a very good introduction to statistics, with thoughtful explanations and a lot of examples. However, keep in mind that the focus of the book is <em>bayesian</em> statistics.</li>
</ul></li>
<li>Large Language Models (LLMs): I have good experiences with LLMs (ChatGPT, Perplexity AI) for learning programming languages. While complex tasks may be out of reach for them, they are very good at explaining basics. “How do I do X in R?” or “Why doesn’t the following code work?” seem to work quite well. RStudio allows you also to use <a href="https://docs.posit.co/ide/user/ide/guide/tools/copilot.html">Copilot</a> which is an LLM model that watches the code you write and tries to guess what you are trying to do. Just remember, that the one thing that LLMs don’t know how to say is “I don’t know”. If something doesn’t make sense, or if it is beyond the scope of their learning, they will “hallucinate” – give good sounding advice which is totally bonkers.</li>
</ul>
</section>
<section id="famous-last-words" class="level3" data-number="5.6.3">
<h3 data-number="5.6.3" class="anchored" data-anchor-id="famous-last-words"><span class="header-section-number">5.6.3</span> Famous last words</h3>
<p>We are now at the end of Five Days of R, but your journey just started. Here is my last advice to you.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Get on with R
</div>
</div>
<div class="callout-body-container callout-body">
<p>Start working with R, right now, for all your projects; for statistics, data management and even preparing reports and documents.</p>
<p>At first, doing the same task in R will take much more time then the same task in programs you used before. You will feel that you are wasting your time.</p>
<p>You are not. Sooner then you think, it will pay off.</p>
</div>
</div>
</section>
</section>
<section id="review" class="level2" data-number="5.7">
<h2 data-number="5.7" class="anchored" data-anchor-id="review"><span class="header-section-number">5.7</span> Review</h2>
<ul>
<li>R markdown / Quarto
<ul>
<li>creating R markdown documents</li>
<li>basic markdown formatting (<strong>bold</strong>, <em>italic</em> etc.)</li>
<li>including R chunks</li>
<li>creating output in different formats</li>
<li>choosing an output format</li>
</ul></li>
<li>Visualisations
<ul>
<li>basic scatterplots with <code>ggplot()</code></li>
<li>understanding esthetics (<code>aes()</code>)</li>
<li>box plots with <code>geom_boxplot()</code> and violin plots with <code>geom_violin()</code></li>
<li>heatmaps</li>
<li>vector vs raster graphics</li>
</ul></li>
<li>Basic statistics
<ul>
<li>t-test and Wilcoxon test</li>
<li>paired tests</li>
<li><span class="math inline">\(\chi^2\)</span> test</li>
<li>principal component analysis with <code>prcomp()</code></li>
</ul></li>
<li>Running a principal component analysis</li>
</ul>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-weiner2022venn" class="csl-entry" role="listitem">
Weiner 3rd, January, Benedikt Obermayer, and Dieter Beule. 2022. <span>“Venn Diagrams May Indicate Erroneous Statistical Reasoning in Transcriptomics.”</span> <em>Frontiers in Genetics</em> 13: 818683. <a href="https://www.frontiersin.org/journals/genetics/articles/10.3389/fgene.2022.818683/full">https://www.frontiersin.org/journals/genetics/articles/10.3389/fgene.2022.818683/full</a>.
</div>
</div>
</section>


</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./day4-manipulating-your-data.html" class="pagination-link" aria-label="Manipulating data frames">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Manipulating data frames</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./appendix-wide-long.html" class="pagination-link" aria-label="Appendix: Wide vs long data">
        <span class="nav-page-text">Appendix: Wide vs long data</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>